{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'PATH_TO_FEATURES' from 'config' (/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/config.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[1;32m/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb Cell 1'\u001b[0m in \u001b[0;36m<cell line: 23>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=19'>20</a>\u001b[0m PATH_TO_GDSC_DATASETS \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m../../datasets/gdsc/\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=20'>21</a>\u001b[0m FILENAME_LANDMARK_GENES \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39mlandmark_genes.csv\u001b[39m\u001b[39m'\u001b[39m \n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=22'>23</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mconfig\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=23'>24</a>\u001b[0m     \u001b[39m# Drug features.\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=24'>25</a>\u001b[0m     PATH_TO_SAVED_DRUG_FEATURES,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=25'>26</a>\u001b[0m     DRUG_FPS_FINAL_FILE_NAME,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=26'>27</a>\u001b[0m     \u001b[39m# Cell line features.\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=27'>28</a>\u001b[0m     PATH_TO_SAVED_CL_FEATURES,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=28'>29</a>\u001b[0m     GENE_EXPR_FINAL_FILE_NAME,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=29'>30</a>\u001b[0m     CNV_GISTIC_FINAL_FILE_NAME,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=30'>31</a>\u001b[0m     CNV_PICNIC_FINAL_FILE_NAME,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=31'>32</a>\u001b[0m     PATH_TO_FEATURES    \n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/04_v2_mutations.ipynb#ch0000000?line=32'>33</a>\u001b[0m )\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'PATH_TO_FEATURES' from 'config' (/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/config.py)"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import time\n",
    "import re\n",
    "import copy\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_theme(style=\"white\")\n",
    "\n",
    "PATH_TO_CNV_DATA = '../../datasets/gdsc/cell_line_genomic_data/'\n",
    "PATH_TO_SAVE_DATA_TO = '../../datasets/gdsc/my_datasets/'\n",
    "\n",
    "MUTATION_FILE = 'mutations_all_20220315.csv'\n",
    "\n",
    "PATH_TO_GDSC_DATASETS = '../../datasets/gdsc/'\n",
    "FILENAME_LANDMARK_GENES = 'landmark_genes.csv' \n",
    "\n",
    "from config import (\n",
    "    # Drug features.\n",
    "    PATH_TO_SAVED_DRUG_FEATURES,\n",
    "    DRUG_FPS_FINAL_FILE_NAME,\n",
    "    # Cell line features.\n",
    "    PATH_TO_SAVED_CL_FEATURES,\n",
    "    GENE_EXPR_FINAL_FILE_NAME,\n",
    "    CNV_GISTIC_FINAL_FILE_NAME,\n",
    "    CNV_PICNIC_FINAL_FILE_NAME,\n",
    "    PATH_TO_FEATURES    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Create Mutations Dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446521, 922)\n",
      "Index(['DRUG_ID', 'CELL_LINE_NAME', 'AUC', 'CELL_ID', 'LN_IC50', 'CONC',\n",
      "       'MASTER_CELL_ID', 'INTENSITY', 'DATASET', 'Z_SCORE', 'DRUG_NAME',\n",
      "       'RMSE', 'COSMIC_ID', 'POSITION', 'TSPAN6', 'SCYL3', 'BAD', 'LAP3',\n",
      "       'SNX11', 'CASP10'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>CONC</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCBP</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>IFRD2</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>CEBPD</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>MIF</th>\n",
       "      <th>PRAF2</th>\n",
       "      <th>LYN</th>\n",
       "      <th>POLG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>3137</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>49</td>\n",
       "      <td>544404</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>...</td>\n",
       "      <td>8.355826</td>\n",
       "      <td>8.951680</td>\n",
       "      <td>7.205590</td>\n",
       "      <td>3.277948</td>\n",
       "      <td>3.465672</td>\n",
       "      <td>6.312806</td>\n",
       "      <td>12.112498</td>\n",
       "      <td>3.010237</td>\n",
       "      <td>8.750848</td>\n",
       "      <td>6.199366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>2366</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>404197</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>...</td>\n",
       "      <td>5.995760</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.716270</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>2368</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>2.0</td>\n",
       "      <td>610</td>\n",
       "      <td>797378</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>...</td>\n",
       "      <td>6.939741</td>\n",
       "      <td>8.688176</td>\n",
       "      <td>7.085349</td>\n",
       "      <td>3.688222</td>\n",
       "      <td>4.572119</td>\n",
       "      <td>6.345090</td>\n",
       "      <td>12.276166</td>\n",
       "      <td>4.114092</td>\n",
       "      <td>5.768098</td>\n",
       "      <td>7.505155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    DRUG_ID CELL_LINE_NAME       AUC  CELL_ID   LN_IC50  CONC  MASTER_CELL_ID  \\\n",
       "0         1         MC-CAR  0.982114     3137  2.395685   2.0              49   \n",
       "9         1            ES3  0.984816     2366  3.140923   2.0            1342   \n",
       "27        1            ES5  0.985693     2368  3.968757   2.0             610   \n",
       "\n",
       "    INTENSITY DATASET   Z_SCORE  ...     MYCBP      FIS1     IFRD2    NPEPL1  \\\n",
       "0      544404   GDSC1 -0.189576  ...  8.355826  8.951680  7.205590  3.277948   \n",
       "9      404197   GDSC1  0.508635  ...  5.995760  9.337588  7.468226  3.716270   \n",
       "27     797378   GDSC1  1.284229  ...  6.939741  8.688176  7.085349  3.688222   \n",
       "\n",
       "       CEBPD   PLEKHM1        MIF     PRAF2       LYN     POLG2  \n",
       "0   3.465672  6.312806  12.112498  3.010237  8.750848  6.199366  \n",
       "9   5.363887  6.188079  12.281947  4.794624  3.588528  6.785201  \n",
       "27  4.572119  6.345090  12.276166  4.114092  5.768098  7.505155  \n",
       "\n",
       "[3 rows x 922 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get GDSC dataset.\n",
    "GDSC = pd.read_pickle(f'{PATH_TO_SAVE_DATA_TO}joined_gdsc_geneexpr.pkl')\n",
    "print(GDSC.shape)\n",
    "print(GDSC.columns[:20])\n",
    "GDSC.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `mutations_all_20220315.csv` took 15.78615 seconds to import. \n",
      "Shape: (8322616, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15  \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15  \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15  \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15  \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mutations dataset.\n",
    "start = time.time()\n",
    "mutations_v1 = pd.read_csv(f'{PATH_TO_CNV_DATA}{MUTATION_FILE}', sep=\",\", header=0)\n",
    "print(f\"File `{MUTATION_FILE}` took {time.time()-start:.5f} seconds to import. \\nShape: {mutations_v1.shape}\")\n",
    "mutations_v1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is now all the mutations information with the `model_id` which can be mapped to the `SANGER_MODEL_ID` from the build GDSC table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC base shape: (446521, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CONC</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>DATASET</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>49</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>3137</td>\n",
       "      <td>544404</td>\n",
       "      <td>14</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>683665</td>\n",
       "      <td>GDSC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>1342</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>2366</td>\n",
       "      <td>404197</td>\n",
       "      <td>14</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>684055</td>\n",
       "      <td>GDSC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>ES5</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>610</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>2368</td>\n",
       "      <td>797378</td>\n",
       "      <td>14</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>684057</td>\n",
       "      <td>GDSC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SIDM00269</td>\n",
       "      <td>ES7</td>\n",
       "      <td>0.088760</td>\n",
       "      <td>71</td>\n",
       "      <td>1</td>\n",
       "      <td>0.972699</td>\n",
       "      <td>2.692768</td>\n",
       "      <td>0.110056</td>\n",
       "      <td>2371</td>\n",
       "      <td>377574</td>\n",
       "      <td>14</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>684059</td>\n",
       "      <td>GDSC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SIDM00203</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>-0.111820</td>\n",
       "      <td>231</td>\n",
       "      <td>1</td>\n",
       "      <td>0.944462</td>\n",
       "      <td>2.478678</td>\n",
       "      <td>0.087011</td>\n",
       "      <td>2375</td>\n",
       "      <td>638065</td>\n",
       "      <td>14</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>684062</td>\n",
       "      <td>GDSC1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CONC SANGER_MODEL_ID CELL_LINE_NAME   Z_SCORE  MASTER_CELL_ID  DRUG_ID  \\\n",
       "0    2.0       SIDM00636         MC-CAR -0.189576              49        1   \n",
       "9    2.0       SIDM00265            ES3  0.508635            1342        1   \n",
       "27   2.0       SIDM00263            ES5  1.284229             610        1   \n",
       "45   2.0       SIDM00269            ES7  0.088760              71        1   \n",
       "54   2.0       SIDM00203          EW-11 -0.111820             231        1   \n",
       "\n",
       "         AUC   LN_IC50      RMSE  CELL_ID  INTENSITY  POSITION  DRUG_NAME  \\\n",
       "0   0.982114  2.395685  0.022521     3137     544404        14  Erlotinib   \n",
       "9   0.984816  3.140923  0.031840     2366     404197        14  Erlotinib   \n",
       "27  0.985693  3.968757  0.026052     2368     797378        14  Erlotinib   \n",
       "45  0.972699  2.692768  0.110056     2371     377574        14  Erlotinib   \n",
       "54  0.944462  2.478678  0.087011     2375     638065        14  Erlotinib   \n",
       "\n",
       "    COSMIC_ID DATASET  \n",
       "0      683665   GDSC1  \n",
       "9      684055   GDSC1  \n",
       "27     684057   GDSC1  \n",
       "45     684059   GDSC1  \n",
       "54     684062   GDSC1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(f'{PATH_TO_SAVE_DATA_TO}gdsc_base_v2.pkl')\n",
    "print(f\"GDSC base shape: {gdsc_base.shape}\")\n",
    "gdsc_base.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`gdsc_base_v2.pkl` was build in notebook `01_gdsc_base_table.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the CELL_LINE_NAME's per SANGER_MODEL_ID.\n",
    "celllines_per_sangermodelid = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']].groupby('SANGER_MODEL_ID')['CELL_LINE_NAME'].nunique()\n",
    "counts_per_sangermodelid = celllines_per_sangermodelid.values\n",
    "\n",
    "assert (counts_per_sangermodelid != 1).any() == False\n",
    "assert (counts_per_sangermodelid == 1).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `CELL_LINE_NAME` has only exactly one `SANGER_MODEL_ID`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDM00003</td>\n",
       "      <td>M14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDM00023</td>\n",
       "      <td>TE-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDM00040</td>\n",
       "      <td>TMK-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SANGER_MODEL_ID CELL_LINE_NAME\n",
       "0       SIDM00003            M14\n",
       "1       SIDM00023          TE-12\n",
       "2       SIDM00040          TMK-1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only take the interested columns for the mapping.\n",
    "gdsc_mapping_subset = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']]\n",
    "\n",
    "# Only take the unique SANGER_MODEL_ID's, since these have a 1-to-1 relationship to the CELL_LINE_NAME's anyways.\n",
    "gdsc_mapping_subset = gdsc_mapping_subset.groupby('SANGER_MODEL_ID').first().reset_index(level=0)\n",
    "gdsc_mapping_subset.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\n",
      "mutations_v2 Shape: (8322616, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \\\n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15   \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15   \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15   \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15   \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15   \n",
       "\n",
       "  SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "0             NaN            NaN  \n",
       "1             NaN            NaN  \n",
       "2             NaN            NaN  \n",
       "3             NaN            NaN  \n",
       "4             NaN            NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the CELL_LINE_NAME's onto the mutations_all dataset, based on the model_id.\n",
    "mutations_v2 = mutations_v1.merge(right    = gdsc_mapping_subset,\n",
    "                                  left_on  = 'model_id',\n",
    "                                  right_on = 'SANGER_MODEL_ID',\n",
    "                                  how      = 'left')\n",
    "print(\"We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\")                                  \n",
    "print(f\"mutations_v2 Shape: {mutations_v2.shape}\")\n",
    "mutations_v2.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We excluded all rows which have CELL_LINE_NAME as NaN.\n",
      "Number of unique `gene_symbol`s in the mutations dataset: 20450\n",
      "mutations_v3 Shape: (4574963, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SIDG40054</td>\n",
       "      <td>UAP1</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2857</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>JHH-6</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>JHH-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SIDG13175</td>\n",
       "      <td>IPO11</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.3511</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>NCI-H727</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>NCI-H727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SIDG14638</td>\n",
       "      <td>LIN9</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SIDG07173</td>\n",
       "      <td>DYM</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.5294</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SIDG08926</td>\n",
       "      <td>FCAMR</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>D-336MG</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>D-336MG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "19  SIDG40054        UAP1  SIDM00615                -          r.?   \n",
       "20  SIDG13175       IPO11  SIDM01119                -          r.?   \n",
       "21  SIDG14638        LIN9  SIDM00933                -          r.?   \n",
       "22  SIDG07173         DYM  SIDM00933                -          r.?   \n",
       "23  SIDG08926       FCAMR  SIDM00620                -          r.?   \n",
       "\n",
       "   cdna_mutation  cancer_driver  cancer_predisposition_variant      effect  \\\n",
       "19             -          False                          False    intronic   \n",
       "20             -          False                          False    intronic   \n",
       "21             -          False                          False  downstream   \n",
       "22             -          False                          False    intronic   \n",
       "23             -          False                          False    intronic   \n",
       "\n",
       "       vaf  coding  source model_name SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "19  0.2857   False  Sanger      JHH-6       SIDM00615          JHH-6  \n",
       "20  0.3511   False  Sanger   NCI-H727       SIDM01119       NCI-H727  \n",
       "21  1.0000   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "22  0.5294   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "23  0.6250   False  Sanger    D-336MG       SIDM00620        D-336MG  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " # Only taking the once with a CELL_LINE_NAME.\n",
    "mutations_v3 = mutations_v2[mutations_v2['CELL_LINE_NAME'].notna()]\n",
    "print(\"We excluded all rows which have CELL_LINE_NAME as NaN.\")\n",
    "print(f\"Number of unique `gene_symbol`s in the mutations dataset: {len(np.unique(mutations_v3.gene_symbol))}\")\n",
    "print(f\"mutations_v3 Shape: {mutations_v3.shape}\")\n",
    "mutations_v3.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We use the landmark genes to sparse down the dimension.\n",
      "File `landmark_genes.csv` took 0.01575 seconds to import. It has shape (978, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Type</th>\n",
       "      <th>RNA-Seq Correlation</th>\n",
       "      <th>RNA-Seq Correlation Self-Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3638</td>\n",
       "      <td>INSIG1</td>\n",
       "      <td>insulin induced gene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2309</td>\n",
       "      <td>FOXO3</td>\n",
       "      <td>forkhead box O3</td>\n",
       "      <td>Forkhead boxes</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>CDH3</td>\n",
       "      <td>cadherin 3</td>\n",
       "      <td>Type I classical cadherins</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entrez ID  Symbol                    Name                 Gene Family  \\\n",
       "0       3638  INSIG1  insulin induced gene 1                         NaN   \n",
       "1       2309   FOXO3         forkhead box O3              Forkhead boxes   \n",
       "2       1001    CDH3              cadherin 3  Type I classical cadherins   \n",
       "\n",
       "       Type  RNA-Seq Correlation  RNA-Seq Correlation Self-Rank  \n",
       "0  landmark                  NaN                            NaN  \n",
       "1  landmark                  NaN                            NaN  \n",
       "2  landmark                  NaN                            NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "landmark_genes = pd.read_csv(f'{PATH_TO_GDSC_DATASETS}{FILENAME_LANDMARK_GENES}', sep=\"\\t\")\n",
    "print(\"We use the landmark genes to sparse down the dimension.\")\n",
    "print(f\"File `{FILENAME_LANDMARK_GENES}` took {time.time()-start:.5f} seconds to import. It has shape {landmark_genes.shape}\")\n",
    "landmark_genes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We merged the most mutations_v3 dataset with the landmark genes to sparse it down.\n",
      "mutations_v4 Shape: (235906, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>splice_region</td>\n",
       "      <td>0.6190</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5prime_UTR_variant</td>\n",
       "      <td>0.4118</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG40702       VGLL4  SIDM00330                -          r.?   \n",
       "1  SIDG40702       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2  SIDG40702       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1           c.?          False                          False   \n",
       "2           c.?          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source model_name SANGER_MODEL_ID  \\\n",
       "0            intronic  0.9926   False  Sanger      LAN-6       SIDM00330   \n",
       "1       splice_region  0.6190   False   Broad      CL-11       SIDM00944   \n",
       "2  5prime_UTR_variant  0.4118   False   Broad      SNG-M       SIDM00373   \n",
       "\n",
       "  CELL_LINE_NAME Symbol  \n",
       "0          LAN-6  VGLL4  \n",
       "1          CL-11  VGLL4  \n",
       "2          SNG-M  VGLL4  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "landmark_genes_v2 = landmark_genes['Symbol']\n",
    "\n",
    "# Take only the rows which have a `gene_symbol` which is also present in the landmark genes table.\n",
    "mutations_v4 = mutations_v3.merge(right    = landmark_genes_v2,\n",
    "                                  left_on  = 'gene_symbol',\n",
    "                                  right_on = 'Symbol')\n",
    "print(\"We merged the most mutations_v3 dataset with the landmark genes to sparse it down.\")                                  \n",
    "print(f\"mutations_v4 Shape: {mutations_v4.shape}\")\n",
    "mutations_v4.head(3)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We only took a chosen set of columns.\n",
      "Number of unique ...\n",
      "    CELL_LINE_NAME's   :    983 (0.42% out of all rows)\n",
      "    gene_symbol's      :    956 (0.41% out of all rows)\n",
      "    protein_mutation's :  25113 (10.65% out of all rows)\n",
      "    rna_mutation's     :  28392 (12.04% out of all rows)\n",
      "    cdna_mutation's    :  22196 (9.41% out of all rows)\n",
      "    cancer_driver's    :      2 ([False  True])\n",
      "\n",
      "mutations_v5 Shape: (235906, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JHOS-2</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00305</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCI-H1436</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00697</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9677</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0          LAN-6       VGLL4  SIDM00330                -          r.?   \n",
       "1          CL-11       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2          SNG-M       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "3         JHOS-2       VGLL4  SIDM00305                -          r.?   \n",
       "4      NCI-H1436       VGLL4  SIDM00697                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver     vaf  \n",
       "0             -          False  0.9926  \n",
       "1           c.?          False  0.6190  \n",
       "2           c.?          False  0.4118  \n",
       "3             -          False  0.5698  \n",
       "4             -          False  0.9677  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v5 = mutations_v4[[\n",
    "    'CELL_LINE_NAME',\n",
    "    'gene_symbol',\n",
    "    'model_id',\n",
    "    'protein_mutation',\n",
    "    'rna_mutation',\n",
    "    'cdna_mutation',\n",
    "    'cancer_driver',\n",
    "    'vaf'\n",
    "]]\n",
    "print(\"We only took a chosen set of columns.\")\n",
    "n_v5 = mutations_v5.shape[0]\n",
    "uniq_cl = len(np.unique(mutations_v5.CELL_LINE_NAME))\n",
    "uniq_gs = len(np.unique(mutations_v5.gene_symbol))\n",
    "uniq_pm = len(np.unique(mutations_v5.protein_mutation))\n",
    "uniq_rna = len(np.unique(mutations_v5.rna_mutation))\n",
    "uniq_cdna = len(np.unique(mutations_v5.cdna_mutation))\n",
    "\n",
    "print(f\"\"\"Number of unique ...\n",
    "    CELL_LINE_NAME's   : {uniq_cl:6.0f} ({100*uniq_cl/n_v5:2.2f}% out of all rows)\n",
    "    gene_symbol's      : {uniq_gs:6.0f} ({100*uniq_gs/n_v5:2.2f}% out of all rows)\n",
    "    protein_mutation's : {uniq_pm:6.0f} ({100*uniq_pm/n_v5:2.2f}% out of all rows)\n",
    "    rna_mutation's     : {uniq_rna:6.0f} ({100*uniq_rna/n_v5:2.2f}% out of all rows)\n",
    "    cdna_mutation's    : {uniq_cdna:6.0f} ({100*uniq_cdna/n_v5:2.2f}% out of all rows)\n",
    "    cancer_driver's    : {len(np.unique(mutations_v5.cancer_driver)):6.0f} ({np.unique(mutations_v5.cancer_driver)})\n",
    "\"\"\")\n",
    "print(f\"mutations_v5 Shape: {mutations_v5.shape}\")\n",
    "mutations_v5.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve_mutation_positions(mut):\n",
    "    positions = re.findall(r'\\d+', mut)\n",
    "\n",
    "    # No positional information was found.\n",
    "    if len(positions) == 0: \n",
    "        return (None, None)\n",
    "    # Only one position was found.\n",
    "    elif len(positions) == 1:\n",
    "        return (int(positions[0]), int(positions[0]))\n",
    "    # Two positions have been found.\n",
    "    elif len(positions) == 2:\n",
    "        return (int(positions[0]), int(positions[1]))\n",
    "    else:\n",
    "        return ValueError(f\"The mutations cell should contain <=2 values, however, it contains {len(positions)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We retrieve the positional information for each mutation columns and added them as new columns.\n",
      "Shape when only taking protein_mutations which have a positional information: (44952, 11)\n",
      "Shape when only taking cancer_drivers = True: (2216, 11)\n",
      "mutations_v6 Shape: (235906, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "      <th>protein_mutation_pos</th>\n",
       "      <th>rna_mutation_pos</th>\n",
       "      <th>cdna_mutation_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6190</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(407, 7)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4118</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(65, 65)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JHOS-2</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00305</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5698</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCI-H1436</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00697</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9677</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCC1395</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00884</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SNU-175</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00216</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CL-40</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00960</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.480-4g&gt;a</td>\n",
       "      <td>c.83-4G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4615</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(480, 4)</td>\n",
       "      <td>(83, 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TOV-21G</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM01169</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.406c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5610</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(406, 406)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Hep3B2-1-7</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00672</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "      <td>(None, None)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0          LAN-6       VGLL4  SIDM00330                -          r.?   \n",
       "1          CL-11       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2          SNG-M       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "3         JHOS-2       VGLL4  SIDM00305                -          r.?   \n",
       "4      NCI-H1436       VGLL4  SIDM00697                -          r.?   \n",
       "5        HCC1395       VGLL4  SIDM00884                -          r.?   \n",
       "6        SNU-175       VGLL4  SIDM00216                -          r.?   \n",
       "7          CL-40       VGLL4  SIDM00960              p.?   r.480-4g>a   \n",
       "8        TOV-21G       VGLL4  SIDM01169              p.?     r.406c>u   \n",
       "9     Hep3B2-1-7       VGLL4  SIDM00672                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver     vaf protein_mutation_pos rna_mutation_pos  \\\n",
       "0             -          False  0.9926         (None, None)     (None, None)   \n",
       "1           c.?          False  0.6190         (None, None)         (407, 7)   \n",
       "2           c.?          False  0.4118         (None, None)         (65, 65)   \n",
       "3             -          False  0.5698         (None, None)     (None, None)   \n",
       "4             -          False  0.9677         (None, None)     (None, None)   \n",
       "5             -          False  1.0000         (None, None)     (None, None)   \n",
       "6             -          False  1.0000         (None, None)     (None, None)   \n",
       "7     c.83-4G>A          False  0.4615         (None, None)         (480, 4)   \n",
       "8           c.?          False  0.5610         (None, None)       (406, 406)   \n",
       "9             -          False  0.4000         (None, None)     (None, None)   \n",
       "\n",
       "  cdna_mutation_pos  \n",
       "0      (None, None)  \n",
       "1      (None, None)  \n",
       "2      (None, None)  \n",
       "3      (None, None)  \n",
       "4      (None, None)  \n",
       "5      (None, None)  \n",
       "6      (None, None)  \n",
       "7           (83, 4)  \n",
       "8      (None, None)  \n",
       "9      (None, None)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v6 = copy.deepcopy(mutations_v5)\n",
    "mutations_v6['protein_mutation_pos'] = mutations_v6.protein_mutation.apply(retrieve_mutation_positions)\n",
    "mutations_v6['rna_mutation_pos'] = mutations_v6.rna_mutation.apply(retrieve_mutation_positions)\n",
    "mutations_v6['cdna_mutation_pos'] = mutations_v6.cdna_mutation.apply(retrieve_mutation_positions)\n",
    "print(\"We retrieve the positional information for each mutation columns and added them as new columns.\")\n",
    "print(f\"Shape when only taking protein_mutations which have a positional information: {mutations_v6[mutations_v6.protein_mutation_pos != (None, None)].shape}\")\n",
    "print(f\"Shape when only taking cancer_drivers = True: {mutations_v6[mutations_v6.cancer_driver==True].shape}\")\n",
    "print(f\"mutations_v6 Shape: {mutations_v6.shape}\")\n",
    "mutations_v6.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "incomplete input (3441924051.py, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [14]\u001b[0;36m\u001b[0m\n\u001b[0;31m    # Set all values wit ha\u001b[0m\n\u001b[0m                            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m incomplete input\n"
     ]
    }
   ],
   "source": [
    "def set_mutation_value(df):\n",
    "    # Set all values wit ha "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (983, 957)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_symbol</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>451Lu</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 957 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_symbol CELL_LINE_NAME  ABCB6  ABCC5  ABCF1  ABCF3  ABHD4  ABHD6  ABL1  \\\n",
       "0                    22RV1    0.0    0.0    1.0    1.0    1.0    0.0   1.0   \n",
       "1                 23132-87    0.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "2                 42-MG-BA    1.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "3                    451Lu    1.0    0.0    1.0    0.0    0.0    0.0   0.0   \n",
       "4                     5637    0.0    1.0    1.0    1.0    1.0    0.0   1.0   \n",
       "\n",
       "gene_symbol  ACAA1  ACAT2  ...  ZMIZ1  ZMYM2  ZNF131  ZNF274  ZNF318  ZNF395  \\\n",
       "0              0.0    0.0  ...    1.0    1.0     0.0     0.0     1.0     0.0   \n",
       "1              0.0    0.0  ...    0.0    1.0     1.0     0.0     0.0     0.0   \n",
       "2              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "3              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "4              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "gene_symbol  ZNF451  ZNF586  ZNF589  ZW10  \n",
       "0               0.0     0.0     0.0   1.0  \n",
       "1               0.0     0.0     1.0   0.0  \n",
       "2               0.0     0.0     0.0   0.0  \n",
       "3               1.0     0.0     0.0   0.0  \n",
       "4               0.0     0.0     0.0   0.0  \n",
       "\n",
       "[5 rows x 957 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_all_final_v1 = pd.pivot_table(\n",
    "    data    = mutations_v6,\n",
    "    values  = 'cancer_driver',\n",
    "    index   = ['CELL_LINE_NAME'],\n",
    "    columns = ['gene_symbol'],\n",
    "    aggfunc = np.sum,\n",
    "    dropna  = False\n",
    ")\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "# Set mutation values: 1.0=mutation, 0.0=no_mutation\n",
    "mutations_all_final_v1[mutations_all_final_v1==0.0] = 1.0\n",
    "mutations_all_final_v1[np.isnan(mutations_all_final_v1)] = 0.0\n",
    "\n",
    "mutations_all_final_v1['CELL_LINE_NAME'] = mutations_all_final_v1.index\n",
    "mutations_all_final_v1.insert(0, 'CELL_LINE_NAME', mutations_all_final_v1.pop('CELL_LINE_NAME'))\n",
    "mutations_all_final_v1.reset_index(drop=True, inplace=True)\n",
    "\n",
    "print(f\"Shape: {mutations_all_final_v1.shape}\")\n",
    "mutations_all_final_v1.head(5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare mutations dataset with the other features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes after by the cell-line column:\n",
      "    gene expr  : (446521, 922)\n",
      "    cnv gistic : (446521, 952)\n",
      "    cnv picnic : (446521, 980)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Read other features.\n",
    "gene_expr = pd.read_pickle(f'{PATH_TO_SAVED_CL_FEATURES}{GENE_EXPR_FINAL_FILE_NAME}')\n",
    "cnv_gistic = pd.read_pickle(f'{PATH_TO_SAVED_CL_FEATURES}{CNV_GISTIC_FINAL_FILE_NAME}')\n",
    "cnv_picnic = pd.read_pickle(f'{PATH_TO_SAVED_CL_FEATURES}{CNV_PICNIC_FINAL_FILE_NAME}')\n",
    "print(f\"\"\"Shapes after by the cell-line column:\n",
    "    gene expr  : {gene_expr.shape}\n",
    "    cnv gistic : {cnv_gistic.shape}\n",
    "    cnv picnic : {cnv_picnic.shape}\n",
    "\"\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "469\n",
      "1\n",
      "(505, 922)\n",
      "(505, 922)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>CONC</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCBP</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>IFRD2</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>CEBPD</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>MIF</th>\n",
       "      <th>PRAF2</th>\n",
       "      <th>LYN</th>\n",
       "      <th>POLG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>188514</th>\n",
       "      <td>133</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.386285</td>\n",
       "      <td>3399</td>\n",
       "      <td>-3.520300</td>\n",
       "      <td>1.024</td>\n",
       "      <td>1027</td>\n",
       "      <td>26702</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-1.052514</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197226</th>\n",
       "      <td>134</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.536710</td>\n",
       "      <td>3399</td>\n",
       "      <td>0.246500</td>\n",
       "      <td>16.000</td>\n",
       "      <td>1027</td>\n",
       "      <td>138738</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.671777</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205830</th>\n",
       "      <td>135</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.395485</td>\n",
       "      <td>3399</td>\n",
       "      <td>-3.456605</td>\n",
       "      <td>1.024</td>\n",
       "      <td>1027</td>\n",
       "      <td>157079</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.392251</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214596</th>\n",
       "      <td>136</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.223736</td>\n",
       "      <td>3399</td>\n",
       "      <td>-2.005449</td>\n",
       "      <td>16.000</td>\n",
       "      <td>1027</td>\n",
       "      <td>20116</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.782996</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223308</th>\n",
       "      <td>140</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.634219</td>\n",
       "      <td>3399</td>\n",
       "      <td>-4.606677</td>\n",
       "      <td>0.064</td>\n",
       "      <td>1027</td>\n",
       "      <td>64405</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.471168</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231975</th>\n",
       "      <td>147</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.979699</td>\n",
       "      <td>3399</td>\n",
       "      <td>5.466523</td>\n",
       "      <td>16.000</td>\n",
       "      <td>1027</td>\n",
       "      <td>685810</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.666688</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240795</th>\n",
       "      <td>150</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.971586</td>\n",
       "      <td>3399</td>\n",
       "      <td>2.298556</td>\n",
       "      <td>1.024</td>\n",
       "      <td>1027</td>\n",
       "      <td>451990</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.547672</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249642</th>\n",
       "      <td>151</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.863996</td>\n",
       "      <td>3399</td>\n",
       "      <td>2.432710</td>\n",
       "      <td>10.240</td>\n",
       "      <td>1027</td>\n",
       "      <td>233360</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.604453</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259929</th>\n",
       "      <td>152</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.792572</td>\n",
       "      <td>3399</td>\n",
       "      <td>2.149573</td>\n",
       "      <td>16.000</td>\n",
       "      <td>1027</td>\n",
       "      <td>150953</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.184076</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269676</th>\n",
       "      <td>153</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.958318</td>\n",
       "      <td>3399</td>\n",
       "      <td>1.115831</td>\n",
       "      <td>0.512</td>\n",
       "      <td>1027</td>\n",
       "      <td>492218</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.942878</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        DRUG_ID CELL_LINE_NAME       AUC  CELL_ID   LN_IC50    CONC  \\\n",
       "188514      133          22RV1  0.386285     3399 -3.520300   1.024   \n",
       "197226      134          22RV1  0.536710     3399  0.246500  16.000   \n",
       "205830      135          22RV1  0.395485     3399 -3.456605   1.024   \n",
       "214596      136          22RV1  0.223736     3399 -2.005449  16.000   \n",
       "223308      140          22RV1  0.634219     3399 -4.606677   0.064   \n",
       "231975      147          22RV1  0.979699     3399  5.466523  16.000   \n",
       "240795      150          22RV1  0.971586     3399  2.298556   1.024   \n",
       "249642      151          22RV1  0.863996     3399  2.432710  10.240   \n",
       "259929      152          22RV1  0.792572     3399  2.149573  16.000   \n",
       "269676      153          22RV1  0.958318     3399  1.115831   0.512   \n",
       "\n",
       "        MASTER_CELL_ID  INTENSITY DATASET   Z_SCORE  ...    MYCBP       FIS1  \\\n",
       "188514            1027      26702   GDSC1 -1.052514  ...  7.98971  10.102478   \n",
       "197226            1027     138738   GDSC1 -0.671777  ...  7.98971  10.102478   \n",
       "205830            1027     157079   GDSC1 -0.392251  ...  7.98971  10.102478   \n",
       "214596            1027      20116   GDSC1 -0.782996  ...  7.98971  10.102478   \n",
       "223308            1027      64405   GDSC1 -0.471168  ...  7.98971  10.102478   \n",
       "231975            1027     685810   GDSC1  0.666688  ...  7.98971  10.102478   \n",
       "240795            1027     451990   GDSC1 -0.547672  ...  7.98971  10.102478   \n",
       "249642            1027     233360   GDSC1 -0.604453  ...  7.98971  10.102478   \n",
       "259929            1027     150953   GDSC1 -0.184076  ...  7.98971  10.102478   \n",
       "269676            1027     492218   GDSC1  0.942878  ...  7.98971  10.102478   \n",
       "\n",
       "           IFRD2   NPEPL1     CEBPD   PLEKHM1        MIF    PRAF2       LYN  \\\n",
       "188514  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "197226  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "205830  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "214596  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "223308  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "231975  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "240795  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "249642  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "259929  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "269676  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "\n",
       "        POLG2  \n",
       "188514  4.826  \n",
       "197226  4.826  \n",
       "205830  4.826  \n",
       "214596  4.826  \n",
       "223308  4.826  \n",
       "231975  4.826  \n",
       "240795  4.826  \n",
       "249642  4.826  \n",
       "259929  4.826  \n",
       "269676  4.826  \n",
       "\n",
       "[10 rows x 922 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(np.unique(gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].DRUG_ID.values)))\n",
    "print(len(np.unique(gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].FIS1.values)))\n",
    "print(gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].shape)\n",
    "print(gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].drop_duplicates().shape)\n",
    "gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- There are multiple rows which contain the same `DRUG_ID`-`CELL_LINE_NAME` tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1010    2\n",
       "1053    2\n",
       "1049    2\n",
       "1048    2\n",
       "1047    2\n",
       "Name: DRUG_ID, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expr[gene_expr.CELL_LINE_NAME=='22RV1'].DRUG_ID.value_counts().head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>CONC</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCBP</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>IFRD2</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>CEBPD</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>MIF</th>\n",
       "      <th>PRAF2</th>\n",
       "      <th>LYN</th>\n",
       "      <th>POLG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1970540</th>\n",
       "      <td>1010</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.981310</td>\n",
       "      <td>3399</td>\n",
       "      <td>2.102708</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1027</td>\n",
       "      <td>37462</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.169802</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3551362</th>\n",
       "      <td>1010</td>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.974411</td>\n",
       "      <td>3399</td>\n",
       "      <td>4.032555</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1027</td>\n",
       "      <td>37462</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>0.542607</td>\n",
       "      <td>...</td>\n",
       "      <td>7.98971</td>\n",
       "      <td>10.102478</td>\n",
       "      <td>7.531275</td>\n",
       "      <td>3.18628</td>\n",
       "      <td>7.831092</td>\n",
       "      <td>5.561236</td>\n",
       "      <td>11.995057</td>\n",
       "      <td>3.86721</td>\n",
       "      <td>5.500169</td>\n",
       "      <td>4.826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         DRUG_ID CELL_LINE_NAME       AUC  CELL_ID   LN_IC50  CONC  \\\n",
       "1970540     1010          22RV1  0.981310     3399  2.102708   2.0   \n",
       "3551362     1010          22RV1  0.974411     3399  4.032555   2.0   \n",
       "\n",
       "         MASTER_CELL_ID  INTENSITY DATASET   Z_SCORE  ...    MYCBP       FIS1  \\\n",
       "1970540            1027      37462   GDSC1  1.169802  ...  7.98971  10.102478   \n",
       "3551362            1027      37462   GDSC2  0.542607  ...  7.98971  10.102478   \n",
       "\n",
       "            IFRD2   NPEPL1     CEBPD   PLEKHM1        MIF    PRAF2       LYN  \\\n",
       "1970540  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "3551362  7.531275  3.18628  7.831092  5.561236  11.995057  3.86721  5.500169   \n",
       "\n",
       "         POLG2  \n",
       "1970540  4.826  \n",
       "3551362  4.826  \n",
       "\n",
       "[2 rows x 922 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expr[(gene_expr.CELL_LINE_NAME=='22RV1') & (gene_expr.DRUG_ID==1010)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Note__:\n",
    "- There can be the same row for the same `DRUG_ID`-`CELL_LINE_NAME` tuple with the same `CONC` and `INTENSITY`\n",
    "- However, this will differ only in the `DATASET` column and one will come from GDSC1, while the other comes from GDSC2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO: Which one do I choose, GDSC1 or GDSC2? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(310946, 922)\n",
      "(135575, 922)\n"
     ]
    }
   ],
   "source": [
    "print(gene_expr[gene_expr.DATASET=='GDSC1'].shape)\n",
    "print(gene_expr[gene_expr.DATASET=='GDSC2'].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TSPAN6', 'SCYL3', 'BAD', 'LAP3', 'SNX11', 'CASP10', 'CFLAR', 'FKBP4',\n",
       "       'RBM6', 'SLC25A13',\n",
       "       ...\n",
       "       'MYCBP', 'FIS1', 'IFRD2', 'NPEPL1', 'CEBPD', 'PLEKHM1', 'MIF', 'PRAF2',\n",
       "       'LYN', 'POLG2'],\n",
       "      dtype='object', length=908)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expr.columns[14:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique rows of column\n",
      "    CELL_LINE_NAME\n",
      "        mutation   : 983\n",
      "        gene-expr  : 988\n",
      "        cnv-gistic : 988\n",
      "        cnv-picnic : 988\n",
      "    GENE_SYMBOL\n",
      "        mutation   : 956\n",
      "        gene-expr  : 908\n",
      "        cnv-gistic : 938\n",
      "        cnv-picnic : 966\n",
      "\n",
      "Size of intersection set for\n",
      "    CELL_LINE_NAME :   983\n",
      "    GENE_SYMBOL    :   858\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mut__uniq_cl = np.unique(mutations_all_final_v1.CELL_LINE_NAME.values)\n",
    "mut__uni_gs = np.unique(mutations_all_final_v1.columns[1:])\n",
    "gexp__uniq_cl = np.unique(gene_expr.CELL_LINE_NAME.values)\n",
    "gexp__uniq_gs = np.unique(gene_expr.columns[14:])\n",
    "cnvg__uniq_cl = np.unique(cnv_gistic.CELL_LINE_NAME.values)\n",
    "cnvg__uniq_gs = np.unique(cnv_gistic.columns[14:])\n",
    "cnvp__uniq_cl = np.unique(cnv_picnic.CELL_LINE_NAME.values)\n",
    "cnvp__uniq_gs = np.unique(cnv_picnic.columns[14:])\n",
    "\n",
    "print(f\"\"\"Unique rows of column\n",
    "    CELL_LINE_NAME\n",
    "        mutation   : {len(mut__uniq_cl)}\n",
    "        gene-expr  : {len(gexp__uniq_cl)}\n",
    "        cnv-gistic : {len(cnvg__uniq_cl)}\n",
    "        cnv-picnic : {len(cnvp__uniq_cl)}\n",
    "    GENE_SYMBOL\n",
    "        mutation   : {len(mut__uni_gs)}\n",
    "        gene-expr  : {len(gexp__uniq_gs)}\n",
    "        cnv-gistic : {len(cnvg__uniq_gs)}\n",
    "        cnv-picnic : {len(cnvp__uniq_gs)}\n",
    "\"\"\")\n",
    "\n",
    "# Find intersection of the cell-lines.\n",
    "intersection_cl = set(mut__uniq_cl).intersection(\\\n",
    "    set(gexp__uniq_cl)).intersection(\\\n",
    "    set(cnvg__uniq_cl)).intersection(\\\n",
    "    set(cnvp__uniq_cl))\n",
    "# Find intersection of the gene-symbol.\n",
    "intersection_gs = set(mut__uni_gs).intersection(\\\n",
    "    set(gexp__uniq_gs)).intersection(\\\n",
    "    set(cnvg__uniq_gs)).intersection(\\\n",
    "    set(cnvp__uniq_gs))\n",
    "\n",
    "print(f\"\"\"Size of intersection set for\n",
    "    CELL_LINE_NAME : {len(intersection_cl):5.0f}\n",
    "    GENE_SYMBOL    : {len(intersection_gs):5.0f}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FBXL12',\n",
       " 'PIN1',\n",
       " 'PAK4',\n",
       " 'GNA15',\n",
       " 'ARPP19',\n",
       " 'EAPP',\n",
       " 'MOK',\n",
       " 'MTHFD2',\n",
       " 'TIPARP',\n",
       " 'CASP3']"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(intersection_gs)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Expr\n",
      "----------\n",
      "Shape total: (444580, 861)\n",
      "Shape GDSC1: (309541, 861)\n",
      "Shape GDSC2: (135039, 861)\n",
      "CNV Gistic\n",
      "----------\n",
      "Shape total: (444580, 861)\n",
      "Shape GDSC1: (309541, 861)\n",
      "Shape GDSC2: (135039, 861)\n",
      "CNV Picnic\n",
      "----------\n",
      "Shape total: (444580, 861)\n",
      "Shape GDSC1: (309541, 861)\n",
      "Shape GDSC2: (135039, 861)\n",
      "Mutation\n",
      "----------\n",
      "Shape total: (983, 859)\n"
     ]
    }
   ],
   "source": [
    "# Create datasets with intersecting CELL_LINE_NAME's and GENE_SYMBOL's only.\n",
    "print(f\"Gene Expr\\n{10*'-'}\")\n",
    "gexpr_inter = gene_expr[['DRUG_ID', 'DATASET', 'CELL_LINE_NAME'] + list(intersection_gs)]\n",
    "gexpr_inter = gexpr_inter[gexpr_inter.CELL_LINE_NAME.isin(list(intersection_cl))]\n",
    "print(f\"Shape total: {gexpr_inter.shape}\")\n",
    "print(f\"Shape GDSC1: {gexpr_inter[gexpr_inter.DATASET=='GDSC1'].shape}\")\n",
    "print(f\"Shape GDSC2: {gexpr_inter[gexpr_inter.DATASET=='GDSC2'].shape}\")\n",
    "\n",
    "print(f\"CNV Gistic\\n{10*'-'}\")\n",
    "cnvg_inter = cnv_gistic[['DRUG_ID', 'DATASET', 'CELL_LINE_NAME'] + list(intersection_gs)]\n",
    "cnvg_inter = cnvg_inter[cnvg_inter.CELL_LINE_NAME.isin(list(intersection_cl))]\n",
    "print(f\"Shape total: {cnvg_inter.shape}\")\n",
    "print(f\"Shape GDSC1: {cnvg_inter[cnvg_inter.DATASET=='GDSC1'].shape}\")\n",
    "print(f\"Shape GDSC2: {cnvg_inter[cnvg_inter.DATASET=='GDSC2'].shape}\")\n",
    "\n",
    "print(f\"CNV Picnic\\n{10*'-'}\")\n",
    "cnvp_inter = cnv_picnic[['DRUG_ID', 'DATASET', 'CELL_LINE_NAME'] + list(intersection_gs)]\n",
    "cnvp_inter = cnvp_inter[cnvp_inter.CELL_LINE_NAME.isin(list(intersection_cl))]\n",
    "print(f\"Shape total: {cnvp_inter.shape}\")\n",
    "print(f\"Shape GDSC1: {cnvp_inter[cnvp_inter.DATASET=='GDSC1'].shape}\")\n",
    "print(f\"Shape GDSC2: {cnvp_inter[cnvp_inter.DATASET=='GDSC2'].shape}\")\n",
    "\n",
    "print(f\"Mutation\\n{10*'-'}\")\n",
    "mut_inter = mutations_all_final_v1[['CELL_LINE_NAME'] + list(intersection_gs)]\n",
    "mut_inter = mut_inter[mut_inter.CELL_LINE_NAME.isin(list(intersection_cl))]\n",
    "print(f\"Shape total: {mut_inter.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We don't have `DRUG_ID` information in the mutations dataset. Which we also don't need. We can sparse down the other features (gene expr, cnv) datasets in the sense that we only have the unique intersecting `CELL_LINE_NAME`'s as rows, the unique intersecting `GENE_SYMBOL`'s as columns, and as cells the value per cell-line-gene tuple. Thus, it should be the same shape as for the mutations dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess feature datasets\n",
    "\n",
    "- [ ] Sparse down the gene expression, cnv gistic and cnv picnic datasets to the same shape as the final mutations dataset from above\n",
    "- [ ] Save the drug-cell-line tuple dataset in a separate file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Feature dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sparse feature sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>FBXL12</th>\n",
       "      <th>PIN1</th>\n",
       "      <th>PAK4</th>\n",
       "      <th>GNA15</th>\n",
       "      <th>ARPP19</th>\n",
       "      <th>EAPP</th>\n",
       "      <th>MOK</th>\n",
       "      <th>...</th>\n",
       "      <th>PDHX</th>\n",
       "      <th>DFFB</th>\n",
       "      <th>FOSL1</th>\n",
       "      <th>ETS1</th>\n",
       "      <th>EBNA1BP2</th>\n",
       "      <th>MYL9</th>\n",
       "      <th>MLLT11</th>\n",
       "      <th>PFKL</th>\n",
       "      <th>FGFR4</th>\n",
       "      <th>SDHB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES7</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>EW-11</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 861 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    DRUG_ID DATASET CELL_LINE_NAME FBXL12 PIN1 PAK4 GNA15 ARPP19 EAPP  MOK  \\\n",
       "0         1   GDSC1         MC-CAR    2.0  2.0  2.0   2.0    2.0  2.0  2.0   \n",
       "9         1   GDSC1            ES3    2.0  2.0  2.0   2.0    2.0  2.0  2.0   \n",
       "27        1   GDSC1            ES5    4.0  4.0  3.0   4.0    4.0  2.0  2.0   \n",
       "45        1   GDSC1            ES7    3.0  3.0  3.0   3.0    3.0  2.0  3.0   \n",
       "54        1   GDSC1          EW-11    3.0  3.0  3.0   3.0    3.0  2.0  3.0   \n",
       "\n",
       "    ... PDHX DFFB FOSL1 ETS1 EBNA1BP2 MYL9 MLLT11 PFKL FGFR4 SDHB  \n",
       "0   ...  2.0  2.0   2.0  2.0      2.0  2.0    2.0  2.0   2.0  2.0  \n",
       "9   ...  2.0  3.0   2.0  2.0      3.0  3.0    3.0  2.0   2.0  3.0  \n",
       "27  ...  2.0  4.0   3.0  3.0      5.0  4.0    5.0  3.0   3.0  4.0  \n",
       "45  ...  3.0  3.0   3.0  2.0      3.0  4.0    3.0  4.0   3.0  3.0  \n",
       "54  ...  4.0  4.0   4.0  4.0      4.0  4.0    5.0  5.0   4.0  4.0  \n",
       "\n",
       "[5 rows x 861 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnvp_inter.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>FBXL12</th>\n",
       "      <th>PIN1</th>\n",
       "      <th>PAK4</th>\n",
       "      <th>GNA15</th>\n",
       "      <th>ARPP19</th>\n",
       "      <th>EAPP</th>\n",
       "      <th>MOK</th>\n",
       "      <th>MTHFD2</th>\n",
       "      <th>TIPARP</th>\n",
       "      <th>...</th>\n",
       "      <th>PDHX</th>\n",
       "      <th>DFFB</th>\n",
       "      <th>FOSL1</th>\n",
       "      <th>ETS1</th>\n",
       "      <th>EBNA1BP2</th>\n",
       "      <th>MYL9</th>\n",
       "      <th>MLLT11</th>\n",
       "      <th>PFKL</th>\n",
       "      <th>FGFR4</th>\n",
       "      <th>SDHB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>7.023759</td>\n",
       "      <td>6.067534</td>\n",
       "      <td>4.318750</td>\n",
       "      <td>3.261427</td>\n",
       "      <td>6.297582</td>\n",
       "      <td>8.313991</td>\n",
       "      <td>5.514912</td>\n",
       "      <td>10.594112</td>\n",
       "      <td>5.222366</td>\n",
       "      <td>...</td>\n",
       "      <td>7.821536</td>\n",
       "      <td>3.601622</td>\n",
       "      <td>3.225596</td>\n",
       "      <td>3.651201</td>\n",
       "      <td>7.895763</td>\n",
       "      <td>3.953414</td>\n",
       "      <td>4.059382</td>\n",
       "      <td>4.376822</td>\n",
       "      <td>3.215209</td>\n",
       "      <td>9.267565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>6.714387</td>\n",
       "      <td>5.695096</td>\n",
       "      <td>4.536146</td>\n",
       "      <td>3.295886</td>\n",
       "      <td>7.021037</td>\n",
       "      <td>8.500080</td>\n",
       "      <td>4.862145</td>\n",
       "      <td>10.609245</td>\n",
       "      <td>6.528668</td>\n",
       "      <td>...</td>\n",
       "      <td>8.094289</td>\n",
       "      <td>3.596762</td>\n",
       "      <td>3.486299</td>\n",
       "      <td>3.127452</td>\n",
       "      <td>7.852436</td>\n",
       "      <td>3.869411</td>\n",
       "      <td>4.248318</td>\n",
       "      <td>4.989945</td>\n",
       "      <td>4.328643</td>\n",
       "      <td>9.515870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>7.752402</td>\n",
       "      <td>5.475753</td>\n",
       "      <td>4.033714</td>\n",
       "      <td>3.176525</td>\n",
       "      <td>7.279671</td>\n",
       "      <td>8.013367</td>\n",
       "      <td>4.957332</td>\n",
       "      <td>11.266705</td>\n",
       "      <td>7.445954</td>\n",
       "      <td>...</td>\n",
       "      <td>7.984052</td>\n",
       "      <td>3.317746</td>\n",
       "      <td>5.106906</td>\n",
       "      <td>5.305024</td>\n",
       "      <td>6.508066</td>\n",
       "      <td>7.840349</td>\n",
       "      <td>8.632889</td>\n",
       "      <td>4.792137</td>\n",
       "      <td>3.078971</td>\n",
       "      <td>8.495921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>451Lu</td>\n",
       "      <td>6.518083</td>\n",
       "      <td>5.462520</td>\n",
       "      <td>4.436039</td>\n",
       "      <td>3.241076</td>\n",
       "      <td>6.657767</td>\n",
       "      <td>7.085021</td>\n",
       "      <td>8.199066</td>\n",
       "      <td>10.550334</td>\n",
       "      <td>5.837560</td>\n",
       "      <td>...</td>\n",
       "      <td>6.970153</td>\n",
       "      <td>3.695460</td>\n",
       "      <td>4.947532</td>\n",
       "      <td>5.198087</td>\n",
       "      <td>7.361412</td>\n",
       "      <td>3.687171</td>\n",
       "      <td>5.965388</td>\n",
       "      <td>4.885217</td>\n",
       "      <td>3.849932</td>\n",
       "      <td>9.726323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5637</td>\n",
       "      <td>6.855088</td>\n",
       "      <td>5.980778</td>\n",
       "      <td>4.382524</td>\n",
       "      <td>6.086206</td>\n",
       "      <td>7.423409</td>\n",
       "      <td>8.120180</td>\n",
       "      <td>5.212472</td>\n",
       "      <td>10.329122</td>\n",
       "      <td>7.212325</td>\n",
       "      <td>...</td>\n",
       "      <td>7.068021</td>\n",
       "      <td>3.386432</td>\n",
       "      <td>6.835862</td>\n",
       "      <td>6.190242</td>\n",
       "      <td>7.431739</td>\n",
       "      <td>3.652418</td>\n",
       "      <td>7.088577</td>\n",
       "      <td>4.761850</td>\n",
       "      <td>3.163742</td>\n",
       "      <td>9.091481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 859 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME    FBXL12      PIN1      PAK4     GNA15    ARPP19      EAPP  \\\n",
       "0          22RV1  7.023759  6.067534  4.318750  3.261427  6.297582  8.313991   \n",
       "1       23132-87  6.714387  5.695096  4.536146  3.295886  7.021037  8.500080   \n",
       "2       42-MG-BA  7.752402  5.475753  4.033714  3.176525  7.279671  8.013367   \n",
       "3          451Lu  6.518083  5.462520  4.436039  3.241076  6.657767  7.085021   \n",
       "4           5637  6.855088  5.980778  4.382524  6.086206  7.423409  8.120180   \n",
       "\n",
       "        MOK     MTHFD2    TIPARP  ...      PDHX      DFFB     FOSL1      ETS1  \\\n",
       "0  5.514912  10.594112  5.222366  ...  7.821536  3.601622  3.225596  3.651201   \n",
       "1  4.862145  10.609245  6.528668  ...  8.094289  3.596762  3.486299  3.127452   \n",
       "2  4.957332  11.266705  7.445954  ...  7.984052  3.317746  5.106906  5.305024   \n",
       "3  8.199066  10.550334  5.837560  ...  6.970153  3.695460  4.947532  5.198087   \n",
       "4  5.212472  10.329122  7.212325  ...  7.068021  3.386432  6.835862  6.190242   \n",
       "\n",
       "   EBNA1BP2      MYL9    MLLT11      PFKL     FGFR4      SDHB  \n",
       "0  7.895763  3.953414  4.059382  4.376822  3.215209  9.267565  \n",
       "1  7.852436  3.869411  4.248318  4.989945  4.328643  9.515870  \n",
       "2  6.508066  7.840349  8.632889  4.792137  3.078971  8.495921  \n",
       "3  7.361412  3.687171  5.965388  4.885217  3.849932  9.726323  \n",
       "4  7.431739  3.652418  7.088577  4.761850  3.163742  9.091481  \n",
       "\n",
       "[5 rows x 859 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gexpr_sparse = gexpr_inter.groupby('CELL_LINE_NAME').first().reset_index()\n",
    "gexpr_sparse = gexpr_sparse.loc[:, ~gexpr_sparse.columns.isin(['DRUG_ID', 'DATASET'])]\n",
    "\n",
    "cnvg_sparse = cnvg_inter.groupby('CELL_LINE_NAME').first().reset_index()\n",
    "cnvg_sparse = cnvg_sparse.loc[:, ~cnvg_sparse.columns.isin(['DRUG_ID', 'DATASET'])]\n",
    "\n",
    "cnvp_sparse = cnvp_inter.groupby('CELL_LINE_NAME').first().reset_index()\n",
    "cnvp_sparse = cnvp_sparse.loc[:, ~cnvp_sparse.columns.isin(['DRUG_ID', 'DATASET'])]\n",
    "\n",
    "gexpr_sparse.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature dataset shapes\n",
      "    Gene Expression : (983, 859)\n",
      "    CNV Gistic      : (983, 859)\n",
      "    CNV Picnic      : (983, 859)\n",
      "    Mutation        : (983, 859)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "assert gexpr_sparse.shape == cnvg_sparse.shape == cnvp_sparse.shape == mut_inter.shape\n",
    "print(f\"\"\"Feature dataset shapes\n",
    "    Gene Expression : {gexpr_sparse.shape}\n",
    "    CNV Gistic      : {cnvg_sparse.shape}\n",
    "    CNV Picnic      : {cnvp_sparse.shape}\n",
    "    Mutation        : {mut_inter.shape}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- We have now a final dataset per feature with cell-lines as rows and gene symbols as columns which are existent in all datasets.\n",
    "- Also only one row per cell-line was kept, since for the same cell-line the rows only different because of different drugs\n",
    "- We can save these datasets and use them for model building \n",
    "  - However, we need to save also the `LN_IC50` values for the cell-line - drug tuples\n",
    "    - NOTE that the gene_symbols are only a feature of the genes and independent of the drug, thus they always have the same value for a cell-line, even for different drugs\n",
    "- NOTE that some cell-line-gene-symbol tuples are `NaN`. That just means that the database had no feature value for the specific gene in the specific cell-line "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save sparsed features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "%mkdir ../../datasets/datasets_for_model_building"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_FEATURES = '../../datasets/datasets_for_model_building/'\n",
    "\n",
    "gexpr_sparse.to_pickle(f'{PATH_TO_FEATURES}gexpr_sparse.pkl')\n",
    "cnvg_sparse.to_pickle(f'{PATH_TO_FEATURES}cnvg_sparse.pkl')\n",
    "cnvp_sparse.to_pickle(f'{PATH_TO_FEATURES}cnvp_sparse.pkl')\n",
    "mut_inter.to_pickle(f'{PATH_TO_FEATURES}mut_sparse.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "READ = False\n",
    "\n",
    "if READ:\n",
    "    with open(f'{PATH_TO_FEATURES}gexpr_sparse.pkl', 'rb') as f: \n",
    "        gexpr_sparse = pickle.load(f)\n",
    "    with open(f'{PATH_TO_FEATURES}cnvg_sparse.pkl', 'rb') as f: \n",
    "        cnvg_sparse = pickle.load(f)\n",
    "    with open(f'{PATH_TO_FEATURES}cnvp_sparse.pkl', 'rb') as f: \n",
    "        cnvp_sparse = pickle.load(f)\n",
    "    with open(f'{PATH_TO_FEATURES}mut_sparse.pkl', 'rb') as f: \n",
    "        mut_sparse = pickle.load(f)                       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get drug dataset\n",
    "\n",
    "- We need drug, cell-line and IC50 information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (446521, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>190089</th>\n",
       "      <td>201T</td>\n",
       "      <td>133</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-3.770673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198783</th>\n",
       "      <td>201T</td>\n",
       "      <td>134</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.814180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207405</th>\n",
       "      <td>201T</td>\n",
       "      <td>135</td>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.298050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216171</th>\n",
       "      <td>201T</td>\n",
       "      <td>136</td>\n",
       "      <td>Mitomycin-C</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-4.472378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224883</th>\n",
       "      <td>201T</td>\n",
       "      <td>140</td>\n",
       "      <td>Vinorelbine</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-5.332884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233550</th>\n",
       "      <td>201T</td>\n",
       "      <td>147</td>\n",
       "      <td>NSC-87877</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>4.680281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242370</th>\n",
       "      <td>201T</td>\n",
       "      <td>150</td>\n",
       "      <td>Bicalutamide</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.754322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251217</th>\n",
       "      <td>201T</td>\n",
       "      <td>151</td>\n",
       "      <td>QS11</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.992590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261873</th>\n",
       "      <td>201T</td>\n",
       "      <td>152</td>\n",
       "      <td>CP466722</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.299660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271251</th>\n",
       "      <td>201T</td>\n",
       "      <td>153</td>\n",
       "      <td>Midostaurin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-1.837530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279918</th>\n",
       "      <td>201T</td>\n",
       "      <td>154</td>\n",
       "      <td>CHIR-99021</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.701566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288567</th>\n",
       "      <td>201T</td>\n",
       "      <td>155</td>\n",
       "      <td>Ponatinib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.494190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297198</th>\n",
       "      <td>201T</td>\n",
       "      <td>156</td>\n",
       "      <td>AZD6482</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.597021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305847</th>\n",
       "      <td>201T</td>\n",
       "      <td>157</td>\n",
       "      <td>JNK-9L</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.979145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314460</th>\n",
       "      <td>201T</td>\n",
       "      <td>158</td>\n",
       "      <td>PF-562271</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.235556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323010</th>\n",
       "      <td>201T</td>\n",
       "      <td>159</td>\n",
       "      <td>HG6-64-1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.271747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331686</th>\n",
       "      <td>201T</td>\n",
       "      <td>163</td>\n",
       "      <td>JQ1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.048784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340344</th>\n",
       "      <td>201T</td>\n",
       "      <td>164</td>\n",
       "      <td>JQ12</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-1.156570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348957</th>\n",
       "      <td>201T</td>\n",
       "      <td>165</td>\n",
       "      <td>DMOG</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.975588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357660</th>\n",
       "      <td>201T</td>\n",
       "      <td>166</td>\n",
       "      <td>FTI-277</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.786602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366444</th>\n",
       "      <td>201T</td>\n",
       "      <td>167</td>\n",
       "      <td>OSU-03012</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.483735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375066</th>\n",
       "      <td>201T</td>\n",
       "      <td>170</td>\n",
       "      <td>Shikonin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.162809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383733</th>\n",
       "      <td>201T</td>\n",
       "      <td>171</td>\n",
       "      <td>AKT inhibitor VIII</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.682822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392553</th>\n",
       "      <td>201T</td>\n",
       "      <td>172</td>\n",
       "      <td>Embelin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.106883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>401355</th>\n",
       "      <td>201T</td>\n",
       "      <td>173</td>\n",
       "      <td>FH535</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.069789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410049</th>\n",
       "      <td>201T</td>\n",
       "      <td>175</td>\n",
       "      <td>PAC-1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.792672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418851</th>\n",
       "      <td>201T</td>\n",
       "      <td>176</td>\n",
       "      <td>IPA-3</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>4.494199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427635</th>\n",
       "      <td>201T</td>\n",
       "      <td>177</td>\n",
       "      <td>GSK650394</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.243024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436473</th>\n",
       "      <td>201T</td>\n",
       "      <td>178</td>\n",
       "      <td>BAY-61-3606</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.459334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445518</th>\n",
       "      <td>201T</td>\n",
       "      <td>179</td>\n",
       "      <td>5-Fluorouracil</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.738474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454446</th>\n",
       "      <td>201T</td>\n",
       "      <td>180</td>\n",
       "      <td>Thapsigargin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-5.512996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463113</th>\n",
       "      <td>201T</td>\n",
       "      <td>182</td>\n",
       "      <td>Obatoclax Mesylate</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.936866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471879</th>\n",
       "      <td>201T</td>\n",
       "      <td>184</td>\n",
       "      <td>BMS-754807</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-3.268468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480645</th>\n",
       "      <td>201T</td>\n",
       "      <td>185</td>\n",
       "      <td>Linsitinib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.972496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489420</th>\n",
       "      <td>201T</td>\n",
       "      <td>186</td>\n",
       "      <td>Bexarotene</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.335698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498069</th>\n",
       "      <td>201T</td>\n",
       "      <td>190</td>\n",
       "      <td>Bleomycin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-3.139471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506844</th>\n",
       "      <td>201T</td>\n",
       "      <td>192</td>\n",
       "      <td>LFM-A13</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>4.179227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515808</th>\n",
       "      <td>201T</td>\n",
       "      <td>193</td>\n",
       "      <td>GW-2580</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>5.879446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524835</th>\n",
       "      <td>201T</td>\n",
       "      <td>194</td>\n",
       "      <td>Luminespib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-4.751446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533781</th>\n",
       "      <td>201T</td>\n",
       "      <td>196</td>\n",
       "      <td>Phenformin</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>7.085319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542844</th>\n",
       "      <td>201T</td>\n",
       "      <td>197</td>\n",
       "      <td>Bryostatin 1</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-4.833935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551601</th>\n",
       "      <td>201T</td>\n",
       "      <td>199</td>\n",
       "      <td>Pazopanib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.328186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560403</th>\n",
       "      <td>201T</td>\n",
       "      <td>200</td>\n",
       "      <td>Dacinostat</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-3.535251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569196</th>\n",
       "      <td>201T</td>\n",
       "      <td>201</td>\n",
       "      <td>Epothilone B</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-7.695381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577998</th>\n",
       "      <td>201T</td>\n",
       "      <td>202</td>\n",
       "      <td>GSK1904529A</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.260885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588690</th>\n",
       "      <td>201T</td>\n",
       "      <td>203</td>\n",
       "      <td>BMS-345541</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>2.912202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598176</th>\n",
       "      <td>201T</td>\n",
       "      <td>204</td>\n",
       "      <td>Tipifarnib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.918427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607158</th>\n",
       "      <td>201T</td>\n",
       "      <td>205</td>\n",
       "      <td>Avagacestat</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>5.125651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618678</th>\n",
       "      <td>201T</td>\n",
       "      <td>206</td>\n",
       "      <td>Ruxolitinib</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>4.036701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627579</th>\n",
       "      <td>201T</td>\n",
       "      <td>207</td>\n",
       "      <td>AS601245</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.561223</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       CELL_LINE_NAME  DRUG_ID           DRUG_NAME DATASET   LN_IC50\n",
       "190089           201T      133         Doxorubicin   GDSC1 -3.770673\n",
       "198783           201T      134           Etoposide   GDSC1 -0.814180\n",
       "207405           201T      135         Gemcitabine   GDSC1 -0.298050\n",
       "216171           201T      136         Mitomycin-C   GDSC1 -4.472378\n",
       "224883           201T      140         Vinorelbine   GDSC1 -5.332884\n",
       "233550           201T      147           NSC-87877   GDSC1  4.680281\n",
       "242370           201T      150        Bicalutamide   GDSC1  2.754322\n",
       "251217           201T      151                QS11   GDSC1  1.992590\n",
       "261873           201T      152            CP466722   GDSC1  2.299660\n",
       "271251           201T      153         Midostaurin   GDSC1 -1.837530\n",
       "279918           201T      154          CHIR-99021   GDSC1  3.701566\n",
       "288567           201T      155           Ponatinib   GDSC1  0.494190\n",
       "297198           201T      156             AZD6482   GDSC1  2.597021\n",
       "305847           201T      157              JNK-9L   GDSC1 -0.979145\n",
       "314460           201T      158           PF-562271   GDSC1  1.235556\n",
       "323010           201T      159            HG6-64-1   GDSC1  0.271747\n",
       "331686           201T      163                 JQ1   GDSC1 -0.048784\n",
       "340344           201T      164                JQ12   GDSC1 -1.156570\n",
       "348957           201T      165                DMOG   GDSC1  2.975588\n",
       "357660           201T      166             FTI-277   GDSC1  1.786602\n",
       "366444           201T      167           OSU-03012   GDSC1  0.483735\n",
       "375066           201T      170            Shikonin   GDSC1 -0.162809\n",
       "383733           201T      171  AKT inhibitor VIII   GDSC1  0.682822\n",
       "392553           201T      172             Embelin   GDSC1  2.106883\n",
       "401355           201T      173               FH535   GDSC1  1.069789\n",
       "410049           201T      175               PAC-1   GDSC1  1.792672\n",
       "418851           201T      176               IPA-3   GDSC1  4.494199\n",
       "427635           201T      177           GSK650394   GDSC1  2.243024\n",
       "436473           201T      178         BAY-61-3606   GDSC1  1.459334\n",
       "445518           201T      179      5-Fluorouracil   GDSC1  3.738474\n",
       "454446           201T      180        Thapsigargin   GDSC1 -5.512996\n",
       "463113           201T      182  Obatoclax Mesylate   GDSC1 -0.936866\n",
       "471879           201T      184          BMS-754807   GDSC1 -3.268468\n",
       "480645           201T      185          Linsitinib   GDSC1 -0.972496\n",
       "489420           201T      186          Bexarotene   GDSC1  3.335698\n",
       "498069           201T      190           Bleomycin   GDSC1 -3.139471\n",
       "506844           201T      192             LFM-A13   GDSC1  4.179227\n",
       "515808           201T      193             GW-2580   GDSC1  5.879446\n",
       "524835           201T      194          Luminespib   GDSC1 -4.751446\n",
       "533781           201T      196          Phenformin   GDSC1  7.085319\n",
       "542844           201T      197        Bryostatin 1   GDSC1 -4.833935\n",
       "551601           201T      199           Pazopanib   GDSC1  2.328186\n",
       "560403           201T      200          Dacinostat   GDSC1 -3.535251\n",
       "569196           201T      201        Epothilone B   GDSC1 -7.695381\n",
       "577998           201T      202         GSK1904529A   GDSC1  1.260885\n",
       "588690           201T      203          BMS-345541   GDSC1  2.912202\n",
       "598176           201T      204          Tipifarnib   GDSC1  0.918427\n",
       "607158           201T      205         Avagacestat   GDSC1  5.125651\n",
       "618678           201T      206         Ruxolitinib   GDSC1  4.036701\n",
       "627579           201T      207            AS601245   GDSC1  0.561223"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_v1 = gene_expr[['CELL_LINE_NAME', 'DRUG_ID', 'DRUG_NAME', 'DATASET', 'LN_IC50']].sort_values(['CELL_LINE_NAME', 'DRUG_ID'])\n",
    "print(f\"Shape: {drug_cl_v1.shape}\")\n",
    "drug_cl_v1.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118673</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40457</th>\n",
       "      <td>CAMA-1</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389675</th>\n",
       "      <td>T-T</td>\n",
       "      <td>1022</td>\n",
       "      <td>AZD7762</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12105</th>\n",
       "      <td>A549</td>\n",
       "      <td>1017</td>\n",
       "      <td>Olaparib</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28933</th>\n",
       "      <td>BT-483</td>\n",
       "      <td>1038</td>\n",
       "      <td>NU7441</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208495</th>\n",
       "      <td>MDA-MB-330</td>\n",
       "      <td>1017</td>\n",
       "      <td>Olaparib</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208021</th>\n",
       "      <td>MDA-MB-231</td>\n",
       "      <td>1011</td>\n",
       "      <td>Navitoclax</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308862</th>\n",
       "      <td>OVKATE</td>\n",
       "      <td>1046</td>\n",
       "      <td>Wee1 Inhibitor</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108257</th>\n",
       "      <td>GRANTA-519</td>\n",
       "      <td>1862</td>\n",
       "      <td>MG-132</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76330</th>\n",
       "      <td>Daudi</td>\n",
       "      <td>1032</td>\n",
       "      <td>Afatinib</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       CELL_LINE_NAME  DRUG_ID       DRUG_NAME  count\n",
       "118673        HCC1937     1192      GSK269962A      6\n",
       "40457          CAMA-1     1192      GSK269962A      4\n",
       "389675            T-T     1022         AZD7762      4\n",
       "12105            A549     1017        Olaparib      4\n",
       "28933          BT-483     1038          NU7441      4\n",
       "208495     MDA-MB-330     1017        Olaparib      4\n",
       "208021     MDA-MB-231     1011      Navitoclax      4\n",
       "308862         OVKATE     1046  Wee1 Inhibitor      4\n",
       "108257     GRANTA-519     1862          MG-132      4\n",
       "76330           Daudi     1032        Afatinib      4"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_v1.groupby(['CELL_LINE_NAME', 'DRUG_ID', 'DRUG_NAME']).size().reset_index()\\\n",
    "    .rename(columns={0: 'count'}).sort_values(['count'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2846209</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.810152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2846216</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.810152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2846244</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.810152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4344253</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>3.781019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4344260</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>3.781019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4344288</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>3.781019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CELL_LINE_NAME  DRUG_ID   DRUG_NAME DATASET   LN_IC50\n",
       "2846209        HCC1937     1192  GSK269962A   GDSC1  3.810152\n",
       "2846216        HCC1937     1192  GSK269962A   GDSC1  3.810152\n",
       "2846244        HCC1937     1192  GSK269962A   GDSC1  3.810152\n",
       "4344253        HCC1937     1192  GSK269962A   GDSC2  3.781019\n",
       "4344260        HCC1937     1192  GSK269962A   GDSC2  3.781019\n",
       "4344288        HCC1937     1192  GSK269962A   GDSC2  3.781019"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_v1[(drug_cl_v1.CELL_LINE_NAME=='HCC1937') & (drug_cl_v1.DRUG_ID==1192)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(310946, 5)\n",
      "(135575, 5)\n"
     ]
    }
   ],
   "source": [
    "print(drug_cl_v1[drug_cl_v1.DATASET=='GDSC1'].shape)\n",
    "print(drug_cl_v1[drug_cl_v1.DATASET=='GDSC2'].shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- NOTE: if there is a same cell-line drug combination the ic50 values only differ for different datasets\n",
    "  - if for the same cell-line drug tuple the dataset is e.g. GDSC1, then the ic50 values will also be the same\n",
    "- Thus, we need to decide which dataset we choose\n",
    "- Since GDSC1 provides more data points, we will go on with GDSC1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(310904, 5)\n",
      "(135242, 5)\n"
     ]
    }
   ],
   "source": [
    "drug_cl_GDSC1 = drug_cl_v1[drug_cl_v1.DATASET=='GDSC1'].drop_duplicates()\n",
    "print(drug_cl_GDSC1.shape)\n",
    "\n",
    "drug_cl_GDSC2 = drug_cl_v1[drug_cl_v1.DATASET=='GDSC2'].drop_duplicates()\n",
    "print(drug_cl_GDSC2.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2846209</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>3.810152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CELL_LINE_NAME  DRUG_ID DATASET   LN_IC50\n",
       "2846209        HCC1937     1192   GDSC1  3.810152"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the duplicated rows from above got removed.\n",
    "drug_cl_GDSC1[(drug_cl_GDSC1.CELL_LINE_NAME=='HCC1937') & (drug_cl_GDSC1.DRUG_ID==1192)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4344253</th>\n",
       "      <td>HCC1937</td>\n",
       "      <td>1192</td>\n",
       "      <td>GSK269962A</td>\n",
       "      <td>GDSC2</td>\n",
       "      <td>3.781019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CELL_LINE_NAME  DRUG_ID   DRUG_NAME DATASET   LN_IC50\n",
       "4344253        HCC1937     1192  GSK269962A   GDSC2  3.781019"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_GDSC2[(drug_cl_GDSC2.CELL_LINE_NAME=='HCC1937') & (drug_cl_GDSC2.DRUG_ID==1192)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_NAME</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1037</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1036</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1033</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         DRUG_NAME\n",
       "DRUG_ID           \n",
       "1                1\n",
       "1039             1\n",
       "1037             1\n",
       "1036             1\n",
       "1033             1\n",
       "1032             1\n",
       "1031             1\n",
       "1030             1\n",
       "1029             1\n",
       "1028             1"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_GDSC2[['DRUG_ID', 'DRUG_NAME']].groupby(['DRUG_ID']).nunique().sort_values(['DRUG_NAME'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Dactinomycin</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fulvestrant</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Oxaliplatin</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Docetaxel</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ulixertinib</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uprosertib</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD173074</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olaparib</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Osimertinib</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P22077</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              DRUG_ID\n",
       "DRUG_NAME            \n",
       "Dactinomycin        2\n",
       "Fulvestrant         2\n",
       "Oxaliplatin         2\n",
       "Docetaxel           2\n",
       "Ulixertinib         2\n",
       "Uprosertib          2\n",
       "PD173074            1\n",
       "Olaparib            1\n",
       "Osimertinib         1\n",
       "P22077              1"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_cl_GDSC2[['DRUG_ID', 'DRUG_NAME']].groupby(['DRUG_NAME']).nunique().sort_values(['DRUG_ID'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These duplicates need to be removed. We need a 1-to-1 mapping from the `DRUG_NAME`'s to the `DRUG_ID`'s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dactinomycin\n",
      "1911    740\n",
      "1811    728\n",
      "Name: DRUG_ID, dtype: int64\n",
      "Fulvestrant\n",
      "1200    764\n",
      "1816    728\n",
      "Name: DRUG_ID, dtype: int64\n",
      "Oxaliplatin\n",
      "1089    802\n",
      "1806    728\n",
      "Name: DRUG_ID, dtype: int64\n",
      "Docetaxel\n",
      "1007    766\n",
      "1819    669\n",
      "Name: DRUG_ID, dtype: int64\n",
      "Ulixertinib\n",
      "1908    752\n",
      "2047    749\n",
      "Name: DRUG_ID, dtype: int64\n",
      "Uprosertib\n",
      "2106    745\n",
      "1553    735\n",
      "Name: DRUG_ID, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "all_non_uniqs = drug_cl_GDSC2[['DRUG_ID', 'DRUG_NAME']].groupby(['DRUG_NAME']).nunique().sort_values(['DRUG_ID'], ascending=False)\n",
    "all_non_uniqs = all_non_uniqs[all_non_uniqs.DRUG_ID>1]\n",
    "\n",
    "for non_unique in all_non_uniqs.index:\n",
    "    print(non_unique)\n",
    "    print(drug_cl_GDSC2[drug_cl_GDSC2.DRUG_NAME==non_unique]['DRUG_ID'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This is now the drug dataset which can be encoded by the drug encode side of the bi-modal network"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save drug dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_FEATURES = '../../datasets/datasets_for_model_building/'\n",
    "\n",
    "drug_cl_GDSC1.to_pickle(f'{PATH_TO_FEATURES}drugs_sparse_gdsc1.pkl')\n",
    "drug_cl_GDSC2.to_pickle(f'{PATH_TO_FEATURES}drugs_sparse_gdsc2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "READ = False\n",
    "\n",
    "if READ:\n",
    "    with open(f'{PATH_TO_FEATURES}drugs_sparse_gdsc1.pkl', 'rb') as f: \n",
    "        drug_cl_GDSC1 = pickle.load(f) \n",
    "    with open(f'{PATH_TO_FEATURES}drugs_sparse_gdsc2.pkl', 'rb') as f: \n",
    "        drug_cl_GDSC2 = pickle.load(f)         "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Classifying the `protein_mutation` column\n",
    "\n",
    "This will be a different approach compared to checking the `cancer_driver` column. Now we will classify the `protein_mutation` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_protein_mutation_col(protein_mut):\n",
    "    # This method is adapted from MOLI.\n",
    "    # https://academic.oup.com/bioinformatics/article/35/14/i501/5529255\n",
    "    mut = protein_mut.split('p.')[1]\n",
    "    if bool(re.match(re.compile(\"^[A-Z][0-9]+[A-Z]$\"), mut)): return \"missense\"\n",
    "    elif bool(re.match(re.compile(\"^[A-Z]+[0-9]+[A-Z]+$\"), mut)): return \"inframe_indel\"\n",
    "    elif \"*\" in mut: return \"stop_gain_or_fs\"\n",
    "    elif \"?\" in mut: return \"ess_splice\"\n",
    "    elif bool(re.match(re.compile(\"^\\-[0-9]+[A-Z]+$\"), mut)): return \"unrecognized_point\"\n",
    "    elif bool(re.match(re.compile(\"^\\-[0-9]+$\"), mut)): return \"unrecognized_truncating\"\n",
    "    elif bool(re.match(re.compile(\"^[A-Z]+[0-9]+\\-$\"), mut)): return \"unrecognized_point2\"\n",
    "    else: return \"unrecognized\"    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We removed the rows which have a missing protein_mutation_position and classified the protein_mutation column.\n",
      "mutations_v7 Shape: (44952, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "      <th>protein_mutation_pos</th>\n",
       "      <th>protein_mutation_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.A95S</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4138</td>\n",
       "      <td>(95, 95)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Daudi</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.S282T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.1212</td>\n",
       "      <td>(282, 282)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>NY</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.V80I</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4706</td>\n",
       "      <td>(80, 80)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>HSC-3</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.E105D</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9444</td>\n",
       "      <td>(105, 105)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>CW-2</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.G276fs*99</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6444</td>\n",
       "      <td>(276, 99)</td>\n",
       "      <td>stop_gain_or_fs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CELL_LINE_NAME gene_symbol protein_mutation  cancer_driver     vaf  \\\n",
       "10         NUGC-4       VGLL4           p.A95S          False  0.4138   \n",
       "19          Daudi       VGLL4          p.S282T          False  0.1212   \n",
       "63             NY       VGLL4           p.V80I          False  0.4706   \n",
       "75          HSC-3       VGLL4          p.E105D          False  0.9444   \n",
       "81           CW-2       VGLL4      p.G276fs*99          False  0.6444   \n",
       "\n",
       "   protein_mutation_pos protein_mutation_group  \n",
       "10             (95, 95)               missense  \n",
       "19           (282, 282)               missense  \n",
       "63             (80, 80)               missense  \n",
       "75           (105, 105)               missense  \n",
       "81            (276, 99)        stop_gain_or_fs  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v7 = mutations_v6[mutations_v6.protein_mutation_pos != (None, None)][\n",
    "    ['CELL_LINE_NAME', 'gene_symbol', 'protein_mutation', 'cancer_driver', 'vaf', 'protein_mutation_pos']\n",
    "]\n",
    "\n",
    "mutations_v7['protein_mutation_group'] = mutations_v7['protein_mutation'].apply(classify_protein_mutation_col)\n",
    "mutations_v7.groupby(['protein_mutation_group']).size().sort_values(ascending=False)\n",
    "\n",
    "print(\"We removed the rows which have a missing protein_mutation_position and classified the protein_mutation column.\")\n",
    "print(f\"mutations_v7 Shape: {mutations_v7.shape}\")\n",
    "mutations_v7.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "      <th>protein_mutation_pos</th>\n",
       "      <th>protein_mutation_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.A95S</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4138</td>\n",
       "      <td>(95, 95)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>p.A95S</td>\n",
       "      <td>False</td>\n",
       "      <td>0.2308</td>\n",
       "      <td>(95, 95)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>PLA2G4A</td>\n",
       "      <td>p.Q513K</td>\n",
       "      <td>False</td>\n",
       "      <td>0.3427</td>\n",
       "      <td>(513, 513)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>789</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>PLA2G4A</td>\n",
       "      <td>p.Q513K</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5354</td>\n",
       "      <td>(513, 513)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9359</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>p.I456T</td>\n",
       "      <td>False</td>\n",
       "      <td>0.8032</td>\n",
       "      <td>(456, 456)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224245</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>SPR</td>\n",
       "      <td>p.N154Y</td>\n",
       "      <td>False</td>\n",
       "      <td>0.1677</td>\n",
       "      <td>(154, 154)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227101</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>KIF20A</td>\n",
       "      <td>p.K827Q</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5000</td>\n",
       "      <td>(827, 827)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227167</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>KIF20A</td>\n",
       "      <td>p.K827Q</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4468</td>\n",
       "      <td>(827, 827)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231058</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>TMEM109</td>\n",
       "      <td>p.R37Q</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6159</td>\n",
       "      <td>(37, 37)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231088</th>\n",
       "      <td>NUGC-4</td>\n",
       "      <td>TMEM109</td>\n",
       "      <td>p.R37Q</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6471</td>\n",
       "      <td>(37, 37)</td>\n",
       "      <td>missense</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CELL_LINE_NAME gene_symbol protein_mutation  cancer_driver     vaf  \\\n",
       "10             NUGC-4       VGLL4           p.A95S          False  0.4138   \n",
       "243            NUGC-4       VGLL4           p.A95S          False  0.2308   \n",
       "423            NUGC-4     PLA2G4A          p.Q513K          False  0.3427   \n",
       "789            NUGC-4     PLA2G4A          p.Q513K          False  0.5354   \n",
       "9359           NUGC-4       BRCA1          p.I456T          False  0.8032   \n",
       "...               ...         ...              ...            ...     ...   \n",
       "224245         NUGC-4         SPR          p.N154Y          False  0.1677   \n",
       "227101         NUGC-4      KIF20A          p.K827Q          False  0.5000   \n",
       "227167         NUGC-4      KIF20A          p.K827Q          False  0.4468   \n",
       "231058         NUGC-4     TMEM109           p.R37Q          False  0.6159   \n",
       "231088         NUGC-4     TMEM109           p.R37Q          False  0.6471   \n",
       "\n",
       "       protein_mutation_pos protein_mutation_group  \n",
       "10                 (95, 95)               missense  \n",
       "243                (95, 95)               missense  \n",
       "423              (513, 513)               missense  \n",
       "789              (513, 513)               missense  \n",
       "9359             (456, 456)               missense  \n",
       "...                     ...                    ...  \n",
       "224245           (154, 154)               missense  \n",
       "227101           (827, 827)               missense  \n",
       "227167           (827, 827)               missense  \n",
       "231058             (37, 37)               missense  \n",
       "231088             (37, 37)               missense  \n",
       "\n",
       "[86 rows x 7 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v7[mutations_v7.CELL_LINE_NAME == 'NUGC-4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c78b81650a0bd32063743affb6953ff71b1a0dba806fbca9e2db842718495748"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4 ('master-thesis-log')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
