{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import time\n",
    "import os\n",
    "import gzip\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "\n",
    "from io import BytesIO\n",
    "\n",
    "sns.set_theme(style=\"white\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Link Interaction Data Insights\n",
    "\n",
    "To create a Graph Neural Network (GNN) we will a way to link two nodes (genes) with one another. For now, we have investigated and developed candidate node-level features. In this notebook we will investigate how we can find linking information for genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material\n",
      "38944496        0 drwxr-xr-x    5 cwoest           staff                 160 Jun 29 11:42 ../../datasets/string/\n",
      "38944659       24 -rw-r--r--    1 cwoest           staff                8196 Jun 29 11:42 ../../datasets/string//.DS_Store\n",
      "38709030   225680 -rw-r--r--    1 cwoest           staff            115547735 Jun  9 16:28 ../../datasets/string//9606.protein.links.detailed.v11.5.txt.gz\n",
      "38944498     3720 -rw-r--r--    1 cwoest           staff             1901833 Jun 29 11:09 ../../datasets/string//9606.protein.info.v11.5.txt.gz\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!find ../../datasets/string/ -ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "According to [this description](https://string-db.org/network/9606.ENSP00000379496) of the STRING database the database contains the following informations.\n",
    "\n",
    "<ins>__Nodes__:</ins> \n",
    "> Network nodes represent proteins.\n",
    "- _splice isoforms or post-translational modifications are collapsed, i.e. each node represents all the proteins produced by a single, protein-coding gene locus._\n",
    "\n",
    "<ins>__Edges__:</ins>\n",
    "> Edges represent protein-protein associations.\n",
    "- _associations are meant to be specific and meaningful, i.e. proteins jointly contribute to a shared function; this does not necessarily mean they are physically binding to each other._\n",
    "- __Known Interactions__\n",
    "  - _from curated databases_ (`database`)\n",
    "  - _experimentally determined_ (`experimental`)\n",
    "- __Predicted Interactions__\n",
    "  - _gene neighborhood_ (`neighborhood`)\n",
    "  - _gene fusions_ (`fusion`)\n",
    "  - _gene co-occurence_ (`cooccurence`)\n",
    "- __Others__\n",
    "  - _textmining_ (`textmining`)\n",
    "  - _co-expression_ (`coexpression`)\n",
    "  - _protein homology_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_LINKING_DATA = '../../datasets/string/'\n",
    "PATH_TO_SAVE_DATA_TO = '../../datasets/string/my_datasets/'\n",
    "\n",
    "PROTEIN_LINKS_FILE = '9606.protein.links.detailed.v11.5.txt.gz'\n",
    "PROTEIN_INFO_FILE = '9606.protein.info.v11.5.txt.gz'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Landmark Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `landmark_genes.csv` took 0.05927 seconds to import. It has shape (978, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Type</th>\n",
       "      <th>RNA-Seq Correlation</th>\n",
       "      <th>RNA-Seq Correlation Self-Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3638</td>\n",
       "      <td>INSIG1</td>\n",
       "      <td>insulin induced gene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2309</td>\n",
       "      <td>FOXO3</td>\n",
       "      <td>forkhead box O3</td>\n",
       "      <td>Forkhead boxes</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>CDH3</td>\n",
       "      <td>cadherin 3</td>\n",
       "      <td>Type I classical cadherins</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entrez ID  Symbol                    Name                 Gene Family  \\\n",
       "0       3638  INSIG1  insulin induced gene 1                         NaN   \n",
       "1       2309   FOXO3         forkhead box O3              Forkhead boxes   \n",
       "2       1001    CDH3              cadherin 3  Type I classical cadherins   \n",
       "\n",
       "       Type  RNA-Seq Correlation  RNA-Seq Correlation Self-Rank  \n",
       "0  landmark                  NaN                            NaN  \n",
       "1  landmark                  NaN                            NaN  \n",
       "2  landmark                  NaN                            NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "PATH_TO_GDSC_DATASETS = '../../datasets/gdsc/'\n",
    "FILENAME_LANDMARK_GENES = 'landmark_genes.csv' \n",
    "\n",
    "start = time.time()\n",
    "landmark_genes = pd.read_csv(f'{PATH_TO_GDSC_DATASETS}{FILENAME_LANDMARK_GENES}', sep=\"\\t\")\n",
    "print(f\"File `{FILENAME_LANDMARK_GENES}` took {time.time()-start:.5f} seconds to import. It has shape {landmark_genes.shape}\")\n",
    "landmark_genes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique landmark genes with an index were written to ´../../datasets/string/my_datasets/uniq_landmark_genes_with_index.csv´.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>INSIG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>FOXO3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CDH3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ORC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ITGAE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  Symbol\n",
       "0      0  INSIG1\n",
       "1      1   FOXO3\n",
       "2      2    CDH3\n",
       "3      3    ORC1\n",
       "4      4   ITGAE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Add a unique number (index) for each distinct landmark gene. \n",
    "# This will be used as an index for the node in the graph instead of the raw \n",
    "# gene symbol name.\n",
    "UNIQ_LANDMARK_GENES = landmark_genes.reset_index()[['index', 'Symbol']].drop_duplicates()\n",
    "UNIQ_LANDMARK_GENES.to_csv(f'{PATH_TO_SAVE_DATA_TO}uniq_landmark_genes_with_index.csv', sep=',', header=True, index=False)\n",
    "print(f\"Unique landmark genes with an index were written to ´{PATH_TO_SAVE_DATA_TO}uniq_landmark_genes_with_index.csv´.\")\n",
    "UNIQ_LANDMARK_GENES = pd.read_csv(f'{PATH_TO_SAVE_DATA_TO}uniq_landmark_genes_with_index.csv', sep=',')\n",
    "UNIQ_LANDMARK_GENES.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Linking Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11938498, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000361263</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000409666</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  neighborhood  fusion  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000379496             0       0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000314067             0       0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000263116             0       0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000361263             0       0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000409666             0       0   \n",
       "\n",
       "   cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0            0            54             0         0         144   \n",
       "1            0             0           180         0          61   \n",
       "2            0            62           152         0         101   \n",
       "3            0             0           161         0          64   \n",
       "4            0            82           213         0          72   \n",
       "\n",
       "   combined_score  \n",
       "0             155  \n",
       "1             197  \n",
       "2             222  \n",
       "3             181  \n",
       "4             270  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ----------------- #\n",
    "# Read Linking Data\n",
    "# ----------------- #\n",
    "\n",
    "# Read content of .gz file as bytes.\n",
    "contents = gzip.open(f'{PATH_TO_LINKING_DATA}{PROTEIN_LINKS_FILE}', \"rb\").read()\n",
    "\n",
    "# Convert bytes to pd.DataFrame.\n",
    "data = BytesIO(contents)\n",
    "protein_links_v1 = pd.read_csv(data, sep=' ')\n",
    "\n",
    "print(protein_links_v1.shape)\n",
    "protein_links_v1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Column | Description |\n",
    "| ------ | ----------- | \n",
    "| `neighboorhood` | |\n",
    "| `fusion` | |\n",
    "| `cooccurence` | |\n",
    "| `coexpression` | |\n",
    "| `experimental` | |\n",
    "| `database` | |\n",
    "| `textmining` | |\n",
    "| `combined_score` | |\n",
    "\n",
    "- The data holds interaction information about two proteins.\n",
    "- `9606`: The taxonomy ID of Homo Sapiens (https://www.ncbi.nlm.nih.gov/protein?term=txid9606[Organism]).'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.20</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.88</td>\n",
       "      <td>62.38</td>\n",
       "      <td>68.92</td>\n",
       "      <td>24.14</td>\n",
       "      <td>180.26</td>\n",
       "      <td>280.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>20.39</td>\n",
       "      <td>7.57</td>\n",
       "      <td>13.41</td>\n",
       "      <td>96.64</td>\n",
       "      <td>119.62</td>\n",
       "      <td>132.79</td>\n",
       "      <td>152.82</td>\n",
       "      <td>167.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>150.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>67.00</td>\n",
       "      <td>175.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>54.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>161.00</td>\n",
       "      <td>216.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>76.00</td>\n",
       "      <td>129.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>241.00</td>\n",
       "      <td>309.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>385.00</td>\n",
       "      <td>900.00</td>\n",
       "      <td>448.00</td>\n",
       "      <td>999.00</td>\n",
       "      <td>999.00</td>\n",
       "      <td>900.00</td>\n",
       "      <td>997.00</td>\n",
       "      <td>999.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      neighborhood       fusion  cooccurence coexpression experimental  \\\n",
       "count  11938498.00  11938498.00  11938498.00  11938498.00  11938498.00   \n",
       "mean          3.20         0.16         0.88        62.38        68.92   \n",
       "std          20.39         7.57        13.41        96.64       119.62   \n",
       "min           0.00         0.00         0.00         0.00         0.00   \n",
       "25%           0.00         0.00         0.00         0.00         0.00   \n",
       "50%           0.00         0.00         0.00        54.00         0.00   \n",
       "75%           0.00         0.00         0.00        76.00       129.00   \n",
       "max         385.00       900.00       448.00       999.00       999.00   \n",
       "\n",
       "          database   textmining combined_score  \n",
       "count  11938498.00  11938498.00    11938498.00  \n",
       "mean         24.14       180.26         280.01  \n",
       "std         132.79       152.82         167.39  \n",
       "min           0.00         0.00         150.00  \n",
       "25%           0.00        67.00         175.00  \n",
       "50%           0.00       161.00         216.00  \n",
       "75%           0.00       241.00         309.00  \n",
       "max         900.00       997.00         999.00  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_links_v1.describe().apply(lambda s: s.apply('{0:.2f}'.format))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19566, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#string_protein_id</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>protein_size</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>180</td>\n",
       "      <td>ADP-ribosylation factor 5; GTP-binding protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000412</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>277</td>\n",
       "      <td>Cation-dependent mannose-6-phosphate receptor;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000001008</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>459</td>\n",
       "      <td>Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000001146</td>\n",
       "      <td>CYP26B1</td>\n",
       "      <td>512</td>\n",
       "      <td>Cytochrome P450 26B1; Involved in the metaboli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000002125</td>\n",
       "      <td>NDUFAF7</td>\n",
       "      <td>441</td>\n",
       "      <td>Protein arginine methyltransferase NDUFAF7, mi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #string_protein_id preferred_name  protein_size  \\\n",
       "0  9606.ENSP00000000233           ARF5           180   \n",
       "1  9606.ENSP00000000412           M6PR           277   \n",
       "2  9606.ENSP00000001008          FKBP4           459   \n",
       "3  9606.ENSP00000001146        CYP26B1           512   \n",
       "4  9606.ENSP00000002125        NDUFAF7           441   \n",
       "\n",
       "                                          annotation  \n",
       "0  ADP-ribosylation factor 5; GTP-binding protein...  \n",
       "1  Cation-dependent mannose-6-phosphate receptor;...  \n",
       "2  Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...  \n",
       "3  Cytochrome P450 26B1; Involved in the metaboli...  \n",
       "4  Protein arginine methyltransferase NDUFAF7, mi...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ---------------------- #\n",
    "# Read Protein Info Data\n",
    "# ---------------------- #\n",
    "\n",
    "# Read content of .gz file as bytes.\n",
    "contents = gzip.open(f'{PATH_TO_LINKING_DATA}{PROTEIN_INFO_FILE}', \"rb\").read()\n",
    "\n",
    "# Convert bytes to pd.DataFrame.\n",
    "data = BytesIO(contents)\n",
    "protein_info_v1 = pd.read_csv(data, sep='\\t')\n",
    "\n",
    "print(protein_info_v1.shape)\n",
    "protein_info_v1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Column              | Description | Example |\n",
    "| ------------------- | ----------- | ------- |\n",
    "| `string_protein_id` | Protein identifier | `9606.ENSP00000000233\t` |\n",
    "| `preferred_name`    | Gene identifier. Same as `GENE_SYMBOL`. | `M6PR` |\n",
    "| `protein_size`      | Size of the protein in aa (amino acids)| `277` |\n",
    "| `annotation`        | Full name of the protein. | `Cation-dependent mannose-6-phosphate receptor;...` |\n",
    "\n",
    "- The `string_protein_id` protein is encoded by the `preferred_name` gene\n",
    "- The `preferred_name` gene encodes the `string_protein_id` protein\n",
    "- `9606` stands for the taxonomy ID of Homo Sapiens (https://www.ncbi.nlm.nih.gov/protein?term=txid9606[Organism]).'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Data Wrangling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Links"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000000233  ENSP00000379496             0       0            0   \n",
       "1  ENSP00000000233  ENSP00000314067             0       0            0   \n",
       "2  ENSP00000000233  ENSP00000263116             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \n",
       "0            54             0         0         144             155  \n",
       "1             0           180         0          61             197  \n",
       "2            62           152         0         101             222  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exclude the Homo Sapiens taxonomy ID from the protein columns.\n",
    "protein_links_v2 = protein_links_v1.copy()\n",
    "protein_links_v2.protein1 = protein_links_v2.protein1.str[5:]\n",
    "protein_links_v2.protein2 = protein_links_v2.protein2.str[5:]\n",
    "protein_links_v2.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>string_protein_id</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>protein_size</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>180</td>\n",
       "      <td>ADP-ribosylation factor 5; GTP-binding protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000412</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>277</td>\n",
       "      <td>Cation-dependent mannose-6-phosphate receptor;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>459</td>\n",
       "      <td>Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  string_protein_id preferred_name  protein_size  \\\n",
       "0   ENSP00000000233           ARF5           180   \n",
       "1   ENSP00000000412           M6PR           277   \n",
       "2   ENSP00000001008          FKBP4           459   \n",
       "\n",
       "                                          annotation  \n",
       "0  ADP-ribosylation factor 5; GTP-binding protein...  \n",
       "1  Cation-dependent mannose-6-phosphate receptor;...  \n",
       "2  Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename protein identifier column.\n",
    "protein_info_v2 = protein_info_v1.rename(columns={'#string_protein_id': 'string_protein_id'}, inplace=False)\n",
    "\n",
    "# Exclude the Homo Sapiens taxonomy ID from the protein columns.\n",
    "protein_info_v2.string_protein_id = protein_info_v2.string_protein_id.str[5:]\n",
    "protein_info_v2.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Understanding The Data\n",
    "\n",
    "> Are the `preferred_name`´s in the protein info data the genes which direct the production of the proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `landmark_genes.csv` took 0.00588 seconds to import. It has shape (978, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "      <th>Gene Family</th>\n",
       "      <th>Type</th>\n",
       "      <th>RNA-Seq Correlation</th>\n",
       "      <th>RNA-Seq Correlation Self-Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3638</td>\n",
       "      <td>INSIG1</td>\n",
       "      <td>insulin induced gene 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2309</td>\n",
       "      <td>FOXO3</td>\n",
       "      <td>forkhead box O3</td>\n",
       "      <td>Forkhead boxes</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1001</td>\n",
       "      <td>CDH3</td>\n",
       "      <td>cadherin 3</td>\n",
       "      <td>Type I classical cadherins</td>\n",
       "      <td>landmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Entrez ID  Symbol                    Name                 Gene Family  \\\n",
       "0       3638  INSIG1  insulin induced gene 1                         NaN   \n",
       "1       2309   FOXO3         forkhead box O3              Forkhead boxes   \n",
       "2       1001    CDH3              cadherin 3  Type I classical cadherins   \n",
       "\n",
       "       Type  RNA-Seq Correlation  RNA-Seq Correlation Self-Rank  \n",
       "0  landmark                  NaN                            NaN  \n",
       "1  landmark                  NaN                            NaN  \n",
       "2  landmark                  NaN                            NaN  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -------------------------------------------------------------------- #\n",
    "# Get the landmark genes which define the set of genes which were used\n",
    "# in the previous notebooks to create candidate features.\n",
    "# -------------------------------------------------------------------- #\n",
    "\n",
    "PATH_TO_GDSC_DATASETS = '../../datasets/gdsc/'\n",
    "FILENAME_LANDMARK_GENES = 'landmark_genes.csv' \n",
    "\n",
    "start = time.time()\n",
    "landmark_genes = pd.read_csv(f'{PATH_TO_GDSC_DATASETS}{FILENAME_LANDMARK_GENES}', sep=\"\\t\")\n",
    "print(f\"File `{FILENAME_LANDMARK_GENES}` took {time.time()-start:.5f} seconds to import. It has shape {landmark_genes.shape}\")\n",
    "landmark_genes.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of 19563 unique ´preferred_name´s in the protein info data 965 can also be found in the 978 unique landmark genes.\n",
      "Thus, 98.67% of all unique landmark genes are existent in the protein info data.\n"
     ]
    }
   ],
   "source": [
    "# Check if the 'Symbol'´s can be found in the 'preferred_name'´s of the protein info data.\n",
    "uniq_preferred_names = np.unique(protein_info_v2.preferred_name.values)\n",
    "uniq_landmark_genes = np.unique(landmark_genes.Symbol.values)\n",
    "count = sum(v in landmark_genes.Symbol.values for v in protein_info_v2.preferred_name.values)\n",
    "print(f\"Out of {len(uniq_preferred_names)} unique ´preferred_name´s in the protein info data {count} can also be found in the {len(uniq_landmark_genes)} unique landmark genes.\")\n",
    "print(f\"Thus, {round(100*count/len(uniq_landmark_genes), 2)}% of all unique landmark genes are existent in the protein info data.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Conclusion__: \n",
    "- We validated that the `preferred_name` is standing for the specific gene, which was called `GENE_SYMBOL` in other notebooks or `Symbol` in the landmark gene dataset respectively.\n",
    "- We can map the `preferred_name` (gene symbol) of the protein info dataset to the protein link dataset to get the corresponding gene per protein. \n",
    "  - This way we map a protein-protein interaction to the corresponding gene-gene interaction."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Mapping Of Genes To Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>string_protein_id</th>\n",
       "      <th>preferred_name</th>\n",
       "      <th>protein_size</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>180</td>\n",
       "      <td>ADP-ribosylation factor 5; GTP-binding protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000412</td>\n",
       "      <td>M6PR</td>\n",
       "      <td>277</td>\n",
       "      <td>Cation-dependent mannose-6-phosphate receptor;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>459</td>\n",
       "      <td>Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  string_protein_id preferred_name  protein_size  \\\n",
       "0   ENSP00000000233           ARF5           180   \n",
       "1   ENSP00000000412           M6PR           277   \n",
       "2   ENSP00000001008          FKBP4           459   \n",
       "\n",
       "                                          annotation  \n",
       "0  ADP-ribosylation factor 5; GTP-binding protein...  \n",
       "1  Cation-dependent mannose-6-phosphate receptor;...  \n",
       "2  Peptidyl-prolyl cis-trans isomerase FKBP4; Imm...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_info_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000000233  ENSP00000379496             0       0            0   \n",
       "1  ENSP00000000233  ENSP00000314067             0       0            0   \n",
       "2  ENSP00000000233  ENSP00000263116             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \n",
       "0            54             0         0         144             155  \n",
       "1             0           180         0          61             197  \n",
       "2            62           152         0         101             222  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_links_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Shape:\n",
      "        before joining : (11938498, 10)\n",
      "        after  joining : (11938498, 12)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>PDE1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>PAK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAB36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000361263</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>181</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAPGEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000409666</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>SUMO3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000000233  ENSP00000379496             0       0            0   \n",
       "1  ENSP00000000233  ENSP00000314067             0       0            0   \n",
       "2  ENSP00000000233  ENSP00000263116             0       0            0   \n",
       "3  ENSP00000000233  ENSP00000361263             0       0            0   \n",
       "4  ENSP00000000233  ENSP00000409666             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \\\n",
       "0            54             0         0         144             155   \n",
       "1             0           180         0          61             197   \n",
       "2            62           152         0         101             222   \n",
       "3             0           161         0          64             181   \n",
       "4            82           213         0          72             270   \n",
       "\n",
       "  gene_symbol1 gene_symbol2  \n",
       "0         ARF5        PDE1C  \n",
       "1         ARF5         PAK2  \n",
       "2         ARF5        RAB36  \n",
       "3         ARF5      RAPGEF1  \n",
       "4         ARF5        SUMO3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------------------ #\n",
    "# Map the corresponding genes to the proteins in the protein link dataset.\n",
    "# ------------------------------------------------------------------------ #\n",
    "# Get the gene symbols for the protein1 column.\n",
    "protein_links_v3 = protein_links_v2.merge(right    = protein_info_v2[['string_protein_id', 'preferred_name']],\n",
    "                                          how      = 'left',\n",
    "                                          left_on  = 'protein1',\n",
    "                                          right_on = 'string_protein_id')\n",
    "protein_links_v3.rename(columns={'preferred_name': 'gene_symbol1'}, inplace=True)\n",
    "protein_links_v3.drop(['string_protein_id'], axis=1, inplace=True)\n",
    "\n",
    "# Get the gene symbols for the protein2 column.\n",
    "protein_links_v4 = protein_links_v3.merge(right    = protein_info_v2[['string_protein_id', 'preferred_name']],\n",
    "                                          how      = 'left',\n",
    "                                          left_on  = 'protein2',\n",
    "                                          right_on = 'string_protein_id')\n",
    "protein_links_v4.rename(columns={'preferred_name': 'gene_symbol2'}, inplace=True)\n",
    "protein_links_v4.drop(['string_protein_id'], axis=1, inplace=True)\n",
    "\n",
    "print(f\"\"\"\n",
    "    Shape:\n",
    "        before joining : {protein_links_v2.shape}\n",
    "        after  joining : {protein_links_v4.shape}\n",
    "\"\"\")                                                                                  \n",
    "protein_links_v4.head(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_symbol1    0\n",
       "gene_symbol2    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if for any protein there hasn't been found a correspoding gene (symbol).\n",
    "protein_links_v4[['gene_symbol1', 'gene_symbol2']].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## EDA\n",
    "\n",
    "- [ ] How many different genes are in the protein linking table (to estimate the maximal size of the graph)?\n",
    "- [ ] How many different gene-gene interactions are there? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    There are \n",
      "        19382 unique gene_symbol1\n",
      "        19382 unique gene_symbol2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Unique number of genes per column.\n",
    "uniq_gene1 = np.unique(protein_links_v4.gene_symbol1.values)\n",
    "uniq_gene2 = np.unique(protein_links_v4.gene_symbol2.values)\n",
    "\n",
    "print(f\"\"\"\n",
    "    There are \n",
    "        {len(uniq_gene1)} unique gene_symbol1\n",
    "        {len(uniq_gene2)} unique gene_symbol2\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>freq</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>11936220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   freq     count\n",
       "0     1  11936220\n",
       "1     2      1139"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Unique number of gene-gene combinations.\n",
    "freq_per_tuple = protein_links_v4.groupby(['gene_symbol1', 'gene_symbol2']).size().reset_index(name='freq', inplace=False).sort_values(['freq'], ascending=False)\n",
    "freq_per_tuple.groupby(['freq']).size().reset_index(name='count', inplace=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Some `(gene_symbol1, gene_symbol2)` tuples occur two times.\n",
    "- However, the majority occurs only one time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Sparse Linking Dataset Using Landmark genes\n",
    "\n",
    "- [ ] Sparse down the linking dataset by only using rows where both, `gene_symbol1` and `gene_symbol2` are in the set of landmark genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (11938498, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>PDE1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>PAK2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "      <td>ARF5</td>\n",
       "      <td>RAB36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000000233  ENSP00000379496             0       0            0   \n",
       "1  ENSP00000000233  ENSP00000314067             0       0            0   \n",
       "2  ENSP00000000233  ENSP00000263116             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \\\n",
       "0            54             0         0         144             155   \n",
       "1             0           180         0          61             197   \n",
       "2            62           152         0         101             222   \n",
       "\n",
       "  gene_symbol1 gene_symbol2  \n",
       "0         ARF5        PDE1C  \n",
       "1         ARF5         PAK2  \n",
       "2         ARF5        RAB36  "
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Shape: {protein_links_v4.shape}\")\n",
    "protein_links_v4.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>INSIG1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>FOXO3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CDH3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  Symbol\n",
       "0      0  INSIG1\n",
       "1      1   FOXO3\n",
       "2      2    CDH3"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UNIQ_LANDMARK_GENES.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtered down such that both columns ´gene_symbol1´ and ´gene_symbol2´ are landmark genes the shape is 102632 (0.86 %). \n",
      "This means 11835866 have either one or both gene symbols not in the set of 978 landmark genes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "protein_links_v5_sparsed = protein_links_v4.loc[(protein_links_v4.gene_symbol1.isin(UNIQ_LANDMARK_GENES.Symbol)) & \n",
    "                                                (protein_links_v4.gene_symbol2.isin(UNIQ_LANDMARK_GENES.Symbol))]\n",
    "n_rows = protein_links_v5_sparsed.shape[0]\n",
    "perc_kept = 100*n_rows/protein_links_v4.shape[0]\n",
    "print(f\"\"\"After filtered down such that both columns ´gene_symbol1´ and ´gene_symbol2´ are landmark genes the shape is {n_rows} ({perc_kept:2.2f} %). \n",
    "This means {protein_links_v4.shape[0]-protein_links_v5_sparsed.shape[0]} have either one or both gene symbols not in the set of {UNIQ_LANDMARK_GENES.shape[0]} landmark genes.\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Add Node Index To DataFrame\n",
    "\n",
    "The simplify the building of a gene-gene interaction graph in the further process of the notebook we are going to assign an index for each unique `gene_symbol`. This index will then hold as an identifier for a node (gene) in the graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000408295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>288</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>RFC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000241337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>218</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>GSTM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1946</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000440485</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>176</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>290</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>PSMD9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "1940  ENSP00000001008  ENSP00000408295             0       0            0   \n",
       "1941  ENSP00000001008  ENSP00000241337             0       0            0   \n",
       "1946  ENSP00000001008  ENSP00000440485             0       0            0   \n",
       "\n",
       "      coexpression  experimental  database  textmining  combined_score  \\\n",
       "1940            81             0         0         258             288   \n",
       "1941            62           175         0          71             218   \n",
       "1946            91           176         0         129             290   \n",
       "\n",
       "     gene_symbol1 gene_symbol2  \n",
       "1940        FKBP4         RFC5  \n",
       "1941        FKBP4        GSTM2  \n",
       "1946        FKBP4        PSMD9  "
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_links_v5_sparsed.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "      <th>index_gene_symbol1</th>\n",
       "      <th>index_gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000408295</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>288</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>RFC5</td>\n",
       "      <td>970</td>\n",
       "      <td>910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000241337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>218</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>GSTM2</td>\n",
       "      <td>970</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000001008</td>\n",
       "      <td>ENSP00000440485</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>176</td>\n",
       "      <td>0</td>\n",
       "      <td>129</td>\n",
       "      <td>290</td>\n",
       "      <td>FKBP4</td>\n",
       "      <td>PSMD9</td>\n",
       "      <td>970</td>\n",
       "      <td>846</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000001008  ENSP00000408295             0       0            0   \n",
       "1  ENSP00000001008  ENSP00000241337             0       0            0   \n",
       "2  ENSP00000001008  ENSP00000440485             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \\\n",
       "0            81             0         0         258             288   \n",
       "1            62           175         0          71             218   \n",
       "2            91           176         0         129             290   \n",
       "\n",
       "  gene_symbol1 gene_symbol2  index_gene_symbol1  index_gene_symbol2  \n",
       "0        FKBP4         RFC5                 970                 910  \n",
       "1        FKBP4        GSTM2                 970                 215  \n",
       "2        FKBP4        PSMD9                 970                 846  "
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Append corresponding index for each gene symbol column as new column(s).\n",
    "protein_links_v6_indexed = protein_links_v5_sparsed.merge(right    = UNIQ_LANDMARK_GENES,\n",
    "                                                          how      = 'left',\n",
    "                                                          left_on  = ['gene_symbol1'],\n",
    "                                                          right_on = ['Symbol'])\n",
    "protein_links_v6_indexed.rename(columns={'index': 'index_gene_symbol1'}, inplace=True)\n",
    "protein_links_v6_indexed.drop(['Symbol'], axis=1, inplace=True)\n",
    "\n",
    "protein_links_v7_indexed = protein_links_v6_indexed.merge(right    = UNIQ_LANDMARK_GENES,\n",
    "                                                          how      = 'left',\n",
    "                                                          left_on  = ['gene_symbol2'],\n",
    "                                                          right_on = ['Symbol'])\n",
    "protein_links_v7_indexed.rename(columns={'index': 'index_gene_symbol2'}, inplace=True)\n",
    "protein_links_v7_indexed.drop(['Symbol'], axis=1, inplace=True)\n",
    "\n",
    "protein_links_v7_indexed.head(3)                                                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Build Graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There will be 965 nodes in the graph.\n",
      "There will be 965 indices in the graph.\n"
     ]
    }
   ],
   "source": [
    "# Set of unique gene symbols.\n",
    "nodes = np.unique(protein_links_v7_indexed[['gene_symbol1', 'gene_symbol2']].values)\n",
    "print(f\"There will be {len(nodes)} nodes in the graph.\")\n",
    "\n",
    "nodes_as_indices = np.unique(protein_links_v7_indexed[['index_gene_symbol1', 'index_gene_symbol2']].values)\n",
    "print(f\"There will be {len(nodes_as_indices)} indices in the graph.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('SMAD3', 'CCNB2'), ('TP53', 'PTPN6'), ('CDK4', 'KDELR2'), ('HSPA4', 'PYCR1'), ('POLR1C', 'RAE1')]\n",
      "[[748, 311], [460, 856], [8, 37], [902, 872], [393, 462]]\n"
     ]
    }
   ],
   "source": [
    "# Create a list of tuples which hold neighbor nodes.\n",
    "gene_symbol_tuples = list(set(list(zip(protein_links_v7_indexed.gene_symbol1, protein_links_v7_indexed.gene_symbol2))))\n",
    "print(gene_symbol_tuples[:5])\n",
    "\n",
    "\n",
    "# Map the list of tuples of neighbor gene nodes to the corresponding int. \n",
    "transform_gene_tuple_to_index_tuple = lambda x : [UNIQ_LANDMARK_GENES[UNIQ_LANDMARK_GENES.Symbol==x[0]]['index'].values[0], \n",
    "                                                  UNIQ_LANDMARK_GENES[UNIQ_LANDMARK_GENES.Symbol==x[1]]['index'].values[0]]\n",
    "input_vec = [transform_gene_tuple_to_index_tuple(tup) for tup in gene_symbol_tuples]\n",
    "print(input_vec[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find neighbors using the gene symbols.\n",
    "dict_as_genes = {gene_node: [] for gene_node in nodes}  # or dict(zip(nodes, []*len(nodes)))\n",
    "\n",
    "# Get the edges.\n",
    "for gene in dict_as_genes:\n",
    "    exists_in_gene_symbol1 = not protein_links_v7_indexed.loc[protein_links_v7_indexed.gene_symbol1==gene].empty\n",
    "    exists_in_gene_symbol2 = not protein_links_v7_indexed.loc[protein_links_v7_indexed.gene_symbol2==gene].empty\n",
    "\n",
    "    neighbor_nodes = None \n",
    "    if exists_in_gene_symbol1:\n",
    "        neighbor_nodes = list(set(protein_links_v7_indexed.loc[protein_links_v7_indexed.gene_symbol1==gene].gene_symbol2))\n",
    "    elif exists_in_gene_symbol2: \n",
    "        neighbor_nodes = list(set(protein_links_v7_indexed.loc[protein_links_v7_indexed.gene_symbol2==gene].gene_symbol1))\n",
    "    else: \n",
    "        print(f\"The gene {gene} couldn't be found in the dataset!\") \n",
    "\n",
    "    # Set neighbors for the gene.\n",
    "    dict_as_genes[gene] += [neighbor_node for neighbor_node in neighbor_nodes if neighbor_node not in dict_as_genes[gene]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "140"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(dict_as_genes['AARS'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find neighbors using the indices.\n",
    "dict_as_indices = {gene_node: [] for gene_node in nodes_as_indices}  # or dict(zip(nodes, []*len(nodes)))\n",
    "\n",
    "# Get the edges.\n",
    "for gene in dict_as_indices:\n",
    "    exists_in_gene_symbol1 = not protein_links_v7_indexed.loc[protein_links_v7_indexed.index_gene_symbol1==gene].empty\n",
    "    exists_in_gene_symbol2 = not protein_links_v7_indexed.loc[protein_links_v7_indexed.index_gene_symbol2==gene].empty\n",
    "\n",
    "    neighbor_nodes = None \n",
    "    if exists_in_gene_symbol1:\n",
    "        neighbor_nodes = list(set(protein_links_v7_indexed.loc[protein_links_v7_indexed.index_gene_symbol1==gene].index_gene_symbol2))\n",
    "    elif exists_in_gene_symbol2: \n",
    "        neighbor_nodes = list(set(protein_links_v7_indexed.loc[protein_links_v7_indexed.index_gene_symbol2==gene].index_gene_symbol1))\n",
    "    else: \n",
    "        print(f\"The gene {gene} couldn't be found in the dataset!\") \n",
    "\n",
    "    # Set neighbors for the gene.\n",
    "    dict_as_indices[gene] += [neighbor_node for neighbor_node in neighbor_nodes if neighbor_node not in dict_as_indices[gene]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On average there are   106 neighbors per node.\n"
     ]
    }
   ],
   "source": [
    "# How many neighbors are there on average?\n",
    "len_of_neighbors = [len(neighbors) for _, neighbors in dict_as_indices.items()]\n",
    "assert len(len_of_neighbors) == len(nodes)\n",
    "print(f\"On average there are {np.mean(len_of_neighbors):5.0f} neighbors per node.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[748, 311], [460, 856], [8, 37], [902, 872], [393, 462]]"
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map the list of tuples of neighbor gene nodes to the corresponding int.\n",
    "input_vec[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[748, 311, 460, 856, 8]"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "# Flatten the list\n",
    "merged = list(itertools.chain(*input_vec))\n",
    "merged[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   count\n",
       "0    156\n",
       "1    576\n",
       "2    196\n",
       "3    290\n",
       "4    136"
      ]
     },
     "execution_count": 329,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = pd.Series(merged).value_counts().to_frame(name='count').sort_index()\n",
    "counts[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[965, 1], edge_index=[2, 102632])"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------ #\n",
    "# Create a torch_geometric Data class using the mapped numbers.\n",
    "# ------------------------------------------------------------ #\n",
    "import torch \n",
    "from torch_geometric.data import Data\n",
    "\n",
    "edge_index = torch.tensor(input_vec, dtype=torch.long)  # All neighbor tuples.\n",
    "x = torch.tensor(counts.values, dtype=torch.float)      # Per index (node/gene) the number of neighbors. \n",
    "\n",
    "data = Data(x          = x, \n",
    "            edge_index = edge_index.t().contiguous())\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `x`: $965$ distinct nodes with each $1$ node feature, which is the number of neighbors.\n",
    "- `edge_index`: $102,632$ links with each link having a start- and an end-node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    data.num_nodes          : 965\n",
      "    data.num_edges          : 102632\n",
      "    data.num_node_features  : 1\n",
      "    data.num_edge_features  : 0\n",
      "    data.has_isolated_nodes : False\n",
      "    data.has_self_loops     : False\n",
      "    data.is_directed        : True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Checking some utility functions.\n",
    "print(f\"\"\"\n",
    "    data.num_nodes          : {data.num_nodes}\n",
    "    data.num_edges          : {data.num_edges}\n",
    "    data.num_node_features  : {data.num_node_features}\n",
    "    data.num_edge_features  : {data.num_edge_features}\n",
    "    data.has_isolated_nodes : {data.has_isolated_nodes()}\n",
    "    data.has_self_loops     : {data.has_self_loops()}\n",
    "    data.is_directed        : {data.is_directed()}\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Model Each Cell-Line As A Graph\n",
    "### Add Nodes-Level Features\n",
    "Each `CELL_LINE_NAME` has one specific node feature value. For example the gene expression level, the copy number or mutational information. Thus, \n",
    "- we can build a graph per cell-line and \n",
    "- add the node feature to each gene node.\n",
    "\n",
    "Note that each graph will have the exact same structure. Meaning, each graph will have the same nodes, only the node values differ. In total there will be $N_{cl}$ graph where $N_{cl}$ stands for the number of distinct cell-lines.\n",
    "\n",
    "The objective is to have a feature vector per node/gene of the form $g_n = [v_{mRnaExpr}, v_{CNV}, v_{isMutated}]$, where \n",
    "- $v_{mRnaExpr}$ holds the value of gene expression\n",
    "- $v_{CNV}$ holds the value of copy numbers\n",
    "- $v_{isMutated}$ holds the value if it mutated or not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [],
   "source": [
    "from my_utils import (\n",
    "    PATH_TO_SAVE_NODE_FEATURE_DATA_TO,\n",
    "    GENE_EXPR_FINAL_FILE_NAME,\n",
    "    CNV_GISTIC_FINAL_FILE_NAME,\n",
    "    CNV_PICNIC_FINAL_FILE_NAME\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446521, 922)\n",
      "Number of unique cell-line & thus number of graph which will be created: 988.\n",
      "Number of gene columns: 908\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>CONC</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCBP</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>IFRD2</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>CEBPD</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>MIF</th>\n",
       "      <th>PRAF2</th>\n",
       "      <th>LYN</th>\n",
       "      <th>POLG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>3137</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>49</td>\n",
       "      <td>544404</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>...</td>\n",
       "      <td>8.355826</td>\n",
       "      <td>8.951680</td>\n",
       "      <td>7.205590</td>\n",
       "      <td>3.277948</td>\n",
       "      <td>3.465672</td>\n",
       "      <td>6.312806</td>\n",
       "      <td>12.112498</td>\n",
       "      <td>3.010237</td>\n",
       "      <td>8.750848</td>\n",
       "      <td>6.199366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>2366</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>404197</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>...</td>\n",
       "      <td>5.995760</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.716270</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>2368</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>2.0</td>\n",
       "      <td>610</td>\n",
       "      <td>797378</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>...</td>\n",
       "      <td>6.939741</td>\n",
       "      <td>8.688176</td>\n",
       "      <td>7.085349</td>\n",
       "      <td>3.688222</td>\n",
       "      <td>4.572119</td>\n",
       "      <td>6.345090</td>\n",
       "      <td>12.276166</td>\n",
       "      <td>4.114092</td>\n",
       "      <td>5.768098</td>\n",
       "      <td>7.505155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    DRUG_ID CELL_LINE_NAME       AUC  CELL_ID   LN_IC50  CONC  MASTER_CELL_ID  \\\n",
       "0         1         MC-CAR  0.982114     3137  2.395685   2.0              49   \n",
       "9         1            ES3  0.984816     2366  3.140923   2.0            1342   \n",
       "27        1            ES5  0.985693     2368  3.968757   2.0             610   \n",
       "\n",
       "    INTENSITY DATASET   Z_SCORE  ...     MYCBP      FIS1     IFRD2    NPEPL1  \\\n",
       "0      544404   GDSC1 -0.189576  ...  8.355826  8.951680  7.205590  3.277948   \n",
       "9      404197   GDSC1  0.508635  ...  5.995760  9.337588  7.468226  3.716270   \n",
       "27     797378   GDSC1  1.284229  ...  6.939741  8.688176  7.085349  3.688222   \n",
       "\n",
       "       CEBPD   PLEKHM1        MIF     PRAF2       LYN     POLG2  \n",
       "0   3.465672  6.312806  12.112498  3.010237  8.750848  6.199366  \n",
       "9   5.363887  6.188079  12.281947  4.794624  3.588528  6.785201  \n",
       "27  4.572119  6.345090  12.276166  4.114092  5.768098  7.505155  \n",
       "\n",
       "[3 rows x 922 columns]"
      ]
     },
     "execution_count": 355,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------- #\n",
    "# GENE EXPRESSION\n",
    "# Read the final gene expression dataset from \n",
    "# notebook 02_GDSC_map_GenExpr.ipynb.\n",
    "# ------------------------------------------- #\n",
    "gene_expr = pd.read_pickle(f'{PATH_TO_SAVE_NODE_FEATURE_DATA_TO}{GENE_EXPR_FINAL_FILE_NAME}')\n",
    "print(gene_expr.shape)\n",
    "print(f\"Number of unique cell-line & thus number of graph which will be created: {len(np.unique(gene_expr.CELL_LINE_NAME.values))}.\")\n",
    "print(f\"Number of gene columns: {len(gene_expr.columns[14:])}\")\n",
    "gene_expr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CNV Gistic: (446521, 952)\n",
      "CNV Picnic: (446521, 980)\n"
     ]
    }
   ],
   "source": [
    "# ---------------------------------------- #\n",
    "# COPY NUMBER VARIATION\n",
    "# Read the final cnv dataset from notebook\n",
    "# 03_GDSC_map_CNV.ipynb.\n",
    "# ---------------------------------------- #\n",
    "# Copy number variation gistic final dataset.\n",
    "final_cnv_gistic = pd.read_pickle(f'{PATH_TO_SAVE_NODE_FEATURE_DATA_TO}{CNV_GISTIC_FINAL_FILE_NAME}')\n",
    "print(f\"CNV Gistic: {final_cnv_gistic.shape}\")\n",
    "\n",
    "# Copy number variation picnic final dataset.\n",
    "final_cnv_picnic = pd.read_pickle(f'{PATH_TO_SAVE_NODE_FEATURE_DATA_TO}{CNV_PICNIC_FINAL_FILE_NAME}')\n",
    "print(f\"CNV Picnic: {final_cnv_picnic.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>AUC</th>\n",
       "      <th>CELL_ID</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>CONC</th>\n",
       "      <th>MASTER_CELL_ID</th>\n",
       "      <th>INTENSITY</th>\n",
       "      <th>DATASET</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>...</th>\n",
       "      <th>MYCBP</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>IFRD2</th>\n",
       "      <th>NPEPL1</th>\n",
       "      <th>CEBPD</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>MIF</th>\n",
       "      <th>PRAF2</th>\n",
       "      <th>LYN</th>\n",
       "      <th>POLG2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>2366</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>404197</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4572</th>\n",
       "      <td>3</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.981415</td>\n",
       "      <td>2366</td>\n",
       "      <td>-0.067752</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1342</td>\n",
       "      <td>561199</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>1.120134</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8622</th>\n",
       "      <td>5</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.941416</td>\n",
       "      <td>2366</td>\n",
       "      <td>3.043634</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>277899</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.330702</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12852</th>\n",
       "      <td>6</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.981549</td>\n",
       "      <td>2366</td>\n",
       "      <td>2.935678</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>524262</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.223649</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17487</th>\n",
       "      <td>9</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.956082</td>\n",
       "      <td>2366</td>\n",
       "      <td>1.296998</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>398551</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.947409</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3386931</th>\n",
       "      <td>1502</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.974498</td>\n",
       "      <td>2366</td>\n",
       "      <td>4.199460</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>2146522</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.369485</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3392246</th>\n",
       "      <td>1526</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.963667</td>\n",
       "      <td>2366</td>\n",
       "      <td>3.119137</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>1827393</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.871292</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3397456</th>\n",
       "      <td>1527</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.816016</td>\n",
       "      <td>2366</td>\n",
       "      <td>0.807811</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>950337</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.014279</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3402698</th>\n",
       "      <td>1529</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.818002</td>\n",
       "      <td>2366</td>\n",
       "      <td>-0.561027</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>274514</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>-0.483495</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3412061</th>\n",
       "      <td>1530</td>\n",
       "      <td>ES3</td>\n",
       "      <td>0.983505</td>\n",
       "      <td>2366</td>\n",
       "      <td>4.663353</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1342</td>\n",
       "      <td>2197018</td>\n",
       "      <td>GDSC1</td>\n",
       "      <td>0.064779</td>\n",
       "      <td>...</td>\n",
       "      <td>5.99576</td>\n",
       "      <td>9.337588</td>\n",
       "      <td>7.468226</td>\n",
       "      <td>3.71627</td>\n",
       "      <td>5.363887</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>12.281947</td>\n",
       "      <td>4.794624</td>\n",
       "      <td>3.588528</td>\n",
       "      <td>6.785201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         DRUG_ID CELL_LINE_NAME       AUC  CELL_ID   LN_IC50  CONC  \\\n",
       "9              1            ES3  0.984816     2366  3.140923   2.0   \n",
       "4572           3            ES3  0.981415     2366 -0.067752   0.1   \n",
       "8622           5            ES3  0.941416     2366  3.043634   8.0   \n",
       "12852          6            ES3  0.981549     2366  2.935678   2.0   \n",
       "17487          9            ES3  0.956082     2366  1.296998   1.0   \n",
       "...          ...            ...       ...      ...       ...   ...   \n",
       "3386931     1502            ES3  0.974498     2366  4.199460  10.0   \n",
       "3392246     1526            ES3  0.963667     2366  3.119137   5.0   \n",
       "3397456     1527            ES3  0.816016     2366  0.807811   4.0   \n",
       "3402698     1529            ES3  0.818002     2366 -0.561027   1.0   \n",
       "3412061     1530            ES3  0.983505     2366  4.663353  10.0   \n",
       "\n",
       "         MASTER_CELL_ID  INTENSITY DATASET   Z_SCORE  ...    MYCBP      FIS1  \\\n",
       "9                  1342     404197   GDSC1  0.508635  ...  5.99576  9.337588   \n",
       "4572               1342     561199   GDSC1  1.120134  ...  5.99576  9.337588   \n",
       "8622               1342     277899   GDSC1  0.330702  ...  5.99576  9.337588   \n",
       "12852              1342     524262   GDSC1  0.223649  ...  5.99576  9.337588   \n",
       "17487              1342     398551   GDSC1  0.947409  ...  5.99576  9.337588   \n",
       "...                 ...        ...     ...       ...  ...      ...       ...   \n",
       "3386931            1342    2146522   GDSC1 -0.369485  ...  5.99576  9.337588   \n",
       "3392246            1342    1827393   GDSC1  0.871292  ...  5.99576  9.337588   \n",
       "3397456            1342     950337   GDSC1 -0.014279  ...  5.99576  9.337588   \n",
       "3402698            1342     274514   GDSC1 -0.483495  ...  5.99576  9.337588   \n",
       "3412061            1342    2197018   GDSC1  0.064779  ...  5.99576  9.337588   \n",
       "\n",
       "            IFRD2   NPEPL1     CEBPD   PLEKHM1        MIF     PRAF2       LYN  \\\n",
       "9        7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "4572     7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "8622     7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "12852    7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "17487    7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "...           ...      ...       ...       ...        ...       ...       ...   \n",
       "3386931  7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "3392246  7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "3397456  7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "3402698  7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "3412061  7.468226  3.71627  5.363887  6.188079  12.281947  4.794624  3.588528   \n",
       "\n",
       "            POLG2  \n",
       "9        6.785201  \n",
       "4572     6.785201  \n",
       "8622     6.785201  \n",
       "12852    6.785201  \n",
       "17487    6.785201  \n",
       "...           ...  \n",
       "3386931  6.785201  \n",
       "3392246  6.785201  \n",
       "3397456  6.785201  \n",
       "3402698  6.785201  \n",
       "3412061  6.785201  \n",
       "\n",
       "[366 rows x 922 columns]"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expr.loc[gene_expr.CELL_LINE_NAME=='ES3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct graph for each cell-line. Set corresponding gene-expression values per node. \n",
    "# Graph predicts IC50 value. However, IC50 value obviously different for each DRUG_ID. \n",
    "#  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_GDSC_SCREENING_DATA = '../../datasets/gdsc/screening_data/'\n",
    "GDSC1_IC50_FILE = 'GDSC1_fitted_dose_response_25Feb20.xlsx'\n",
    "gdsc1_ic50s = pd.read_excel(f'{PATH_TO_GDSC_SCREENING_DATA}{GDSC1_IC50_FILE}', header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975300</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975551</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>3</td>\n",
       "      <td>Rapamycin</td>\n",
       "      <td>MTORC1</td>\n",
       "      <td>PI3K/MTOR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.1</td>\n",
       "      <td>-0.067752</td>\n",
       "      <td>0.981415</td>\n",
       "      <td>0.074930</td>\n",
       "      <td>1.120134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975618</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>5</td>\n",
       "      <td>Sunitinib</td>\n",
       "      <td>PDGFR, KIT, VEGFR, FLT3, RET, CSF1R</td>\n",
       "      <td>RTK signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.043634</td>\n",
       "      <td>0.941416</td>\n",
       "      <td>0.023747</td>\n",
       "      <td>0.330702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975633</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>6</td>\n",
       "      <td>PHA-665752</td>\n",
       "      <td>MET</td>\n",
       "      <td>RTK signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.935678</td>\n",
       "      <td>0.981549</td>\n",
       "      <td>0.057955</td>\n",
       "      <td>0.223649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1558</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975644</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>9</td>\n",
       "      <td>MG-132</td>\n",
       "      <td>Proteasome, CAPN1</td>\n",
       "      <td>Protein stability and degradation</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.003906</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.296998</td>\n",
       "      <td>0.956082</td>\n",
       "      <td>0.020726</td>\n",
       "      <td>0.947409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306729</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975447</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1502</td>\n",
       "      <td>Bicalutamide</td>\n",
       "      <td>AR</td>\n",
       "      <td>Hormone-related</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.039063</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.199460</td>\n",
       "      <td>0.974498</td>\n",
       "      <td>0.056111</td>\n",
       "      <td>-0.369485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307604</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975450</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1526</td>\n",
       "      <td>Refametinib</td>\n",
       "      <td>MEK1, MEK2</td>\n",
       "      <td>ERK MAPK signaling</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.019531</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.119137</td>\n",
       "      <td>0.963667</td>\n",
       "      <td>0.107487</td>\n",
       "      <td>0.871292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308478</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975451</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1527</td>\n",
       "      <td>Pictilisib</td>\n",
       "      <td>PI3K (class 1)</td>\n",
       "      <td>PI3K/MTOR signaling</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.807811</td>\n",
       "      <td>0.816016</td>\n",
       "      <td>0.069947</td>\n",
       "      <td>-0.014279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309358</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975452</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1529</td>\n",
       "      <td>Pevonedistat</td>\n",
       "      <td>NAE</td>\n",
       "      <td>Other</td>\n",
       "      <td>1046</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.003906</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.561027</td>\n",
       "      <td>0.818002</td>\n",
       "      <td>0.159992</td>\n",
       "      <td>-0.483495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310059</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975454</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1530</td>\n",
       "      <td>PFI-3</td>\n",
       "      <td>SMARCA2, SMARCA4, PB1</td>\n",
       "      <td>Chromatin other</td>\n",
       "      <td>1025</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.039063</td>\n",
       "      <td>10.0</td>\n",
       "      <td>4.663353</td>\n",
       "      <td>0.983505</td>\n",
       "      <td>0.040524</td>\n",
       "      <td>0.064779</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "1        GDSC1             281       12975300     684055            ES3   \n",
       "395      GDSC1             281       12975551     684055            ES3   \n",
       "753      GDSC1             281       12975618     684055            ES3   \n",
       "1151     GDSC1             281       12975633     684055            ES3   \n",
       "1558     GDSC1             281       12975644     684055            ES3   \n",
       "...        ...             ...            ...        ...            ...   \n",
       "306729   GDSC1             281       12975447     684055            ES3   \n",
       "307604   GDSC1             281       12975450     684055            ES3   \n",
       "308478   GDSC1             281       12975451     684055            ES3   \n",
       "309358   GDSC1             281       12975452     684055            ES3   \n",
       "310059   GDSC1             281       12975454     684055            ES3   \n",
       "\n",
       "       SANGER_MODEL_ID     TCGA_DESC  DRUG_ID     DRUG_NAME  \\\n",
       "1            SIDM00265  UNCLASSIFIED        1     Erlotinib   \n",
       "395          SIDM00265  UNCLASSIFIED        3     Rapamycin   \n",
       "753          SIDM00265  UNCLASSIFIED        5     Sunitinib   \n",
       "1151         SIDM00265  UNCLASSIFIED        6    PHA-665752   \n",
       "1558         SIDM00265  UNCLASSIFIED        9        MG-132   \n",
       "...                ...           ...      ...           ...   \n",
       "306729       SIDM00265  UNCLASSIFIED     1502  Bicalutamide   \n",
       "307604       SIDM00265  UNCLASSIFIED     1526   Refametinib   \n",
       "308478       SIDM00265  UNCLASSIFIED     1527    Pictilisib   \n",
       "309358       SIDM00265  UNCLASSIFIED     1529  Pevonedistat   \n",
       "310059       SIDM00265  UNCLASSIFIED     1530         PFI-3   \n",
       "\n",
       "                            PUTATIVE_TARGET  \\\n",
       "1                                      EGFR   \n",
       "395                                  MTORC1   \n",
       "753     PDGFR, KIT, VEGFR, FLT3, RET, CSF1R   \n",
       "1151                                    MET   \n",
       "1558                      Proteasome, CAPN1   \n",
       "...                                     ...   \n",
       "306729                                   AR   \n",
       "307604                           MEK1, MEK2   \n",
       "308478                       PI3K (class 1)   \n",
       "309358                                  NAE   \n",
       "310059                SMARCA2, SMARCA4, PB1   \n",
       "\n",
       "                             PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  \\\n",
       "1                          EGFR signaling        1045          Y  0.007813   \n",
       "395                   PI3K/MTOR signaling        1045          Y  0.000391   \n",
       "753                         RTK signaling        1045          Y  0.031250   \n",
       "1151                        RTK signaling        1045          Y  0.007813   \n",
       "1558    Protein stability and degradation        1045          Y  0.003906   \n",
       "...                                   ...         ...        ...       ...   \n",
       "306729                    Hormone-related        1046          Y  0.039063   \n",
       "307604                 ERK MAPK signaling        1046          Y  0.019531   \n",
       "308478                PI3K/MTOR signaling        1046          Y  0.015625   \n",
       "309358                              Other        1046          Y  0.003906   \n",
       "310059                    Chromatin other        1025          Y  0.039063   \n",
       "\n",
       "        MAX_CONC   LN_IC50       AUC      RMSE   Z_SCORE  \n",
       "1            2.0  3.140923  0.984816  0.031840  0.508635  \n",
       "395          0.1 -0.067752  0.981415  0.074930  1.120134  \n",
       "753          8.0  3.043634  0.941416  0.023747  0.330702  \n",
       "1151         2.0  2.935678  0.981549  0.057955  0.223649  \n",
       "1558         1.0  1.296998  0.956082  0.020726  0.947409  \n",
       "...          ...       ...       ...       ...       ...  \n",
       "306729      10.0  4.199460  0.974498  0.056111 -0.369485  \n",
       "307604       5.0  3.119137  0.963667  0.107487  0.871292  \n",
       "308478       4.0  0.807811  0.816016  0.069947 -0.014279  \n",
       "309358       1.0 -0.561027  0.818002  0.159992 -0.483495  \n",
       "310059      10.0  4.663353  0.983505  0.040524  0.064779  \n",
       "\n",
       "[366 rows x 19 columns]"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc1_ic50s.loc[gdsc1_ic50s.CELL_LINE_NAME=='ES3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[748, 311], [460, 856], [8, 37], [902, 872], [393, 462]]"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_vec = [transform_gene_tuple_to_index_tuple(tup) for tup in gene_symbol_tuples[:5]]\n",
    "input_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[4, 1], [2, 7], [15, 9], [12, 13], [14, 8], [5, 1], [11, 0], [5, 6], [3, 10]]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Map the list of tuples of neighbor gene nodes to the corresponding int. \n",
    "transform_gene_tuple_to_index_tuple = lambda x : [mapping_gene_to_number[x[0]], mapping_gene_to_number[x[1]]]\n",
    "input_vec = [transform_gene_tuple_to_index_tuple(tup) for tup in t]\n",
    "input_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_index_to_gene = dict(zip(nodes_as_numbers, nodes))\n",
    "mapping_gene_to_index = dict(zip(nodes, nodes_as_numbers))\n",
    "mapping_number_to_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Build Graph Only By Connection Or Not\n",
    "\n",
    "- For now, build a graph only using the information if there is an existing tuple in the dataset or not.\n",
    "- The graph will be undirected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSP00000394168',\n",
       " 'ENSP00000357308',\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 303,\n",
       " 303,\n",
       " 'SLCO1B7',\n",
       " 'CENPW']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.iloc[0].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000394168</td>\n",
       "      <td>ENSP00000357308</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>303</td>\n",
       "      <td>303</td>\n",
       "      <td>SLCO1B7</td>\n",
       "      <td>CENPW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000351108</td>\n",
       "      <td>ENSP00000432660</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>178</td>\n",
       "      <td>HNRNPAB</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000361803</td>\n",
       "      <td>ENSP00000361852</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>565</td>\n",
       "      <td>565</td>\n",
       "      <td>SMAP2</td>\n",
       "      <td>TMCO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000365435</td>\n",
       "      <td>ENSP00000308938</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151</td>\n",
       "      <td>161</td>\n",
       "      <td>TNFRSF1B</td>\n",
       "      <td>PLG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000396402</td>\n",
       "      <td>ENSP00000341243</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>434</td>\n",
       "      <td>471</td>\n",
       "      <td>ZNF793</td>\n",
       "      <td>RBPJL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSP00000323901</td>\n",
       "      <td>ENSP00000328596</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>253</td>\n",
       "      <td>316</td>\n",
       "      <td>FZD2</td>\n",
       "      <td>SIX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSP00000222982</td>\n",
       "      <td>ENSP00000312286</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>311</td>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>PLA2G1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000394168  ENSP00000357308             0       0            0   \n",
       "1  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "2  ENSP00000351108  ENSP00000432660             0       0            0   \n",
       "3  ENSP00000361803  ENSP00000361852             0       0            0   \n",
       "4  ENSP00000365435  ENSP00000308938             0       0            0   \n",
       "5  ENSP00000396402  ENSP00000341243             0       0            0   \n",
       "6  ENSP00000323901  ENSP00000328596             0       0            0   \n",
       "7  ENSP00000222982  ENSP00000312286             0       0            0   \n",
       "8  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "9  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \\\n",
       "0             0             0         0         303             303   \n",
       "1             0             0         0         284             284   \n",
       "2           121             0         0         104             178   \n",
       "3             0             0         0         565             565   \n",
       "4            53             0         0         151             161   \n",
       "5             0           104         0         434             471   \n",
       "6           122             0         0         253             316   \n",
       "7            63             0         0         295             311   \n",
       "8             0             0         0         284             284   \n",
       "9             0             0         0         284             284   \n",
       "\n",
       "  gene_symbol1 gene_symbol2  \n",
       "0      SLCO1B7        CENPW  \n",
       "1        KCNG2       KLHL28  \n",
       "2      HNRNPAB         CFL1  \n",
       "3        SMAP2        TMCO2  \n",
       "4     TNFRSF1B          PLG  \n",
       "5       ZNF793        RBPJL  \n",
       "6         FZD2         SIX6  \n",
       "7       CYP3A5      PLA2G1B  \n",
       "8        KCNG2       KLHL28  \n",
       "9        KCNG2         CFL1  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEED = 42\n",
    "SAMPLE_SIZE = 8\n",
    "\n",
    "# Random sample some rows for testing purposes, to get some distinct gene-gene tuples.\n",
    "test = protein_links_v4.sample(SAMPLE_SIZE, random_state=SEED)\n",
    "\n",
    "# For testing purposes duplicate a row.\n",
    "sample = test.sample(1, random_state=SEED)\n",
    "test = pd.concat([test, sample], ignore_index=True)\n",
    "test = pd.concat([test, sample], ignore_index=True)\n",
    "\n",
    "test.loc[test.shape[0]-3, 'gene_symbol2'] = 'KLHL28'  # Case where the tuple (gene1, gene2) already exists.\n",
    "test.loc[test.shape[0]-2, 'gene_symbol2'] = 'CFL1'\n",
    "test.loc[test.shape[0]-1, 'gene_symbol2'] = 'KCNG2'   # Case where gene1 and gene2 are the same.\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = np.unique(test[['gene_symbol1', 'gene_symbol2']].values)\n",
    "len(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'CENPW',\n",
       " 1: 'CFL1',\n",
       " 2: 'CYP3A5',\n",
       " 3: 'FZD2',\n",
       " 4: 'HNRNPAB',\n",
       " 5: 'KCNG2',\n",
       " 6: 'KLHL28',\n",
       " 7: 'PLA2G1B',\n",
       " 8: 'PLG',\n",
       " 9: 'RBPJL',\n",
       " 10: 'SIX6',\n",
       " 11: 'SLCO1B7',\n",
       " 12: 'SMAP2',\n",
       " 13: 'TMCO2',\n",
       " 14: 'TNFRSF1B',\n",
       " 15: 'ZNF793'}"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_as_numbers = np.arange(0, len(nodes), 1)\n",
    "mapping_number_to_gene = dict(zip(nodes_as_numbers, nodes))\n",
    "mapping_gene_to_number = dict(zip(nodes, nodes_as_numbers))\n",
    "mapping_number_to_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_as_genes = {gene_node: [] for gene_node in nodes}  # or dict(zip(nodes, []*len(nodes)))\n",
    "len(dict_as_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking gene : CENPW\n",
      "Checking gene : CFL1\n",
      "Checking gene : CYP3A5\n",
      "Checking gene : FZD2\n",
      "Checking gene : HNRNPAB\n",
      "Checking gene : KCNG2\n",
      "Checking gene : KLHL28\n",
      "Checking gene : PLA2G1B\n",
      "Checking gene : PLG\n",
      "Checking gene : RBPJL\n",
      "Checking gene : SIX6\n",
      "Checking gene : SLCO1B7\n",
      "Checking gene : SMAP2\n",
      "Checking gene : TMCO2\n",
      "Checking gene : TNFRSF1B\n",
      "Checking gene : ZNF793\n"
     ]
    }
   ],
   "source": [
    "dict_as_genes = {gene_node: [] for gene_node in nodes}  # or dict(zip(nodes, []*len(nodes)))\n",
    "\n",
    "# Get the edges.\n",
    "for gene in dict_as_genes:\n",
    "    print(f\"Checking gene : {gene}\")\n",
    "    exists_in_gene_symbol1 = not test.loc[test.gene_symbol1==gene].empty\n",
    "    exists_in_gene_symbol2 = not test.loc[test.gene_symbol2==gene].empty\n",
    "\n",
    "    neighbor_nodes = None \n",
    "    if exists_in_gene_symbol1:\n",
    "        neighbor_nodes = list(set(test.loc[test.gene_symbol1==gene].gene_symbol2))\n",
    "    elif exists_in_gene_symbol2: \n",
    "        neighbor_nodes = list(set(test.loc[test.gene_symbol2==gene].gene_symbol1))\n",
    "    else: \n",
    "        print(f\"The gene {gene} couldmn't be found in the dataset!\") \n",
    "\n",
    "    # Set neighbors for the gene.\n",
    "    #print(f\"    appending neighbors {neighbor_nodes}\")\n",
    "    dict_as_genes[gene] += [neighbor_node for neighbor_node in neighbor_nodes if neighbor_node not in dict_as_genes[gene]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CENPW': ['SLCO1B7'],\n",
       " 'CFL1': ['KCNG2', 'HNRNPAB'],\n",
       " 'CYP3A5': ['PLA2G1B'],\n",
       " 'FZD2': ['SIX6'],\n",
       " 'HNRNPAB': ['CFL1'],\n",
       " 'KCNG2': ['CFL1', 'KLHL28'],\n",
       " 'KLHL28': ['KCNG2'],\n",
       " 'PLA2G1B': ['CYP3A5'],\n",
       " 'PLG': ['TNFRSF1B'],\n",
       " 'RBPJL': ['ZNF793'],\n",
       " 'SIX6': ['FZD2'],\n",
       " 'SLCO1B7': ['CENPW'],\n",
       " 'SMAP2': ['TMCO2'],\n",
       " 'TMCO2': ['SMAP2'],\n",
       " 'TNFRSF1B': ['PLG'],\n",
       " 'ZNF793': ['RBPJL']}"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict_as_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: [11],\n",
       " 1: [5, 4],\n",
       " 2: [7],\n",
       " 3: [10],\n",
       " 4: [1],\n",
       " 5: [1, 6],\n",
       " 6: [5],\n",
       " 7: [2],\n",
       " 8: [14],\n",
       " 9: [15],\n",
       " 10: [3],\n",
       " 11: [0],\n",
       " 12: [13],\n",
       " 13: [12],\n",
       " 14: [8],\n",
       " 15: [9]}"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------- #\n",
    "# Map the genes to numbers.\n",
    "# ------------------------- #\n",
    "\n",
    "# Return dictionary as the mapped numbers.\n",
    "dict_as_numbers = {mapping_gene_to_number[gene]: [mapping_gene_to_number[neighbor] for neighbor in neighbors] for gene, neighbors in dict_as_genes.items()}\n",
    "dict_as_numbers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Transform to pytorch geometric graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLCO1B7</td>\n",
       "      <td>CENPW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HNRNPAB</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMAP2</td>\n",
       "      <td>TMCO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TNFRSF1B</td>\n",
       "      <td>PLG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ZNF793</td>\n",
       "      <td>RBPJL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FZD2</td>\n",
       "      <td>SIX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>PLA2G1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>KCNG2</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol1 gene_symbol2\n",
       "0      SLCO1B7        CENPW\n",
       "1        KCNG2       KLHL28\n",
       "2      HNRNPAB         CFL1\n",
       "3        SMAP2        TMCO2\n",
       "4     TNFRSF1B          PLG\n",
       "5       ZNF793        RBPJL\n",
       "6         FZD2         SIX6\n",
       "7       CYP3A5      PLA2G1B\n",
       "8        KCNG2       KLHL28\n",
       "9        KCNG2         CFL1"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test[['gene_symbol1', 'gene_symbol2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('HNRNPAB', 'CFL1'),\n",
       " ('CYP3A5', 'PLA2G1B'),\n",
       " ('ZNF793', 'RBPJL'),\n",
       " ('SMAP2', 'TMCO2'),\n",
       " ('TNFRSF1B', 'PLG'),\n",
       " ('KCNG2', 'CFL1'),\n",
       " ('SLCO1B7', 'CENPW'),\n",
       " ('KCNG2', 'KLHL28'),\n",
       " ('FZD2', 'SIX6')]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a list of tuples which hold neighbor nodes.\n",
    "t = list(set(list(zip(test.gene_symbol1, test.gene_symbol2))))\n",
    "t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[4, 1], [2, 7], [15, 9], [12, 13], [14, 8], [5, 1], [11, 0], [5, 6], [3, 10]]"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map the list of tuples of neighbor gene nodes to the corresponding int. \n",
    "transform_gene_tuple_to_index_tuple = lambda x : [mapping_gene_to_number[x[0]], mapping_gene_to_number[x[1]]]\n",
    "input_vec = [transform_gene_tuple_to_index_tuple(tup) for tup in t]\n",
    "input_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4, 1, 2, 7, 15, 9, 12, 13, 14, 8, 5, 1, 11, 0, 5, 6, 3, 10]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import itertools\n",
    "# Flatten the list\n",
    "merged = list(itertools.chain(*input_vec))\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    count\n",
       "0       1\n",
       "1       2\n",
       "2       1\n",
       "3       1\n",
       "4       1\n",
       "5       2\n",
       "6       1\n",
       "7       1\n",
       "8       1\n",
       "9       1\n",
       "10      1\n",
       "11      1\n",
       "12      1\n",
       "13      1\n",
       "14      1\n",
       "15      1"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = pd.Series(merged).value_counts().to_frame(name='count').sort_index()\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1],\n",
       "       [2],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [2],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1],\n",
       "       [1]])"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>gene_symbol1</th>\n",
       "      <th>gene_symbol2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000394168</td>\n",
       "      <td>ENSP00000357308</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>303</td>\n",
       "      <td>303</td>\n",
       "      <td>SLCO1B7</td>\n",
       "      <td>CENPW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000351108</td>\n",
       "      <td>ENSP00000432660</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>178</td>\n",
       "      <td>HNRNPAB</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000361803</td>\n",
       "      <td>ENSP00000361852</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>565</td>\n",
       "      <td>565</td>\n",
       "      <td>SMAP2</td>\n",
       "      <td>TMCO2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000365435</td>\n",
       "      <td>ENSP00000308938</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151</td>\n",
       "      <td>161</td>\n",
       "      <td>TNFRSF1B</td>\n",
       "      <td>PLG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000396402</td>\n",
       "      <td>ENSP00000341243</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>434</td>\n",
       "      <td>471</td>\n",
       "      <td>ZNF793</td>\n",
       "      <td>RBPJL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSP00000323901</td>\n",
       "      <td>ENSP00000328596</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>253</td>\n",
       "      <td>316</td>\n",
       "      <td>FZD2</td>\n",
       "      <td>SIX6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSP00000222982</td>\n",
       "      <td>ENSP00000312286</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>311</td>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>PLA2G1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>KLHL28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSP00000315654</td>\n",
       "      <td>ENSP00000379434</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>284</td>\n",
       "      <td>284</td>\n",
       "      <td>KCNG2</td>\n",
       "      <td>CFL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0  ENSP00000394168  ENSP00000357308             0       0            0   \n",
       "1  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "2  ENSP00000351108  ENSP00000432660             0       0            0   \n",
       "3  ENSP00000361803  ENSP00000361852             0       0            0   \n",
       "4  ENSP00000365435  ENSP00000308938             0       0            0   \n",
       "5  ENSP00000396402  ENSP00000341243             0       0            0   \n",
       "6  ENSP00000323901  ENSP00000328596             0       0            0   \n",
       "7  ENSP00000222982  ENSP00000312286             0       0            0   \n",
       "8  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "9  ENSP00000315654  ENSP00000379434             0       0            0   \n",
       "\n",
       "   coexpression  experimental  database  textmining  combined_score  \\\n",
       "0             0             0         0         303             303   \n",
       "1             0             0         0         284             284   \n",
       "2           121             0         0         104             178   \n",
       "3             0             0         0         565             565   \n",
       "4            53             0         0         151             161   \n",
       "5             0           104         0         434             471   \n",
       "6           122             0         0         253             316   \n",
       "7            63             0         0         295             311   \n",
       "8             0             0         0         284             284   \n",
       "9             0             0         0         284             284   \n",
       "\n",
       "  gene_symbol1 gene_symbol2  \n",
       "0      SLCO1B7        CENPW  \n",
       "1        KCNG2       KLHL28  \n",
       "2      HNRNPAB         CFL1  \n",
       "3        SMAP2        TMCO2  \n",
       "4     TNFRSF1B          PLG  \n",
       "5       ZNF793        RBPJL  \n",
       "6         FZD2         SIX6  \n",
       "7       CYP3A5      PLA2G1B  \n",
       "8        KCNG2       KLHL28  \n",
       "9        KCNG2         CFL1  "
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------ #\n",
    "# Create tensor which holds all tuples. \n",
    "# ------------------------------------ #\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[4, 1], [2, 7], [15, 9], [12, 13], [14, 8], [5, 1], [11, 0], [5, 6], [3, 10]]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[16, 1], edge_index=[2, 9])"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------ #\n",
    "# Create a torch_geometric Data class using the mapped numbers.\n",
    "# ------------------------------------------------------------ #\n",
    "import torch \n",
    "from torch_geometric.data import Data\n",
    "\n",
    "edge_index = torch.tensor(input_vec, dtype=torch.long)\n",
    "x = torch.tensor(counts.values, dtype=torch.float)\n",
    "\n",
    "data = Data(x=x, edge_index=edge_index.t().contiguous())\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Data(x=[3, 1], edge_index=[2, 4])"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Resource: https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html\n",
    "import torch\n",
    "from torch_geometric.data import Data\n",
    "\n",
    "edge_index = torch.tensor([[0, 1, 1, 2],\n",
    "                           [1, 0, 2, 1]], dtype=torch.long)\n",
    "x = torch.tensor([[-1], [0], [1]], dtype=torch.float)\n",
    "\n",
    "data = Data(x=x, edge_index=edge_index)\n",
    "# >>> Data(edge_index=[2, 4], x=[3, 1])\n",
    "\n",
    "data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "too many dimensions 'str'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb Cell 41'\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb#ch0000087?line=1'>2</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtorch_geometric\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdata\u001b[39;00m \u001b[39mimport\u001b[39;00m Data\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb#ch0000087?line=3'>4</a>\u001b[0m edge_index \u001b[39m=\u001b[39m torch\u001b[39m.\u001b[39mtensor([[\u001b[39m0\u001b[39m, \u001b[39m1\u001b[39m, \u001b[39m1\u001b[39m, \u001b[39m2\u001b[39m],\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb#ch0000087?line=4'>5</a>\u001b[0m                            [\u001b[39m1\u001b[39m, \u001b[39m0\u001b[39m, \u001b[39m2\u001b[39m, \u001b[39m1\u001b[39m]], dtype\u001b[39m=\u001b[39mtorch\u001b[39m.\u001b[39mlong)\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb#ch0000087?line=5'>6</a>\u001b[0m x \u001b[39m=\u001b[39m torch\u001b[39m.\u001b[39;49mtensor([[\u001b[39m'\u001b[39;49m\u001b[39mA\u001b[39;49m\u001b[39m'\u001b[39;49m], [\u001b[39m'\u001b[39;49m\u001b[39mB\u001b[39;49m\u001b[39m'\u001b[39;49m], [\u001b[39m'\u001b[39;49m\u001b[39mC\u001b[39;49m\u001b[39m'\u001b[39;49m]], dtype\u001b[39m=\u001b[39;49mtorch\u001b[39m.\u001b[39;49mcomplex64)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/07_Linking.ipynb#ch0000087?line=7'>8</a>\u001b[0m data \u001b[39m=\u001b[39m Data(x\u001b[39m=\u001b[39mx, edge_index\u001b[39m=\u001b[39medge_index)\n",
      "\u001b[0;31mValueError\u001b[0m: too many dimensions 'str'"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from torch_geometric.data import Data\n",
    "\n",
    "edge_index = torch.tensor([[0, 1, 1, 2],\n",
    "                           [1, 0, 2, 1]], dtype=torch.long)\n",
    "x = torch.tensor([['A'], ['B'], ['C']], dtype=torch.complex64)\n",
    "\n",
    "data = Data(x=x, edge_index=edge_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# -------------------------------------------------------- #\n",
    "# Build dictionary holding node (gene_symbol1) as the keys \n",
    "# and connected nodes (gene_symbol2) as the value.\n",
    "# -------------------------------------------------------- #\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>string_protein_id</th>\n",
       "      <th>preferred_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000379496</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>144</td>\n",
       "      <td>155</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000314067</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>197</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000263116</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>62</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>101</td>\n",
       "      <td>222</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000361263</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>181</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ENSP00000409666</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>270</td>\n",
       "      <td>ENSP00000000233</td>\n",
       "      <td>ARF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938493</th>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>ENSP00000354800</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>213</td>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>OR6Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938494</th>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>ENSP00000308270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>151</td>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>OR6Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938495</th>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>ENSP00000335660</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>181</td>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>OR6Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938496</th>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>ENSP00000300127</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>155</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>154</td>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>OR6Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11938497</th>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>ENSP00000310488</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>ENSP00000485678</td>\n",
       "      <td>OR6Q1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11938498 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 protein1         protein2  neighborhood  fusion  cooccurence  \\\n",
       "0         ENSP00000000233  ENSP00000379496             0       0            0   \n",
       "1         ENSP00000000233  ENSP00000314067             0       0            0   \n",
       "2         ENSP00000000233  ENSP00000263116             0       0            0   \n",
       "3         ENSP00000000233  ENSP00000361263             0       0            0   \n",
       "4         ENSP00000000233  ENSP00000409666             0       0            0   \n",
       "...                   ...              ...           ...     ...          ...   \n",
       "11938493  ENSP00000485678  ENSP00000354800             0       0            0   \n",
       "11938494  ENSP00000485678  ENSP00000308270             0       0            0   \n",
       "11938495  ENSP00000485678  ENSP00000335660             0       0            0   \n",
       "11938496  ENSP00000485678  ENSP00000300127             0       0            0   \n",
       "11938497  ENSP00000485678  ENSP00000310488             0       0            0   \n",
       "\n",
       "          coexpression  experimental  database  textmining  combined_score  \\\n",
       "0                   54             0         0         144             155   \n",
       "1                    0           180         0          61             197   \n",
       "2                   62           152         0         101             222   \n",
       "3                    0           161         0          64             181   \n",
       "4                   82           213         0          72             270   \n",
       "...                ...           ...       ...         ...             ...   \n",
       "11938493           213             0         0           0             213   \n",
       "11938494           152             0         0           0             151   \n",
       "11938495           182             0         0           0             181   \n",
       "11938496           155             0         0           0             154   \n",
       "11938497           168             0         0           0             167   \n",
       "\n",
       "         string_protein_id preferred_name  \n",
       "0          ENSP00000000233           ARF5  \n",
       "1          ENSP00000000233           ARF5  \n",
       "2          ENSP00000000233           ARF5  \n",
       "3          ENSP00000000233           ARF5  \n",
       "4          ENSP00000000233           ARF5  \n",
       "...                    ...            ...  \n",
       "11938493   ENSP00000485678          OR6Q1  \n",
       "11938494   ENSP00000485678          OR6Q1  \n",
       "11938495   ENSP00000485678          OR6Q1  \n",
       "11938496   ENSP00000485678          OR6Q1  \n",
       "11938497   ENSP00000485678          OR6Q1  \n",
       "\n",
       "[11938498 rows x 12 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_links_v2.merge(right=protein_info_v2[['string_protein_id', 'preferred_name']],\n",
    "                                          how='left',\n",
    "                                          left_on='protein1',\n",
    "                                          right_on='string_protein_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--- \n",
    "## EDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print('ENSP00000243052' in protein_links_v2.protein1.values) # This would have the gene 'PDE1B'\n",
    "print('ENSP00000379496' in protein_links_v2.protein1.values) # This would have the gene 'PDE1C'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count_nulls</th>\n",
       "      <th>count_zeros</th>\n",
       "      <th>perc_zeros</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>neighborhood</th>\n",
       "      <td>0</td>\n",
       "      <td>11528086</td>\n",
       "      <td>96.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fusion</th>\n",
       "      <td>0</td>\n",
       "      <td>11915600</td>\n",
       "      <td>99.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cooccurence</th>\n",
       "      <td>0</td>\n",
       "      <td>11877356</td>\n",
       "      <td>99.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>coexpression</th>\n",
       "      <td>0</td>\n",
       "      <td>5412870</td>\n",
       "      <td>45.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experimental</th>\n",
       "      <td>0</td>\n",
       "      <td>6987602</td>\n",
       "      <td>58.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database</th>\n",
       "      <td>0</td>\n",
       "      <td>11537060</td>\n",
       "      <td>96.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>textmining</th>\n",
       "      <td>0</td>\n",
       "      <td>1769058</td>\n",
       "      <td>14.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>combined_score</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                count_nulls  count_zeros  perc_zeros\n",
       "neighborhood              0     11528086       96.56\n",
       "fusion                    0     11915600       99.81\n",
       "cooccurence               0     11877356       99.49\n",
       "coexpression              0      5412870       45.34\n",
       "experimental              0      6987602       58.53\n",
       "database                  0     11537060       96.64\n",
       "textmining                0      1769058       14.82\n",
       "combined_score            0            0        0.00"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for missing values\n",
    "df_stats = protein_links_v2[protein_links_v2.columns[~protein_links_v2.columns.isin(['protein1', 'protein2'])]].apply(lambda x: sum(x.isnull())).to_frame(name='count_nulls')\n",
    "df_stats['count_zeros'] = protein_links_v2[protein_links_v2.columns[~protein_links_v2.columns.isin(['protein1', 'protein2'])]].apply(lambda x: sum(x==0))\n",
    "df_stats['perc_zeros'] = round(100*df_stats['count_zeros'] / protein_links_v2.shape[0], 2)\n",
    "df_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "      <td>11938498.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.20</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.88</td>\n",
       "      <td>62.38</td>\n",
       "      <td>68.92</td>\n",
       "      <td>24.14</td>\n",
       "      <td>180.26</td>\n",
       "      <td>280.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>20.39</td>\n",
       "      <td>7.57</td>\n",
       "      <td>13.41</td>\n",
       "      <td>96.64</td>\n",
       "      <td>119.62</td>\n",
       "      <td>132.79</td>\n",
       "      <td>152.82</td>\n",
       "      <td>167.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>150.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>67.00</td>\n",
       "      <td>175.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>54.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>161.00</td>\n",
       "      <td>216.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>76.00</td>\n",
       "      <td>129.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>241.00</td>\n",
       "      <td>309.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>385.00</td>\n",
       "      <td>900.00</td>\n",
       "      <td>448.00</td>\n",
       "      <td>999.00</td>\n",
       "      <td>999.00</td>\n",
       "      <td>900.00</td>\n",
       "      <td>997.00</td>\n",
       "      <td>999.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      neighborhood       fusion  cooccurence coexpression experimental  \\\n",
       "count  11938498.00  11938498.00  11938498.00  11938498.00  11938498.00   \n",
       "mean          3.20         0.16         0.88        62.38        68.92   \n",
       "std          20.39         7.57        13.41        96.64       119.62   \n",
       "min           0.00         0.00         0.00         0.00         0.00   \n",
       "25%           0.00         0.00         0.00         0.00         0.00   \n",
       "50%           0.00         0.00         0.00        54.00         0.00   \n",
       "75%           0.00         0.00         0.00        76.00       129.00   \n",
       "max         385.00       900.00       448.00       999.00       999.00   \n",
       "\n",
       "          database   textmining combined_score  \n",
       "count  11938498.00  11938498.00    11938498.00  \n",
       "mean         24.14       180.26         280.01  \n",
       "std         132.79       152.82         167.39  \n",
       "min           0.00         0.00         150.00  \n",
       "25%           0.00        67.00         175.00  \n",
       "50%           0.00       161.00         216.00  \n",
       "75%           0.00       241.00         309.00  \n",
       "max         900.00       997.00         999.00  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_links_v1.describe().apply(lambda s: s.apply('{0:.2f}'.format))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `neighborhood`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_by_neighborhood = protein_links_v1.groupby(['neighborhood']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAEZCAYAAACTsIJzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAABiDklEQVR4nO3dd0CU9R/A8fcNpmzFrai498gBTnILbiu1TLNyZFpamaZlmZqalaudLfXnyNDCvdIchJqKK7eAAiIiU9at3x/HPd7JOOZxyPf1T9xzd8/z4Umez/Mdz+cr0+l0OgRBEAQhi7y0AxAEQRCsi0gMgiAIggmRGARBEAQTIjEIgiAIJkRiEARBEEyIxCAIgiCYEIlBEMqZ77//ntmzZ5d2GGVKeTtnZT4x3LlzhyZNmjB48GAGDx7MwIEDGTlyJDt37pQ+s2LFCrZt25bnflavXs3+/ftzfM/4+40aNeLBgwcFivHcuXN88MEHAJw/f55p06YV6PvmBAcHM2bMGPr06cOgQYN46aWXOHXqlPR+dHQ0AQEBDB48mDNnzmT7/qVLl5g4cSK9e/dm8ODBjBo1KtdzUR5s2LCBQYMGMWDAAPz9/XnnnXeIioqS3j969Ch+fn6MGDGC9PR0k+8+/fTTnD9/3mTb+fPnefrpp80ed86cORw/frx4fok8jBkzhgsXLhAREVHofeT333Fufy+BgYFMnDix0McH/d9+mzZtirSPvOzevZsxY8YAeZ+zMWPG8N1332Xb/uOPPzJ58uRc979q1Srmz59ffAEXJ10Zd/v2bV3r1q1Ntt25c0fXq1cv3e7du/O9nxdeeEG3a9cus59r2LChLi4urkAx/v7777oJEyYU6Dv5tX//fl2vXr10p0+flradOXNG16NHD92hQ4d0Op1Ot3XrVt3YsWNz/H5oaKiua9euur/++kvadv36dZ2/v79uw4YNJRKzNVu8eLFu3LhxuqioKJ1Op9NpNBrd1q1bdV26dNFFR0frdDqdbtasWbovv/wyx+/7+fnpzp07Z7Lt3LlzOj8/v5IN3Erl9vdSHH8TOf3tF6ddu3bpXnjhBbOf27lzp65Pnz7Ztvft21d35MiRXL+3cuVK3UcffVSkGEuKsrQTU0moUaMG06ZNY82aNfTt25dZs2bRoEEDXn75ZVauXMm+ffuwsbHB3d2dTz75hH379nHhwgWWLl2KQqHgwIEDJCQkcPv2bXr06EFcXJz0fYDly5dz/vx5tFotb775Jn5+fgQGBrJnzx6+/fZbAOn1hx9+yMqVK0lOTmb27NkMGTKEjz/+mO3bt5OcnMxHH33E5cuXkclkdO3alRkzZqBUKmnRogUTJkzg2LFj3Lt3j1deeYXRo0dn+12XLl3K3LlzTe6cWrduzXvvvcfSpUuxs7Nj+fLlJCcnM2bMGNauXWvy/RUrVjBp0iR69OghbfP29mbp0qWMGzeOYcOGIZfL+fTTTzl06BAKhYI2bdowb968XLd/++23xMfHS62kVatWSa/HjBmDq6srN2/eZNSoUQwZMoSFCxdy9epVVCoVPj4+zJw50+w5+Pbbb9m6dStKpRIvLy8WL16Ms7Mzv/32Gxs2bECr1eLm5sb777+Pt7c3p06dYvHixWi1WgAmTpxI3759Tc7F3bt32bhxI4cOHcLV1RUAuVzOkCFDuHDhAt9++y01atTgwIED2NnZkZyczLvvvlugf5uBgYHs27cPuVxOeHg49vb2LFmyBG9vb8aMGcPzzz9Pv379WL9+Pb/88gvOzs506dKFoKAgDh48aHIuHz+3ycnJuZ7LvMyaNQsnJyeuXLnC3bt3adSoEUuWLKFChQrcuHGDhQsXkpCQgEajYcyYMYwYMYKQkBDp3/GDBw+YPXs2ERERuLm54enpSYMGDZg6daoUY2hoKAkJCbz88ss8//zzAMTGxvLyyy9z7949atSowccff4ynpyd3797lww8/JDIyEp1Ox5AhQ3jllVe4c+cOzz//PN7e3kRGRrJ48WI0Gg0ffPAB58+fJzk5mXfeeYe+ffuiUqlYvHgxwcHBKBQKWrZsyezZs3FycuLatWvMnz+fhIQEZDIZ48ePZ8iQIdLfQ1BQEG5ubnh5eeXr/2nv3r1ZtGgRp06d4qmnngLgxIkT6HQ6OnfuzDfffMOBAwdIT08nLS2Nd999l969e5vs4+mnn2bFihW0aNEi2+vTp0+zbNky0tLSkMvlvP766/j5+REbG8u7775LfHw8AN27d+fNN9/MV8zmlPmupNw0btyYq1evmmyLjo7ml19+4ffffycwMJDOnTtz7tw5nn/+eZo3b87MmTOl/2Hp6ens2LGDd955J9u+a9asydatW/n000+ZNWtWnl1L1apVY9q0aTz11FN88sknJu8tWLAANzc3goKC+P3337ly5Qo//vgjAJmZmbi7u7Nx40ZWrlzJJ598QkZGhsn34+PjCQsLo3379tmO6+Pjw/Xr12nSpIl0/MeTAsDp06dz/H7Tpk2RyWRcv36d//3vf1y8eJE//viD7du38/DhQ3bu3JnrdnNcXFzYuXMnY8aMYdGiRTRr1ozAwEC2bdtGfHw8P/30U57n4MCBAwQGBrJp0ya2b99OzZo1WbduHSdOnGDbtm2sX7+ebdu28corr/D6668D+ovTSy+9RGBgIIsWLeKff/7JFldoaCj16tWTkoIxX19f/v33X1555RWefvppxo0bV+CkYHDy5Enef/99tm/fTqtWrbJ1Q1y4cIEvv/yS9evXs2XLlnx3XeZ1Ls25cOECa9asYefOnURGRrJ7927UajXTpk3jrbfeIjAwkHXr1vHjjz9y9uxZk+8uWLCA+vXrs2vXLlasWMHp06dN3q9VqxaBgYGsXr2axYsXo1KpALh16xYffPABQUFBNGzYkIULFwLw9ttv07FjR4KCgtiwYQN//vknO3bsAPTJ+7XXXmPPnj14enqSkZFB586d2bp1K++++y6ffvopAF9//TX37t3jjz/+4I8//kCr1bJ06VLUajWTJ09mzJgxBAUF8f333/P5559z5swZ9u/fz969e9m2bRsbN24kJSUlX+dOqVTy7LPPsmXLFmnbpk2bGD16NFFRURw/fpy1a9cSFBTE9OnTWblyZb72C5CYmMjs2bNZunQpW7du5auvvuLDDz8kKiqKzZs3S9ei9evXEx4eTnJycr73nefvVCx7sUIymQx7e3uTbVWqVKFx48YMHTqUbt260a1bN3x8fHL8frt27XLd96hRowBo2LAh3t7eOfbb58fff//Nhg0bkMlk2NraMnLkSH755RcmTJgAQM+ePQFo1qwZmZmZpKamYmdnl20/arU62zbDH59MJjMbR07fB/2FWSaTcfz4cQYPHiydz+XLlwMwadKkHLevWrUqz+MZ7qoADh06xPnz56U/qsf77HM6B8HBwfTr10+6gBsGBZcuXUp4eDgjR46Uvp+UlERCQgL9+/dn/vz5HDx4EF9fX2bMmFGoc2FOTp/R6XTI5Y/uwZo1a0bVqlUBfQLet2+fyedDQkLo3Lkznp6eAIwePZpjx46ZPba5c5mXrl27YmtrC+j/XScmJhIWFkZERATvvfee9Ln09HQuXbqEt7e3tO3w4cNs3boVgMqVK9OvXz+TfQcEBADQpEkTMjMzpQuur6+vdFc+YsQIRowYQWpqKqdPn5ZukJydnRk2bBh///03rVq1QqlU0rp1a2nfNjY2UsuvcePGxMXFAfq/renTp2NjYwPoxwGmTJlCWFgYGRkZ9OnTB9BfE/r06cORI0dITEykd+/eODk5ATB8+PAcb6Zy8uyzz+Lv709KSgpqtZqjR4/y4Ycf4uzszNKlSwkKCiI8PJzQ0FAePnyYr30CnD17ltjYWKZMmSJtk8lkXLlyha5duzJhwgSio6Px9fXlrbfewtnZOd/7zssTmxjOnz9Pw4YNTbbJ5XLWrVvH+fPnCQ4OZtGiRXTt2pWZM2dm+76jo2Ou+zb+I9dqtSiVSmQyGTqjeoSGC3NetFqtyYVEq9WaXJgMScDwGd1j9Q7d3d2pW7cuJ06coFevXibv/fPPP3h7e+Pi4pJnDG3btiUkJIQmTZqYbD937hw2NjbUq1cvW1fE/fv3pd87p+3mzoXxudVqtaxYsUK60CQlJZmck5zOgUKhMPlMUlISSUlJaLVaBg8eLLXytFot9+7dw9XVlZEjR+Ln58exY8c4cuQIq1evZvfu3SaJtnXr1oSHhxMbGytdlA1CQkLyNdDp7u5OQkJCtvPi5uYmvTa+YXn8XBneN95muLjl9Hnjc2vuXOYlp5g0Gg3Ozs788ccfJr+Ls7OzSatBqVSaxGT892F437BfePTvWKFQmMSuVCrRarXZzofx34Wtra3Jv7vHz43xdx5/rVKp0Gg02c6JTqeT9m98bOP4zKlSpQq+vr7s3LmT1NRU+vbti7OzMxcvXuS1115j3LhxdO7cmfbt2/PRRx/luA/jY2dmZgKg0Wjw9vbmt99+k96LiYnBw8MDGxsbDhw4QHBwMP/88w/PPPMM33//Pc2bN8933Ll5IruSbt26xVdffcX48eNNtl++fJmAgAC8vb2ZOHEi48aNk2aQKBSKXO8WH2e4O7p48SIRERG0atUKDw8Prl27RkZGBiqVij179kifz23fXbp0Yd26deh0OjIzM9m8eTO+vr4F+l1nz57NokWLTP5Qz5w5w+LFi3n77bfNfv+tt97ihx9+4PDhw9K2GzduMHv2bN544w3s7Ozw8fFh+/btZGZmotVq+fDDD9mxY0eu293d3bl48SI6nY6UlBT++uuvXI/fpUsXfv75Z+kcTJ48mXXr1uUZs6+vL/v27ZPuPFetWsXPP/9Mly5d2LFjB/fu3QP0s4vGjh0LwMiRI/nvv/8YNmwYH3/8MUlJScTGxprst0qVKowZM4YZM2YQExMjbf/999/Zu3cvr776qtnz2a1bNzZs2CD9YaelpbFp0ya6d+9u9rvGv19wcLA0EyowMFB6L69zW5hzmZe6detib28vJQbD7LYLFy6YfK579+5SKyU+Pp79+/fnKyGFhIRIv+PGjRvp1q0bTk5OtGrVivXr1wOQnJzMtm3bCvx30bVrVzZs2IBKpUKr1bJ+/Xo6d+4s3ejs3bsX0F9k9+zZg6+vL926dWP37t3STYZxQsyP559/nqCgILZt2yaNo5w8eZLmzZvz0ksv0aFDBw4cOIBGo8n2XQ8PD+m8hoSESP82DTcrJ0+eBOC///6jb9++xMTEsGzZMr766it69erFnDlzqF+/PteuXStQzLl5IloM6enpDB48GNDfrdjZ2TFjxgyTAVXQNzX79+/P8OHDcXR0xN7enrlz5wL6wZ7PP/88X3f6t2/fZsiQIchkMj7//HPc3Nyku4H+/fvj6elJx44duXLlCqD/n/vll1/y+uuvS9PfAObOncuCBQsYOHAgKpWKrl27MmnSpAL97t27d2fJkiWsWLGCmJgYtFotVatWZcmSJXTq1Mns95s2bcqaNWtYsWIFixYtQqFQ4OLiwtSpU6UugZEjRxIZGcmwYcPQ6XR06NCBMWPGIJPJctyelpbGkSNH6NOnD1WqVKFDhw7Z7gIN5syZw8KFC6Vz4OvryyuvvGL2d75+/brUpVe/fn0+/vhjnJycePXVVxk/fjwymQwnJydWr16NTCbj7bffZtGiRSxfvhyZTMbrr79OzZo1s+37rbfe4rfffmPy5MlkZmaSmZlJixYt2LhxIzVq1DB7PidNmsTnn3/O0KFDpRuCnj17FmhqZt26dfnggw+YOHEiNjY21K1bV3pv0KBBuZ7bwpzLvNja2vLVV1+xcOFCfvjhB9RqNW+88Qbt2rUjJCRE+tzs2bOZO3cuAwcOxM3NjerVq2frxs1Jw4YNee+997h//z716tWTpm4uW7aM+fPnExgYSGZmJgMHDmTYsGFERkbmO/bJkyezZMkShgwZglqtpmXLlrz//vvY2Njw1VdfsWDBAlatWoVGo2HKlCnS38qVK1cYPnw4Li4uNG7cWBrYzY+OHTuyYMECXF1dadSoEaDvRtu7dy/9+/dHq9Xi5+dHYmJitvGLt99+mw8//JBNmzbRrFkzmjVrBugTxsqVK1m6dCkZGRnodDqWLl1KzZo1GTt2LLNmzSIgIABbW1saNWqEv79/vuPNi0yX21+sIAhW4fz587zxxhscPHiwtEPJ0fr162natClt2rQhMzOT0aNHM3Xq1AK1kgTr8kS0GARBKD2GFpuhH79fv34iKZRxosUgCIIgmHgiB58FQRCEwhOJQRAEQTBRZsYY0tPTuXDhAp6engWaXywIglCeaTQaYmNjad68eb5mi0EZSgwXLlyQ5gYLgiAIBbN+/XqTqgN5KTOJwfAk6vr166VyAoIgCELe7t69y/PPP5/taf68lJnEYOg+qlq1ao4PJgmCIAi5K0gXvBh8FgRBEEyIxCAIgiCYEIlBEARBMCESgyAIgmBCJAZBEATBhEgMgiAIgokSTQwpKSkEBARw584dQL+AjGEJvBkzZkiLmZS0E5fusnLTmVzXBBAEQRAeKbHEEBoayqhRowgLCwP0SWLq1KnMnz9fWtjbePHskpSQnMG+ExFcDsv/ohuCIAjlVYklhs2bNzNv3jwqV64MwLFjx2jdujWNGzcG9KuX9e7du6QOb6Jb6xo42ivZefyWRY4nCIJQlpVYYli4cKFJXY7w8HAcHR2ZPn06gwcPZtWqVWYXqi8u9nZKeravzdHQKBKSMyxyTEEQhLLKYoPPGo2Go0ePMmPGDAIDA0lLS+O7776z1OHp71MHtUbLvhPhFjumIAhCWWSxxFCpUiVatWpFrVq1UCgU9O/fn3Pnzlnq8NSq4kyrBpXYFRyGRisGoQVBEHJjscTQpUsXLl68SHR0NAB//fUXzZo1s9ThARjgW5fY+DT+/S/GoscVBEEoSyyWGKpVq8b8+fOZNGkS/fr1IzExkYkTJ1rq8AB0bFaViq727DgmBqEFQRByU+Jltw8ePCj93KNHD3r06FHSh8yVQiGnn08d1u++TNT9FKpXciq1WARBEKxVuXvyuU9HLxRyGbuOh5V2KIIgCFap3CUGDxd7fFpUY/+JCDJUmtIORxAEweqUu8QAMKBzXVLSVBw5c6e0QxEEQbA65TIxNK9XkdpVndkhupMEQRCyKZeJQSaTMcC3LtdvJ3A1QtRPEgRBMFYuEwOAX7uaONgpRP0kQRCEx5TbxOBob4Nfu1ocORNJ0kPLlP8WBEEoC8ptYgD9k9CZai37T0SUdiiCIAhWo1wnBq9qLjSrV5FdwbfQivpJgiAIQDlPDAD+vnW5G5fK6Sv3SjsUQRAEq1DuE0OnFtVwd7YTg9CCIAhZyn1isFHK6dPJi1P/xRDzILW0wxEEQSh15T4xAPTrVAeZTMbu4LDSDkUQBKHUicQAVHJzoGOzquwNCSdT1E8SBKGcE4khi79vXZIeZnLsXFRphyIIglCqRGLI0rJBJWp4OrFTLOIjCEI5JxJDFn39pDpcDo/nxp2E0g5HEASh1JRoYkhJSSEgIIA7d0zLW69bt44xY8aU5KEL5en2tbGzVbBTVF0VBKEcK7HEEBoayqhRowgLCzPZfv36db777ruSOmyRODnY0KNtTQ6dvkNKmqq0wxEEQSgVJZYYNm/ezLx586hcubK0LTMzkw8++IBp06aV1GGLbIBvXTJVGg6eFPWTBEEon0osMSxcuJCnnnrKZNtnn33G8OHDqVWrVkkdtsjq1XClsZc7O4+L+kmCIJRPFht8PnbsGNHR0QwfPtxShyw0/851iYx9yLnrsaUdiiAIgsVZLDFs376da9euMXjwYObOncuFCxd48803LXX4AuncqjouFWzFILQgCOWS0lIH+uSTT6SfQ0JCWL16NcuXL7fU4QvERqmgT0cvAv+6xv2ENCq5OZR2SIIgCBYjnmPIRX+fOuhA1E8SBKHcKfEWw8GDB7Nt69ixIx07dizpQxdJZQ9H2jepyp6QcJ7r3QgbpcihgiCUD+Jql4cBneuQkJzBP+ejSzsUQRAEixGJIQ9tGlamWsUK7BCL+AiCUI6IxJAHuVxGf986XLwZR1h0UmmHIwiCYBEiMZjRs31tbJVysfSnIAjlhkgMZrhUsKVrmxoc+vc2qemifpIgCE8+kRjywb9zXdIyNPx16nZphyIIglDiRGLIhwa13GlQy40dx8PQ6UT9JEEQnmwiMeTTAN+63I5J5sLNuNIORRAEoUSJxJBPXdvUwNnRRiz9KQjCE8/sk88HDx5k37593Lp1C7lcTr169ejXrx9dunSxRHxWw85GQa8OXvz59w0eJKXj4WJf2iEJgiCUiFxbDDdv3uS5555jw4YNNGrUiFdeeYUJEybQuHFj1q5dy8iRI7l+/bolYy11/X3qoNHq2PNPeGmHIgiCUGJybTF8/fXXLFu2LMdFdV544QUiIiJYsWIFn332WYkGaE2qVapA28aV2R0cxjM9G6BUiJ44QRCePLle2T799NM8V1qrXbt2uUoKBv6+dXmQlE7IxbulHYogCEKJMHvLe//+fQ4cOADok8XYsWO5fPlyiQdmrdo1qUJldwcxCC0IwhPLbGKYNWsWt2/fJjg4mCNHjjB48GAWLFhgidiskkIuo59PHc5dv8/tmOTSDkcQBKHYmU0MCQkJjBs3jr///puAgACGDRtGWlqaJWKzWn06eqFUyNklFvERBOEJZDYxqFQqVCoVR44cwdfXl7S0NFJTU/O185SUFAICArhz5w4AmzZtIiAggIEDBzJ79mwyMzOLFn0pcXWyo0vr6hw4GUFahrq0wxEEQShWZhNDz5498fHxwd3dnebNm/PMM88QEBBgdsehoaGMGjWKsLAwAG7dusWaNWvYuHEjf/75J1qtlv/9739F/gVKi79vXVLT1Rw+fae0QxEEQShWZhPDtGnT2L59O2vXrgVg2bJlTJkyxeyON2/ezLx586hcuTIAtra2zJs3DycnJ2QyGQ0bNiQqKqqI4ZeeRl7u1Kvuyo5jt0T9JEEQnihmE4NWq2XHjh3MmjWLlJQUDh8+jEajMbvjhQsX8tRTT0mva9SoQefOnQF48OAB69evp2fPnkUIvXTJZDIGdK5DWHQS/4U9KO1wBEEQio3ZxLB06VKuXLlCaGgoAEeOHOGTTz4p9AFjYmIYO3Ysw4cPp2PHjoXejzXo3qYmFeyV7DwWVtqhCIIgFBuziSE4OJjFixdjZ2eHk5MTP/74I8eOHSvUwW7cuMHIkSMZOnRovrqjrJ29nZKe7Wtz7FwkCckZpR2OIAhCsTCbGJRKJXL5o4/Z2tqiVJqtvZdNSkoKL7/8Mm+88Qbjx48v8PetVX/fOqg1OvadEPWTBEF4MphNDA0bNmT9+vVoNBpu3rzJBx98QOPGjQt8oC1btnD//n1++uknBg8ezODBg1mxYkWhgrYmNSs707qBJ7uCw9BoxSC0IAhln9lb/zlz5rBo0SLi4uIYPXo0Xbp0Yc6cOfk+wMGDBwEYN24c48aNK3Sg1mxA5zos+vkkpy7dpWPzaqUdjiAIQpGYTQxOTk5MnjyZRYsWkZKSQkREBO7u7paIrczo0LQqFV3t2Xk8TCQGQRDKPLNdSWvXruW1114DID4+nqlTp/Lbb7+VeGBliUIhp59PHU5fuUfU/ZTSDkcQBKFIzCaGTZs2sWHDBgBq1arFtm3b+PXXX0s8sLKmb0cvFHIZu46HlXYogiAIRWI2MWg0GpycnKTXzs7OyGSyEg2qLHJ3sce3ZXX2n4ggPVPUTxIEoewymxjq1avHsmXLuH37Nrdv32bFihXUqVPHAqGVPQN865CSpuLo2cjSDkUQBKHQzCaGjz76iLCwMIYMGcKIESMICwvjww8/tEBoZU+zehXxquos6icJglCmmZ2VVKlSJVavXm2JWMo8ff2kunz9+zmu3U6gYW0xe0sQhLLHbGK4efMm33//PQkJCSZ3wd98802JBlZW9Whbk5+3X2THsVsiMQiCUCaZTQyzZs2iZcuWtG/fXgw654OjvQ1+7Wqx70QELw9qjksF29IOSRAEoUDMJoa0tDTmzp1riVieGAM612Xn8TD2nwhnmF+D0g5HEAShQMwOPnt5eXHv3j1LxPLE8KrqQnPviuw8HoZW1E8SBKGMMdti0Gq1BAQE0KxZM+zs7KTtYowhbwN867J07SlOX7nHU02qlHY4giAI+WY2MfTu3ZvevXtbIpYnik+Larg727Hz+C2RGARBKFPMJoahQ4dy9+5drly5QpcuXYiJiaF69eqWiK1MUyrk9O1Uh037r3A37iFVK1Yo7ZAEQRDyxewYw+HDhxk5ciQfffQRcXFx+Pv7s3//fkvEVub17eSFTCZjd3BYaYciCIKQb2YTw+rVq9m8eTMuLi5UrlyZ//3vf6xcudISsZV5ldwc6NisKvtORJCp0pR2OIIgCPmSryJ6lStXll43adIk388zpKSkEBAQwJ07dwA4fvw4AwcOpE+fPnzxxReFDLls8e9cl6SHmRwNjSrtUARBEPLFbGJwcHAgKipKSganTp0ymZ2Um9DQUEaNGkVYWBgA6enpvPfee3z11Vfs3LmTCxcucPjw4aJFXwa0rF+JGp5O7Dx+q7RDEQRByBezieGtt95i/PjxRERE8NxzzzFlyhTefvttszvevHkz8+bNk1ob586dw8vLi1q1aqFUKhk4cCC7d+8u+m9g5fT1k+pwJTye63cSSjscQRAEs8zOSqpduzabN2/mzJkzaLVaWrVqhYeHh9kdL1y40OT1vXv38PT0lF5XrlyZmJiYQoRc9jz9VG1+3fkfu46HMfXZ1qUdjiAIQp7MthheeOEFXFxc6N69O35+fvlKCjnRarUmYxM6na7c1F5ycrChR9uaHDp9h9R0VWmHIwiCkCeziaFGjRqcPn0arVZbpANVrVqV2NhY6XVsbKzJoPaTrl3jKmSqNETFPiztUIpdfHI6D9NEwhOEJ4XZrqQbN24wevRolEoltra20p3+6dOnC3SgVq1acevWLcLDw6lZsybbt29n+PDhhQ68rPF0dwAgNiGN+rXcSjeYYvbih3twsFOyeZF/aYciCEIxMJsY1q9fXywHsrOzY/HixUydOpWMjAy6d+9Ov379imXfZYGnmz4xfP17KPHJ6QzwrVvKERWvtAyxzrUgPCnMJoaEhIQct9eoUSNfBzh48KD0s4+PD3/++Wf+InvCuFSwxVYpJz45g69/P0d/nzrFNsZSnsZrBEEoeWYTw9SpU6WfVSoVsbGxNG/enC1btpRoYE8amUyGSvNonOaHPy/wzNMNcXO2Iz1Tzd9nIun5VC0UCrPDPiauRsTz1oq/WTylC83qVSzusAVBKIfMJgbjO36AkJAQgoKCSiygJ5nRyqj8+fdNDp++w7Jp3fh+2wVOXLqLh4t9gSuxhl7TD+iHXLwrEoMgCMWiYLenQMeOHbl48WJJxPLEq5ZVYfWjCT588HJH1Bodry7az4lLdwGIvl/wGUs2SgUAKrWoxSQIQvEw22IwTgI6nY4LFy6Qnp5eokE9qRa/3oUHienSrKR5L3fix6AL1K7qwqF/bxMVm1Lgfdoo9bldpS7adGJBEASDAo0xyGQyPDw8+PDDD0sypieWh4s9Hi720usmdT34dFo3AG5GJhBZiMRgaCmI6q2CIBSXfI0xpKSk4OTkREZGBikpKVSsKPqyi1t1Tycuh8dn2x5xN4k3Pj/EJ1O60Ngr+1Pnaen6aaKZKtFiEASheJgdY9i5cyfDhg0DICoqioCAgGwD0kLRVa/kRGx8arYuoW8Cz6PW6Lh080GO30vNen4gOTWzxGN8UkTcTeLQ6TulHYYgWC2zieGbb77h119/BaBu3boEBgayatWqEg+svPFwtUeng8SUDJPt52/cB8Clgk2O3zM8WJb0UCSG/Jry6V98tv7f0g5DEKyW2cSg1WqpWrWq9LpatWpFrpskZOfmpF/jIsEoMaQbPU2s0eqyfQcedSU9nlCEsi01XcXFm3GlHYZQTplNDB4eHmzcuBG1Wo1Go2HLli1UqlTJErGVK1JiSH50gU80agXklhhSjVoMl27FFWrKa3ml0+V8Tq3B0rWnmPXlUVJEF6FQCswmhvnz57N582ZatWpFy5YtpQV4hOLl5qxPDMZ3/sY/azS5tBiyEoNGq+Pd1UeZsdyyq+Jpc0lYhbX/RDjnrsea/2AxUOdyTq3BtdsJgHXHKDy5zM5KqlOnDmvXrkWtVqNQKMjIyBCzkkqAITGYtBiME0NuXUkZampVceJ2jH6qa0qaCq1Wh1xumdpJucVVWCs2nQUg6LPBxbrfnKjUGuk5EGtT3AlXEAoiX7OShg4diqurK7GxsWJWUgmxt1Vga6MwGWMwTgy5jeukpaupW92VWlWcpG1DZ/7J5fCcZzEVN60Vd8eYY81344aEqxHjeUIpELOSrIRMJsPN2c4kMSSkmB9jSMtQ42CnZEz/JnjXdAVAq4PFv5wk6MhN7ieklWjcGk3ZvXCpCxh7Smom63dfLvZWUk4M4x+WOJYgPE7MSrIibk62JD7WlaTMqraa1+Czg50SnxbVmTyspbQ9LUPNd9vOs3HflRKNuSxftwpaRuSHPy+wcd8VTly8W0IRPWLoShJdSkJpELOSrIi7sz23Y5J5eeE+fgq6SEJKBu4u+rGHnAafNRotmSoNjvb6Zxy8qrlQ2cOR2WPb87+PB1CrilOJz1IqzguXpWcJFbTFkJo1NdgS3Wda0WIQSpHZweePPvqIt956i/nz5yOTyWjWrBmfffaZJWIrdxrX8SAk62408NB1ABrWduNBYnqOfc2nr9wDoHZVZwDsbZWsmdNber9uNVeuRGQvs1GcirMPvKAX6qIqaIvBkATlFlgUyXCsstxVJ5RdZlsMhnGF4OBgQkJC2Lx5M7Vq1SrSQf/44w/8/f3x9/dnyZIlRdrXk6R1Q09Av9qbT4tqKBUynuvdCIVcluOd+e7gcDxc7OjYrGq29wCqVHQkNiGtRC8uxnEV9Y7f0vWe1AVMDIa7d4UFZnwZTqtoMQilIc8WQ0xMDN999x3//vsvMpmMtm3b8uqrr5qMORRUWloaCxcuZPfu3bi4uDBq1CiOHz+Or69voff5pKhX3RUPFzvaNKrM1Gdak/QwE3cXexQKWY4zaKLup9CkTkVpHOJxVTwqoNXquJ+YThUPxxKJ2fjCpdWBogjXzEwLrClh3CopdIvBQlOBQSQGoXTk2mKIjo7mmWeeQaFQ8MYbbzBp0iR0Oh0jRowgMjKy0AfUaDRotVrS0tJQq9Wo1Wrs7OwKvb8niVwu47M3ujNhSAsUCjnuWSW65XJ5jl02Go0uz3n4VbOSQciFaHYHh5VIzMYthqJOSlAVscWQkqYyKSOSk4zMR8mnoF1XpdG9IwafhdKQa4th+fLlzJgxgyFDhkjb+vbtS7NmzVi+fDmffvppoQ7o5OTEG2+8Qf/+/XFwcKB9+/a0bdu2UPt6ElVyc8i2TSGX5XjnqNJoUeRxi+7pod/X939cAKCfT53iCdKI8UCsWqPDxuyoVe6K2mL46PtgvKq58PozrXP9TIbRuhWqgiaGrN/Vks8/WHrcpbSo1Fouhz+ghbeY2GINcr3dvHTpkklSMBg+fDjnzp0r9AEvX77M77//zl9//cWRI0eQy+WsWbOm0PsrD5SKnMcYNBptrt1IAM6OtiavS2KVN+PZUkXt9ihqfHfjUol5kJrnZ9IzH7UoCjvGYMllVMtLV9KPQRd476tj3IpKLO1QBPJIDHkNJNra2ub6njlHjx7Fx8eHihUrYmtry7Bhwzhx4kSh91ceyOXyHKerqjW6PAdCHe1Mb9+NL4rFxbjFUNQulqKsQqfT6UhJy+RhmirPzxl3JRW4xaA1tBgs2JVkxU9nF6dbUUmAvjtQKH25JgaFQkFMTEy27TExMUVKDI0bN+b48eOkpqai0+k4ePAgLVq0KPT+ygN9V1IOYwxaLco8xhgUCjn2tgrpdZqZ/vfCMB1jKOKspCK0GNIzNag1OrMXFpOupIIOPmclQZUFL9blpcWgteCML8G8XK8qI0eO5L333iMl5dE6xHFxccycOZPRo0cX+oBdunTB39+fYcOGMWjQINRqNRMmTCj0/sqD3MYY1GotSnneM44ND79B3olBrdHy7uojBa5sahxXkbuSijD4bFjBzmyLIcNo8LmQs5Is25VUPsYYDL+nSAzWIdehwlGjRhEREUHXrl2pX78+arWasLAwXnzxRYYPH16kg06YMEEkgwJQKHJJDFpdnoPPABUclDzQt9LznLHzME3FpVsPuH47gZb1PfMdm3EroahdLEUZfE5J1SeElDQVOp0OWS4PoRm3GAo8K8kw+KwWLYbipimFqcBC7vKcQ/Luu+8ybtw4QkNDAWjVqhVVqlSxSGDCIwq5PFs3jVarQ6vVYZPH4DNAhXy2GDTaws24Kc6uJOMWQ14X95wYWgxarY60DLVJS8nYw/RHLYqCdiUZxnksOcZQXhKD4d+ODJEYrIHZyYVVqlShT58+lohFyIVcLss2+Cw1vc0kBkcH48SQ+x254WJX0AHkvLqS4hLTuHY7gU7Nq+VrX8YtBo1Wh7IAT8sZjy08TMs9MZy/fh+5TP8wXkEv8I9mJYnB5+JmOLdluYz7k8Q6VykRTOQ0+Gy4szd38TRuMeQ1K8mQeAo7UweyJ4Zdx8P45OcT+b7rNR58Lmj/v/ESmClpOS+HqdHqOHkpho5ZiaqgF3jDrCnLthjKyRiDRlSTtSYiMZQBOQ0+G+7szbYY7B81Cs0NPuv3W/iupMcvmA/TVGh1kJHPabKqIvT/J6catxhyHoC+FZVIQkoGPi2qFeoYlkoMxq02a15MqDhJT5WLxGAVcr2qvPLKKwD88ssvFgtGyJlCkX2M4VGLIf9jDHkNPhsuduoC3qFq8hhjSM06Xn6nyRq3GAracjFtMeScGC7digOgeb1KKBWyArcYDAPXJd2VZHweysuF0tAyEi0G65DrGMONGzcICgpi7dq1VKuWvY9YjDtYTk4tBsOF3FxXkqPDo//FqXkNPhsGVgs5t994HwaGhJCemb/ZRiYthgLO/MmrxaDRaPl5xyW2Hb5BRVd7PN0dUCrkVttiME485WVRLGmMQSQGq5BrYpg2bRpbtmwhLi6OtWvXmrwnk8lEYrAguVxmMs0SHl2cFGaeYzBtMWhITMng6NlI+vnWNZkzbmgpFPQO1XSMwfQiZkgMaekFbzEU9OIbdT+FmpWduHMvhROX7tK0bkVS01Xcikrk7NX7HD5zB4CalfVrY9so5QVKghqNVmqllXSLQfXYIHx5IK1xLQafrUKuiWHo0KEMHTqUTz75hNmzZ1syJuExOY4xaPM3+OzkYDpddd3uy+wODmPj/quMeLoBg7rWQyZ7NOupwFM4tbl3e0iJIZ9jDMazkgqSGHQ6HeHRSfi2rI5Pi2r8duAax89Fm3ymd4fatGtSBe8a+nWxlQp5gbqrTJ5/KPHEUP66ksRSptYlz+mqqampVK5cmSlTpkjrMTz77LM4OTlZKj4B/UXs8WmLhotTXiUxADo0q8org5uz49gt9oaES9sTkjP44Y8LxDxI5dXBzaX9FXzw+dHPuXUl5XeMwfg5hoIkhgdJ6SSnqqhTzQX/znU5fi6KyFj9kqa1qjjTsLYbY/2b4ur0qLy7jVKe7ySYnqEmwWgt7oKOfxSUSWIoJ4PPoivJuuSaGBITE3nuuefw9vbGx8eHzMxMQkJC2LRpExs3bsTd3d2ScZZr8hynq+avK8nR3obB3bz5Iav0NsC8VzrRon4lftlxiaAjN+naqobUlVTwwWdtjj+D0RiDmcSQmq7CwU5Z6BZDWLT+0W6vai7IZDKGdK/P+t2X+XLm0zg72uT4oJxSkXtXUmRsCsmpmTT28gBgxorD3I55VBqmpMcYjPdfXqaragvZlSmUjFwTw8qVKxkxYoQ0Owlg/PjxfPPNN6xYsYIPP/zQEvEJFK0ryaBjs6qcv3GfH+f2oUJW91Kv9rUJOnKT+OR0aXZTYesHGcdkYBhbyOvBuvRMNc/N2cmgrvVITTcuiZ3/C0R4VmKoU80FgL6dvOjbySvPJ6eVyuxdSXfjHuLqZMekxQcAGN2nEXa2Sikp2CjlVK3oWPKzkoy6rcrLHbT0HIMYY7AKuSaGEydOsG3btmzbJ0yYwIABA0oyJuExCrk8j1lJ+XsUZfbY9mh1mKz4Zqi8mp6pwT6rYG6BB591OScGnU6Xr+mqVyPiAfjzyE0quzvg5GBDSpqqQHflt6KTqOhqL60/kZ9SGjZKObHxqbz5xSHG9G9CeqaGpWtPmVyI/7f3CgC2SjmZai0dmlUlMSWjxFsM6caF/spJV5Lh31F5edLb2uU5xqBQKLJtk8vlRSq7LRRcTkX0CpoYFAo5j//ftMtKDBmZaqnlUeAyEUZ/yMZ/1JlqrXSRzeuJ60u3Hkg/34tPo3UDT85eiy1QP354dBJeWa2F/LJVKvgvTH/sD7//R9reyMudZnUrEnjoOgCvDW9JpxbVSH6YiauTHcvW/1si5cuN3Yp+tFhNeelKErOSrEuuiUGr1ZKcnIyzs7PJ9qSkpBIPSjClkMvQah4fY8iqX1+AekKPs7fV/+9Pz9Rgl/VzYSuOgun4hPEU1bzGGEKvxVKrijO3Y5IB8K7pytlrsfmOQ63RcjsmmbaNKhco7k7Nq0qJAaBeDVcGdqlLrw5eAHhVc8bTzZEW9fVLTbo769ffLszzDwV1LSKBSm4OJKVklJuuJMM/o/Ly+1q7XG83AwIC+Oijj0xWcsvIyGDevHkMHjzYIsEJevI8SmLkt8WQE+OuJMP+CjoLxqSIntF3je+qc3uw7mhoJBduxPH0U7XwcNFfeL1rugH5H+vYfvQWao2uwC2Gvp3qADCqTyM2L/JnxYweUlIAePqp2lJSMFbQ5x8KKkOl4d8r92hY2y3XcutPMpEYrEOuLYYJEyYwffp0/Pz8aN26NWq1mnPnztG+fXvGjx9vyRjLvbyffC58YlAo5CgVcjIy1agLuWxlboPPxonB0GLQ6fSlwhf/epKOzapyJDSKapUqMLS7Nx2bVeXAyQhqV3XONY70TDVarQ5bGwVhUUk4OihZ8+cFWjWohE8+K7gaVHCw4ffFASgV8gKtAWCjyP8018KY/8M/PExT0aZhZUKv3S93iaG8/b7WKtfEoFAoWLlyJWfPnuXMmTMAvPrqq7Rq1arIBz148CCrV68mLS2Nzp07M3fu3CLv80mmUOQ0+Fz0riTQtxqMWwxFSQx//XubzQeu8uU7fiQ9fDTv31AS48stoez5R/8sxT8X7gL6GUQKhZxaVZwZF9CM6PsPs+Iw/X3vxj3k1UX7aVm/Ek81qcKPQRepVcUZmQzeHNkWezuzFeSzsbXJPoZmjlJZcl1J9+JTOXf9PgN869C3kxdrd/1X5HW0yxoxK8k6mP1rat26Na1bty62A96+fZt58+bx22+/UbFiRcaOHcvhw4fp3r17sR3jSaOQy7LVzCmOriQwJAb1oyJ6RehKunhTX6Tu+LlovthwOis+GWnpai7dipOSgrGmdSuavJamzWq0/Lz9IvHJGbzxXBt+O3ANgHPX70sXj9sxybRvWoVKbg4FirkobEooMRw/F8Unv5wEYFA3b2QyWa5Luj7JRFeSdSj4bVYR7du3jwEDBlC1alUAvvjiC+zs7Mx8q3xT5LBQT3F0JQHY2ylJz9RICaEoLQaD77edB6Bn+1rEJ2UQcTeJn4Iumjxt3M+nDruDw2hW77HEoNS3gFRqLX/9e4cHSekcPHVb/17WwO+FG3E0q1eRHm1r0r1tzYL9wkWkzOpKUmu0/Lb/Kv186uCeNT5ikKHS8NH3//Bsrwa0bqgfFL94M47Fv55kyZQuVPd04nZMMpXcHDh+LorTl+/x99lIAFo39KSGp76ygP6G4Mm/UOZVb0soHRZPDOHh4djY2DBp0iSio6Pp0aMHb775pqXDKFNyGnzO70I95tjbKsgwGXwufNltAA8XOx4kZWCjlPPGc204czWWed8Fcz8xndF9G3MrKhEHOyWThrVkUNd6VPFwNPm+g60SuVzGyUt3eZCUbvLec70bsn73ZQBG9W5Eq4b5X5u6uNhkPdNwOewB/9t7hUOn7/Dt7F6cuHSXBjXdcHex51hoFOdv3CchJYOvZj4NwG8HrpKQnMEff9/A76lavLPySLZ9TxrWEv/OdaXX8hy6EJ9ExlOTRV6wDmYTw/79++nVq1e2nwtLo9Fw6tQp1q5di6OjI5MnT2br1q0MGzasSPt9khkecDNeB1laqMdMSQxz7GyVWV1JhhZD4R5w++DljlRwsMHZ0ZYFP4bQtXUNfX2tRpWZPbY9aRlqerStabKwUK0qztn2Z2+npOdTtdh3IiLbe91a16BmZSfcnOxo7p19xpAl1KnmQkamhv0n9fFF3X/IxZtxfLwmBNBXbzWMXRie3zgaGsm/l+9hb6tg5/Ewdh4PA2BId2+a1q1Ih6ZVuHgrjmb1TH8nhVxm0dXiSotpmfEnPxGWBbkmhhdeeIEOHTqwe/duOnbsiLOzM19++WWRE0OlSpXw8fHBw0Nfh6ZXr16cO3dOJIY8GAaYtVqd9LNhFlFxtBgSH2YajTEU7gG3p5pUkZLWt7N7mUxz9m1ZvUD7HNO/ibTYjo1STsv6lfBrVwtbGwXVPUu3gKNvy+p8E3hO6t4C+GXHJennO/f05TOqVnTkblwq+0+E8+WWUJrW9eDNkW05GhrJzuNh9PPx4rlejaTvtayfvfVTXsYYjMuMi8Fn65BrYli2bBnBwcHcu3ePsWPHkpqayoMHD/j555/p0KEDTZs2LdQB/fz8ePfdd0lKSqJChQocOXKEnj17FvoXKA8M6yZsOXiN53rrLyZSddUiDz4ruRefarS0Z8EfcJPLspehyE9Zity4u9jz3rgOhf5+SXKwU9K5VXUOnLxNvequyBUy6UG5iUNbcDQ0ish7Kbw3rgPTPjvEik1nqeHpxPvjO+LkaMszPRvyTM+GJokzN8ocVu57EpmWGX/yW0hlQa5XFVdXV4YOHUrNmjUJDAwkKCgIDw8PZDIZ33zzTaEP2KpVK1555RVGjx7NgAEDqF69OsOHDy/0/soDQ3fRuqz+dXj0lLG5NZ/NsTNMVzU8x1CIhXrkRezOKmt6tq8N6LuNBnapB+jHVgK61OP98R1ZPqM7dau70iKru+v5vo1xcjQtI5OfxCnPYdLBk0itFmMM1ibXFsOzzz5LhQoVePDgATt37qRJkybY29szduxYxo4dW6SDjhgxghEjRhRpH+WJ8bMKkxYfYLhffemCURxdSekZmkddSQVeqEdXoAfEngTN6lakc6vqdG5VHd+W1XFzssOlgv7CX8HBRqpeO7pvI/aG2OPbsmAP3xkocii3/iSKjU+Tfi4PLaSyINfEEBQURGxsLCNGjODSpUsEBgZy69YtXn75ZTp06MDEiRMtGWe5ZrwEZ2RsCis3n+WZng2Qy2VF6rIBfVdSRqZaSjSPD3Kbo9XqKGKjpcyRy2XMerG99Lpt45zrNDX3rlSkQfLyMsZwJavCLogxBmuR55+0p6cn7u7uvP322/zwww/UrVuXOXPmSAPHgmXk9Mei1uiKPL4A+hZDplprskhOQS5GGq223HUlWYoin2MMOp0u29TesuRKeDw1PJ1QKuTl7klva5XrX/T16/qyw8ZrMmzbto169erxzDPPAHDt2rWSjU4AMFlW0kCj0Ra5GwmQqqo+zJoFBAXrTtJqdciL2GoRcpbfFsPGvVcY+9GeMpkcUtNV/BcWRyMvd+RyGeWggVQm5JoYvvvuOxYsWMCNGzeyvXfz5k0++uijIg1CC/l3PyEt2za1RlvkZxgA3Jz1T50b9/MWZABaqyt6vSYhZ/rBZ/NJenNWuZCc/p1YC61Wx7+XY7LNxlq/+zIpaSr6+9RBIRdjDNYi1zGGpUuXsmvXLqZOnUpGRgZeXl5otVoiIiKwt7dnypQp+Pv7WzLWcqt906r89e8dk21qjQ4bZdEvyIYnjw3z7yH3FkPSw0xpkNVAo9GKFkMJUchluSbpsOgkLty4Tz+fOtLEAWtuMZy5eo8Pv/+HxVO6SGVQ0jPU7DsRjl+7WjSu44FcLi8Xg+1lQZ5PPvfv35/+/ftz7do1bt68iUwmo27dujRo0MBS8QlA19Y1cHO2472vjknbElMyqOBQ9JX0qlbUJwZzC9BH33/IhE/289qIVvT3qWP0WZ1oMZQQpUJusipehkrDkl9P0ql5Ndbt+o/45AwuZBUuBIjPocvRWkRm3XhExaZIieGfC9GkZWjo3UE//VcuKx+1ocoCs30RUVFRVKhQgRYtWtC8eXOcnJyIj4839zWhmDk8Vlb6VnQSdQq4OE1O3J3tTdaBhpzLYhjKYe88dstke2x8GhUfKyInFA+5XMbNqETis1oC3wae4+SlGFZtPislgWOhUdLn45PSUak17A4OI1OlyXGfpSXmQarJfwEOnrpNZQ9HqcKuQowxWA2ztZJGjRrFvXv3cHJyQiaTkZycjEKhwN3dnRUrVtC2bVtLxFnuPZ4Y7j1IpW9Hr1w+nX9yuYzK7g5Exj6UtuVUFuNBkr7/+l58qsn26PsptGxg+WJ25YGbk37857tt5+nSqgb7TkTgUsGWpIeZONgp8W1ZjQMnH5XmeJCUzs/bL/HnkZso5DJ6Z/370Ol0WV2POd8HHjkTybkb95kyIu+1Vm5FJTL/h39YMrUrld0d8/yswbpd/3HhZhyO9vp/v3fj9P9+wqKTCL0WyzO9GkrPwcjLSTXZssBsi8HX15dPPvmEkydPcuLECZYvX86wYcP49ttv+eSTTywRo8CjZTiN1ale9BYDQBWPCiavV/92Ntsg4f1E/V1rarqa1HT9DKb0TDX3E9OpVsn0+0LxeHVIC5rU8eDizTj2n4ygsocjLw9qDuiftDZeh6JapQocOn2Hncf1LbobkYmAfiW96csPM/7jvTm2IjQaLUvXnWJ3cBgqtYbQq7GMX7CX1HQVKamZfLb+X5JTMwE4ffke9xPTWbbuX/48cgOdTkd8sum4RnxSOreiEqXXm/Zf5eLNOE5eigHg7oOH3I5JZsbyw9jZKuiV9RQ5GKoIizEGa2C2xXD58mWTBNC3b1++/fZbmjZtikqlyuObQnGyt83+v6o4upJAX+X09JV70usLN+J4kJRORVf9hWfV5rPsDXm0yE7w+WhiHqRKx68uEkOJsFHK8XuqFl9tCeXUfzEM7FpPKrMx3K+BNDBdwV6Js6MN0fcf4pW1NOrlcH39puDzUdy4o79Qh0Un0bC2OwApaSpslHJCr8ZKx4t5kMr6PZeJjU/jclg81+8kcOj0HapUdOSFfk24ficBgP/CHvBf2AMq2NuwfOMZnuvdkKZ1K6KQyVi95Sx341IJXDKQuMTss6Ri4lLZsPcKCrmM1e88bdLyEC0G62E2MajVaq5evUrDhg0BuHr1KlqtloyMDNTqnBd5F4rf4y2GCvZKPItp5bJGXu7ZtkXGpkiJwTgpACzfqF/q1ZAQqlcq3YqnT7KmdR89TNqpeVU83R3YsjgAOxsFJy7ql0d1c7aXCtHNGN2OY+ei2HLwGlGxKVy48Whw+kZkIg1ru3PgZATLN56hfdMq0v9j0HfzeLjqx4vmfR+MXda/uYyspVkNrRCDn7Oqym7adzVb3Bv3XeHwaf1MuiHdvdl2+AYNa7txNSKBY6GRDO5eP1t3lEImE7WSrITZxPD2228zZswYGjRogFarJTw8nGXLlrFy5coil+AW8k+hkEsrmAHUqe5a5HIYBjklhve/DWZgl3qM9W+S6/eisgakq1TMX3+zUHC1qzgzaWgLKro5SKW57bLWezBcxD1c7Jk4tAURMcnUq+GKo72S3cFhLFv/Lw/TVHRoWpVLt+K4kXXHH3pN30oIj04i/G4yjbzcuRIeT/T9hybPTRgSQkamhodpKmkCgkFOD14abN5/lSoejrw6uDkDu9ZjrH9T4pMyeHnhXrQ66PlUrWzf0T/gJloM1sBsYujevTt79uzh1KlTKJVK2rRpg6urKy1atMDJSdwpWpKDnYLk1KzEUEzdSIDU8nB2tGXV2z0YN38vWq2OP/6+Qe2qjxbTaVDLjWu3E0y+q1TIccoqGicUP5lMhn9WBdfHGWaDuTvb4VXNBa+sfxNVK1YgoHNdNuy7gk6nX0Y1Q6XmalZNIsMF/l7WQ41Du3sTHp3E3biHOT4LcS8+lRuRCQC8P74jDWu7s3LzGW5FJTF9VBtux6TwTeA56fM1KzvRrF5FJgxpIS1apFTI8HR3oFPzasQlpkmxGhNjDNbDbGLQarX89ttv/P3336jVajp37sykSZNEUigFdrZKklP14zrFmRhkMhmr3/bD0d7GpGsBkO4ye7avxTj/Zoz5cLfJ+27OdsXWchEKxsXJDlul3GQQ2sC7phuGm+9m9SqiUmtZu+s/4pPSpZaegVc1F6pXcuLPIzcBaN3AE5cKttI61HfjUqVxikZe7rg62fHBy52k7xu6hFwq2FK9UgXmT/TNNovO4J0X2uVa5qO8rHFdFpidlfTZZ5/xzz//MHbsWF566SXOnDnD0qVLLRGb8BgHOwV1qrnQ37cOPi0KV8o5N17VXPB0119ghvvVp01DTyo42BAWnQRAh6ZVpfIZ8CgxuRttEyxLIZcxf6IvQ3p4Z3vPu6YroB+b8q7hSrusCrBHQiNJephp0hKs4u7I5BEtpemsDWq70aV1Den9yNgUfgy6SEVXe1ydsv//ruLhSGMvd14KaMqn07rlmhQAbJSKHCdSgOEBt3z84kKJM9tiOHLkCL///js2Nvrugh49ejBo0CDee++9Eg9OMOVob4OjnZLXhuc937yoxgU0A2DS4v2EZyUGd2fTh9ga1HIjLDrJJFkIlmd4ivhxHi72uDnbUaeaCwqFnHo1XKnkas/32y4A0LRuRSLuJgNQ0dWeyh6OdGhWlWOhUTg52JpMQfZ0dyA2Po2alXPuJZDJZHw6rVuRfxe5ouS6kk5cuktsfBr+neuWyP6fNGYTg06nk5ICgK2trclrwXImDm1RLKW288ulgp304Ju7iz4BdG9Tk8Nn7kgDzhXE+IJVksn0a0YYalvJZDJe6N9EmlHWrnFldgeHAY9WAWzdwJNjoVHY2silUimj+zZmVJ9G/Hs5psRnnylKsCTGx2tCABjgW0d0feaD2cTQuHFjFi1axAsvvIBMJmPdunXS1NWiWrJkCfHx8SxevLhY9veka1Ar++yhkmTcGjD8PH10W94Y2Zr9JyIARAE9K/Z4a+Lpp2ohl8uoV8M1WzFEgD4dvbC3U9K5ZXVslHJ+XxwgdS+1a1ylxOMtrllJiSkZ7A0Jp0fbWni6O5Ce8WhafVh0El5VXcrdqoMFZfb2c968eSQmJjJy5EieeeYZ4uLieP/994t84ODgYLZu3Vrk/Qglx9Cf7OpkK/ULK+QybJQKyEoIIjGUHTKZDL92tfCq6oJLhexdgHK5jB5ta0rJwNZGYdG7a/0DbkXfz+HTd/h153/M/uooiSkZjJu/R3pv2meHCDp6s+gHecKZbTE4OTmxZMmSYj1oQkICX3zxBZMmTeLy5cvmvyCUClcn/V1lTi0VwwCjGGMomxRyGS/0b0zzeoVferS4KeQy1Bot05cfZoBPHanWU0HFZZVviU1I4+8zkTxMN30Q99KtOAZ38+bd1UfwaVGdId2zD94bux2TTEJyBi3qW8+5Kmm5JoaBAwfm+cWgoKBCH/SDDz5g+vTpREdHF3ofQslTZN0t5jTo2LV1DRKSM+jnU/RCfkLpeK5Xo9IOwYRcJuNhmorwu8msvH220InB8CyGVqvj8Ok7VKtYgZVv9+CZ2TsAsFUqyFRpuHTrAZduPWBId28u3oxDp9PluEb3a0sPAhD02eBC/mbZ7QsJp0OzqjnO8rIGuSaG4uguyslvv/1GtWrV8PHxITAwsESOIRQPw7Kfhvo6xhRymdk7LUEoCLlcRuLDTOm1TqcrVFeW8UN6VyLiGdi1Hva2Sj55rTOLfz3J/cQ07saZPssx68ujQPaLv8po0Sq1Rlsskz8iY1NYufksbRp6Mn+ib5H3VxJyTQwdOnQokQPu3LmT2NhYBg8eTGJiIqmpqSxatEhMf7VCA7vWo0pFR3yL+ZkJQciJXC4jKeVRmY2YB6lUrWi+QOPFm3HM+vIoP87tg6e7A3GJ6Tg72kpVYb1r6J/paO5diVb1Pbl2O8GkzLwxjUYrzdI6dz2WVKNuqAeJ6VT2KHr5F8P66ob4rJHZMYbi9tNPP0k/BwYGcuLECZEUrJSNUk7nltVLOwyhnHh8oZ6ImOR8JYaDp/RrUgSfj2JQN28eJKXRqLYHZ7NqQhk/k1HRzYF/LkQTGatfUU6pkJu0CqLjHlKzsjPpGWrmfH3c5Dj3E9OyJQaVWktquqpAXUKG8ueGciHWyHKT4gVBEPLw+BRSlSrvKUoXb8Zx9uo96ZmLO7EppKarSMvQSE9+w2OJwdWeTLWWK1llyTVaLbEJjxafCs966O/uA9MFqQDuJ2QvI/75//7lhXm7s61fkpfUrOmzIjHkYtiwYeIZBkEQgOxTnzPVjxYWCr0Wy/vfHic5NRO1Rn+XPuvLo7z/bTDqrDv+W5GJ0roitas+qiXmZnQ3XymrFtg/F/Qly3U6uGlUTnzxLye5fifBpJJs20b6ciKGLqtz1x+tYXE0a2lVQyXa/EjJqndml8/EkJ6ptngNKdFiEATBKigeazFkGrUY1u78j7NXYxn9/i7GzNvNc3N2Su8lZfXVXw6PZ8mvp6hWsYJUGwowGcBu1dBTqu/1XC/9g7rXsyoG92yvLwW+559wk8TQ3LsiDnZKdh4P4+LNOOZ8fVzqfrLNeubjYXr+Fy0zjDHkJzFotTqemb2Db7aeM/vZ4mTxMQZBEIScZOtKymox7A0JlxYNAv3qc8Zu3ElEIZcxY3RbZMho3cgTZ0dbKjjYZFtd0MnBhq/f7UnMg1RS01Vs2n9Vqhw7cWhLVCotx89F0am5fsKFUiGjZf1KnLh4l8vh8dJ+kh5mUNHVATtbBZlqLQ/TVNkqE+fGEL+Njfn7csP6K7uOh5V4jTRjIjEIgmAVDImhgr2Sh+lqMlVakh5msmrz2Ty/dyUinqoejnRrU9Nk+/r5/XP8fAUHG+rVcOXOPf14wo3IRBztlTjYKeneriZ/n41kb0g4DWq5seT1LtgoFUwf1ZZl6/+lXg1X9vwTTkKyITHoS+Gnpud/NcuUNH0LJz/dQ2pN6ZSbFV1JgiBYBUNXkmFlOpVaI3W7ALRtXJk3nmuT7XtarQ4H++z3uAq5LFv3lDHD2ENyaiauWSVC2jWuIj2r0K5xFX35F6C6pxOfv9kdv3b67qY1f17k1H8x2Nnk3JWUnqlm+cbTxCdnX/jI8DsZz4bKjVpTOutTiBaDIAhWwdBicHXSV/XNVGtN5vq7OdlRp3rOC1Q52hW8ym8FBxuUChlqjQ6XrPIvCrmMueM7cPZqLCN7Zy8WaigTc/7Gfc7fuE+96vrZT6lpj1oMe/4J52hoJGevxiKXyZj2WDIzDD7nLzGIFoMgCOXYnXv6Zwua16uErVLO+ev32XHslvS+e9b6EoZBYtAvRATkuThQbmQy2aNCkUZFBds1rsLLg5pLD7oZe/x5BcPYh6HFoNPpWP3bWc5e1c9cungzjr/+vW3ynZQCtRhEYhAEoRwb4FuHDk2r8myvhtgoFfwX9kB6eA3064MoFXLeHNlW2uZd0w0Axxy6kvLDuIJwflSwN22ZGGYWpaarOHHpbrZlU6PuP5QWRzIwdCUZT8fNTWklBtGVJAiCVejWpqY0gGybw4wd44fIRjzdgC0Hr1G7qjMXb8YVqsUAj6oD57Q+RU4enzmlUOhfX72dwE/bL+X4nZS0TDRanTTeYWhd5KvFkI/PlATRYhAEwerYKrPP8TesIggw1r8pQZ8NpmLWQLXhAl1QhgHonNanyA9N1syi2zHJuX5Gp4MUo7ESwwym/CQGjYUfbDMQiUEQBKvz+Bz/uS91kGYEGTOMDRTkyWNjbgXsSnqcYXA8Mmt8ZGiP+jTyyl6NOCmraqxOpyMtw5AYzMecn+RREkRiEATB6hieKDbo2LxajiW4DV1IaRn5f47A2KMxhvy3GEb1ebSOxYOsRYEMd/Yjnm4gPRxnzJAYMtVa6fkFMfgsCIJQADY5dCXlxDlrbMBGWbhLWUHHGABG923MBy93BCA++VGZcIVchpODTY4D4YbEkGr0vEOmmSKBABrxHIMgCIJeToPPOWndwJMXBzShb6c6hTpOh2ZVeaFfY2l2U37l1MJwdbJDLpfhaJ/9mYqkh/oEYmjZONgp89ViUIlZSYIgCHr5bTHI5TKe6Zn9QbT8cnKw4bneBV/iNK8WRl4thrSsgWeXCrZSssiL6EoSBEF4TJM6Hnz2RrfSDiMb4+mxhmqtCVmrzz3+rAMYJYasFoOrk20+u5IefaYgaz4UlUgMgiBYHcMAbdfWNXJcc7y0GZfMrl/LDXgUc54thgxDi8EOjVZndjqqWv3ofUu2HkqlK2n16tXs2rULgO7duzNz5szSCEMQBCul0eovgoV9cK2kGa++5uZkR6fmVenaugaANMbg6e6As4MtWp1Omr2UatSVBPopqwrb3H9H4zGGDJU2311sRWXxFsPx48c5evQoW7duZdu2bVy8eJF9+/ZZOgxBEKyY4U7aWhODXC6TqrAqFXLmvNRRemrb0GJo17gKK97qQY3KTtyL1y8V+qgrSd/9ZO7JZuOuJMNa0ZZg8cTg6enJrFmzsLW1xcbGBm9vb6KioiwdhiAIVswwTdPeznrXRTbMnFI+NlXWMSuZGZJaFXdHYhPSuBoRz5dbQgFwlVoMWn744wKzvjxKeHRStmOoSykxWDwdN2jQQPo5LCyMXbt2sWHDBkuHIQiCFTN0Jdnn0c1S2mxtFKSmq7Ot+aBQyBndpxHtmlQBoLK7Ayq1li0Hr0mfMTx/kanWsuPYTdQaHUdCI/GqZlpW3Hg9hsI+3V0YpXbWr127xsSJE5k5cyZ16tQprTAEQbBChq6k/D7PUBoM4wzKHMpzj+rbWPrZ08MRgODz0dm+m6nSSIPWV42WDjVQm4wxPMFdSQD//vsv48aN46233mLo0KGlEYIgCFbM0JWkkFtvYjCs3pZTYjBW2d0x2zZDyY/7CWkYJiZdvZ2AWqM1GVcora4ki5/16OhopkyZwrJly/D397f04QVBKAO8a+pXRnNyLPjKbJZiSFpKM5VdK7s7ZNtmGKD+4LtgAOpWd+FhmoqZq47w+YbT0ueMu5IeL/ldkizelbRmzRoyMjJYvHixtG3kyJGMGjXK0qEIgmClJg9vRT+fOjnebVsLQ6nvnFZ6M+Zob8PCyb4kPcykUW0P0jPVVKtUgcZe7lzO6j6qXsmJW1FJXLudYPIgm1qjRS6DBZM706SOR8n9Mo+xeGKYO3cuc+fOtfRhBUEoQ+xsFDT2styFsDAMg87mupIAWtb3zLatV4faUmLwyFpXAuBh+qNKsWq1FqVCTgvvSkUNt0CstwNPEATBiuW3Kyk3jnaPusk8XB4lhjTjxKDVZpsOawkiMQiCIBSCoc/fXFdSbhwdHnXYGCeGh0aludVqbakMwIvEIAiCUAiGriSbQrYYjIvtVTRKDCq1FpVaQ9CRm+w8HoaN0nKDzgYiMQiCIBRCUVsMDkbF9txcTNd3SE1XE3Ixukj7LwqRGARBEArBkBiUhezqMW4xONrZ4GBU/uNhuopbUfoSGZZ84tlAJAZBEIRCkGYlFbKrx7g8t4OdwmTlt6jYh1Kp7pTUzCJEWTgiMQiCIBSCXGZICIVLDMZ1oOxslSaJ4uLNOOlnM0s2lAiRGARBEArB0JWkLeTKasZPMtso5SYthku34nL6isWIxCAIglAIhq4kraZ4bukdjdaeuHTrAfa2pVdyXCQGQRCEQjA8X6ApprWYHR1M60JZsgTG40RiEARBKATDZCSttnjWYnZ8bLW6ejVci2W/hSESgyAIQiHUr+kGgGcxFfrzcLXH2dFWGnvwztp/abDe5ZEEQRCs2MCu9Whar6KUIArj1w/7SmtPDOtRH792tcjI1BCXmEabRpX5cW6fQk+HLQqRGARBEApBJpMVKSkAuDs/KoXhaG8jzUwydCN55rCWgyWIriRBEATBhEgMgiAIgolSSQxBQUEMGDCAPn36sH79+tIIQRAEQciFxccYYmJi+OKLLwgMDMTW1paRI0fSsWNH6tevb+lQBEEQhBxYvMVw/PhxOnXqhJubG46OjvTt25fdu3dbOgxBEAQhFxZPDPfu3cPT89H6p5UrVyYmJsbSYQiCIAi5sHhXklarRSZ7NC9Xp9OZvM6NRqOvSX737t0Si00QBOFJY7hmGq6h+WHxxFC1alVOnTolvY6NjaVy5cpmvxcbGwvA888/X2KxCYIgPKliY2Px8vLK12ctnhh8fX1ZtWoVDx48wMHBgb179/Lxxx+b/V7z5s1Zv349np6eKBSlV3VQEAShLNFoNMTGxtK8efN8f0em0xVTacACCAoK4ttvv0WlUjFixAheffVVS4cgCIIg5KJUEoMgCIJgvcSTz4IgCIIJkRgEQRAEEyIxCIIgCCZEYhAEQRBMiMQgCIIgmBCJQRAEQTBRLhKDtZX5HjNmDP7+/gwePJjBgwcTGhrK8ePHGThwIH369OGLL76weEwpKSkEBARw584dgFzj+e+//xg2bBh9+/Zlzpw5qNXqUolv9uzZ9OnTRzqH+/btK5X4Vq9ejb+/P/7+/ixduhSwrnOXU3zWcu4AVqxYwYABA/D39+enn34CrOf85RSbNZ07gCVLljBr1iygmM+b7gl39+5dnZ+fny4+Pl738OFD3cCBA3XXrl0rtXi0Wq2uS5cuOpVKJW1LS0vTde/eXRcREaFTqVS68ePH6w4dOmSxmM6ePasLCAjQNWvWTHf79u084/H399edOXNGp9PpdLNnz9atX7/e4vHpdDpdQECALiYmJttnLRnfsWPHdM8995wuIyNDl5mZqXvxxRd1QUFBVnPucopv7969VnHudDqdLiQkRDdy5EidSqXSpaWl6fz8/HT//fefVZy/nGK7ceOG1Zw7nU6nO378uK5jx466d999t9j/Zp/4FoO1lfm+efMmAOPHj2fQoEGsW7eOc+fO4eXlRa1atVAqlQwcONCiMW7evJl58+ZJNatyiycyMpL09HRat24NwLBhwywS5+PxpaWlERUVxXvvvcfAgQNZuXIlWq3W4vF5enoya9YsbG1tsbGxwdvbm7CwMKs5dznFFxUVZRXnDqBDhw78+uuvKJVK4uLi0Gg0JCUlWcX5yyk2e3t7qzl3CQkJfPHFF0yaNAko/r/ZJz4xWFuZ76SkJHx8fPjyyy/5+eef2bhxI1FRUaUa48KFC3nqqaek17mds8e3e3p6WiTOx+O7f/8+nTp1YtGiRWzevJlTp06xZcsWi8fXoEED6Q8uLCyMXbt2IZPJrObc5RRf165dreLcGdjY2LBy5Ur8/f3x8fGxqn97j8emVqut5tx98MEHTJ8+HRcXF6D4/2af+MRQ2DLfJaVNmzYsXboUZ2dnPDw8GDFiBCtXrrSqGHM7Z9ZyLmvVqsWXX35J5cqVcXBwYMyYMRw+fLjU4rt27Rrjx49n5syZ1KpVy+rOnXF89erVs6pzBzBt2jSCg4OJjo4mLCzMqs6fcWzBwcFWce5+++03qlWrho+Pj7StuP9mLV5d1dIKW+a7pJw6dQqVSiX9T9XpdNSoUUMqKw6lH2PVqlVzjOfx7ffv3y+VOK9cuUJYWBh9+/YF9OdQqVSWSnz//vsv06ZN47333sPf358TJ05Y1bl7PD5rOnc3btwgMzOTJk2a4ODgQJ8+fdi9e7dJ9eTSOn85xbZz507c3NxK/dzt3LmT2NhYBg8eTGJiIqmpqURGRhbreXviWwy+vr4EBwfz4MED0tLS2Lt3L926dSu1eJKTk1m6dCkZGRmkpKSwdetWZsyYwa1btwgPD0ej0bB9+/ZSjbFVq1Y5xlOjRg3s7Oz4999/Afjjjz9KJU6dTseiRYtITExEpVKxadMmevfubfH4oqOjmTJlCsuWLcPf3x+wrnOXU3zWcu4A7ty5w9y5c8nMzCQzM5MDBw4wcuRIqzh/OcXWvn17qzh3P/30E9u3b+ePP/5g2rRpPP300/zwww/Fet6e+BZDlSpVmD59Oi+++KJU5rtly5alFo+fnx+hoaEMGTIErVbL6NGjadOmDYsXL2bq1KlkZGTQvXt3+vXrV2ox2tnZ5RrPsmXLmDt3LikpKTRr1owXX3zR4vE1btyYCRMmMGrUKNRqNX369CEgIMDi8a1Zs4aMjAwWL14sbRs5cqTVnLvc4rOGcwfQvXt3zp07x5AhQ1AoFPTp0wd/f388PDxK/fzlFNvrr7+Ou7u7VZy7xxX336wouy0IgiCYeOK7kgRBEISCEYlBEARBMCESgyAIgmBCJAZBEATBhEgMgiAIggmRGIQyacGCBaVShbYsE+dMyC+RGIRsGjVqxIMHD0y27d69mzFjxpj97quvvsr169dLKjTJ5MmT2b17NykpKYXex4EDB1iwYEGen7lz5w5t2rTJ8b1Vq1Yxf/78Qh8fICQkRJoLXxLWrFkjlWXO7ZxptVp69erFjh07sn3/448/zvMczZo1izVr1hRv0EKpe+IfcBMs6/vvv7fIcSpWrMiePXuKtI+ePXvSs2fPYorI+uV2zuRyOSNHjmTLli3SE9IA6enpBAUFsXHjRkuGKVgBkRiEAlu1ahWRkZHExsYSGRlJlSpV+PTTT6lcuTJPP/00K1asoEWLFqxYsYKgoCDc3d156qmnuHDhAmvXrmXWrFk0aNCAl19+GcDkdUxMDPPnzyc6OhqVSoW/v79UWjgvY8aMoXXr1pw+fZro6Gh8fHz4+OOPkcvlnD59mmXLlpGWloZcLuf111/Hz8+PwMBA9uzZw7fffkt4eDjvvfceiYmJeHp6otPpGDRoEB06dECj0fDBBx9w/vx5kpOTeeedd6R6OTdu3OD5558nMTGRJk2aMG/ePJycnLh27Rrz588nISEBmUzG+PHjGTJkCCEhISxcuBBHR0cePnzIzJkzSU1NZfr06dy8eZOMjAwWLFjAU089RXJyMh999BGXL19GJpPRtWtXZsyYgVKp5NSpUyxdupS0tDRsbGx488036datGyqVigULFnD8+HEqVqxIxYoVcXZ2Nnv+hg8fzurVq4mMjKRGjRoA7Nq1i+bNm1OnTh0WLFhAaGgoDx8+RKfTsWDBAtq1a2eyj0aNGhEcHIyHh0e21wcPHuTrr79GpVJhb2/Pu+++m2tLTCh9IjEIhXLq1Cm2bduGk5MTkyZNYuPGjUybNk16f+/evezdu5dt27ZhZ2fHa6+9lq/9vvPOO4wbN46nn36ajIwMXn31VWrXrs2AAQPMfjciIoK1a9eSmppK//79OXHiBE2aNGH27NmsWbOGmjVrEhMTw7PPPkujRo1Mvjtz5kwGDx7M6NGjuXHjBsOHD2fQoEEAZGRk0LlzZ+bPn8++fftYsmSJlBgiIiL4/fffcXd355133uHrr79m+vTpTJ48mZkzZ9KnTx9iYmJ45pln8PLyAvTVTvfv30+NGjUICQnh7t27fPHFF7Rq1Yqff/6ZVatW8csvv7BgwQLc3NwICgpCpVIxefJkfvzxR5555hmmTZvG119/TatWrbh27RovvPACW7Zs4eDBg4SFhbFjxw7UajUvvPBCvhKDu7s7/fr1IzAwkKlTpwKwadMmXnnlFUJDQ7l37x6bNm1CLpfz3Xff8f3332dLDLkJCwvjiy++4Ndff8Xd3Z1r167x0ksvsXfvXhwdHfO1D8GyRGIQssmpLK9Wq0UufzQk1aFDB5ycnABo2rQpiYmJJp//559/6N27t/SZ5557jl9++SXP46ampnLy5EkSExNZsWKFtO3y5cv5Sgx+fn7I5XKcnJzw8vIiMTGRs2fPEhsby5QpU0x+vytXrkivExMTOXfuHOvWrQPA29ubTp06Se/b2NhIiaBx48bExcVJ7/Xu3Vu6Qx4+fDhLly5l6NChZGRk0KdPH0Bfr6tPnz4cOXKEjh07Uq1aNemuHPRlxFu1aiXt//fffwfg77//ZsOGDchkMmxtbRk5ciS//PILjRo1onbt2tJ3GjRoQNu2bTlx4gTBwcEEBARga2uLra0tAwcONPld8zJ69GjefPNNpkyZws2bN7l37x5+fn4oFApcXV3ZuHEjt2/fJiQkhAoVKuRrnwDHjh3j3r17jBs3Ttomk8mIiIigcePG+d6PYDkiMQjZuLu7k5CQIF3wAOLi4nBzc5Ne29vbSz/LZDIeL7llZ2dnss3GxibXz6tUKkCffHQ6HRs3bsTBwQGABw8eYGdnl6+4c4pJo9Hg7e3Nb7/9Jr0XExODh4cHQUFBAFK5YuOYjEsYPx67MePPabValEolGo0m2+d0Op201u7jd8m5nZvHa+lrtVrUarXZ/ecWnzktW7bE3d2d4OBgDh06xMiRI1EoFBw6dIiFCxfy0ksv0bNnT+rVq8eff/6Z574yMzNN4vbx8WH58uXStujo6FItLS/kTcxKErLp1q0ba9euRavVAvo76q1bt9K9e/d876NHjx7s3r2bxMREtFot27Ztk95zd3fnwoULgP4ifeLECQCcnJxo3bq1tPB6UlISo0aN4sCBA4X+XVq3bk14eDgnT54E9Auj9+3b12QVKycnJ9q2bUtgYCAAt2/fJjg4OF8Lmhw8eJDExEQ0Gg2bN2+mW7du1KtXD6VSyd69e6Xfcc+ePfj6+hYo9i5durBu3Tp0Oh2ZmZls3rwZX19fWrduzc2bNzl37hyg75o6efIkHTp0oGvXrmzbto2MjAwyMjLYuXNngY75/PPPs3XrVvbt28eIESMA/R2/n58fo0ePpnnz5uzfvx+NRpPtux4eHpw/fx6A7du3S9t9fHw4duwYN27cAODw4cMMGjSI9PT0AsUmWI5oMQjZzJkzh8WLFxMQECDdcQ4ePJihQ4fmex8dO3bkxRdfZPTo0djZ2Zl0nYwZM4a3336bvn37UrNmTZNum2XLlvHxxx8zcOBAMjMzCQgIkPr6C8PDw4OVK1dKa2DodDqWLl1KzZo1pYQEsGTJEubMmcP//vc/qlSpQs2aNU1aILnx9vZm4sSJJCUl0a5dOyZMmICNjQ1fffUVCxYsYNWqVWg0GqZMmUKnTp0ICQnJd+xz585lwYIFDBw4EJVKRdeuXZk0aRK2trasWLGCjz/+mPT0dGQyGZ988gl169aldu3aREREEBAQgJubmzSukV/+/v4sXbqU7t27Sy3GkSNH8tZbbzFw4EDUajWdO3dm79690o2Dcbzz58/HxcUFX19faUnJ+vXrM3/+fGbMmCEtbvP1118XqDtKsCxRdluwiN27d7N+/XrWrl1b2qHk6Ouvv6ZPnz54e3uTnJzMoEGD+P7776lfv35phyYIFidaDIIA1KlTh+nTpyOXy9FoNLz66qsiKQjllmgxCIIgCCbE4LMgCIJgQiQGQRAEwYRIDIIgCIIJkRgEQRAEEyIxCIIgCCZEYhAEQRBM/B/QF8qpvP0xzgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(grouped_by_neighborhood.index, np.log(grouped_by_neighborhood.values.tolist()));\n",
    "plt.xlabel(\"Unique `neighborhood` Value\");\n",
    "plt.ylabel(\"log(# Of Occurences)\");\n",
    "plt.title(\"Distribution Of Occurences Of Unique `neighborhood` Values\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Most `neighborhood` values are very small and under 50.\n",
    "- The higher the `neighborhood` value, the less it occurs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `coexpression`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXkAAAD7CAYAAACPDORaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAQf0lEQVR4nO3df0iVd//H8de5c5a1jRWcM4dJsVYUSWULso0pFWo5D65WoR20cqOIjGrDSJMcgi4qCvpWLCI2mI48c0Utli5qCKNgyz+MAxHRMjKXHbHVTqT547r/GB2+rXvznOPljn32fEDg9TkXXe/+8NnFdZ1zLodlWZYAAEb6T7QHAAAMHSIPAAYj8gBgMCIPAAYj8gBgsJhoD/BEV1eXfD6fnE6nRowYEe1xAOC50NfXJ7/fr6SkJI0aNeqZ14dN5H0+nzweT7THAIDnUk1NjebMmfPM+rCJvNPplPTHoPHx8VGeBgCeD3fu3JHH4wk29M+GTeSfXKKJj4/X+PHjozwNADxf/uoyNzdeAcBgRB4ADEbkAcBgRB4ADEbkAcBgRB4ADEbkAcBgRB7/SklJSXI4HEP6JykpKdr/TGD4fBgK+Cf5fL6w9nc4HOIhangecSYPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgsJC+1iAQCCg3N1efffaZrl+/rr179wZfa29v18yZM3X48GEdOHBA33zzjV5++WVJ0ooVK+TxeIZmcgDAgAaMfHNzs8rKytTS0iJJSktLU1pamiTJ7/crLy9PJSUlkv74PpC9e/cqOTl56CYGAIRswMs1Xq9X5eXlcrlcz7y2a9cu5ebmauLEiZL+iPzhw4fldrtVUVGh7u5u2wcGAIRuwMhXVlZqzpw5z6y3tLTop59+UkFBgSTp4cOHmjZtmoqLi3XixAk9ePBAhw4dsn9iAEDIIr7xWltbq5UrVyo2NlaSNGbMGB05ckSTJk1STEyMCgsL1djYaNugAIDwRRz5c+fOKSsrK7jd1tamurq64LZlWYqJ4evqASCaIop8Z2enurq6lJiYGFwbNWqUdu/erVu3bsmyLNXU1Cg9Pd22QQEA4YvoVLu1tVXx8fFPrY0bN04VFRVav369enp6NHv2bK1Zs8aWIQEAkQk58ufPnw/+PGPGDHm93mf2yczMVGZmpj2TAQAGjU+8AoDBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBiDwAGIzIA4DBQop8IBBQdna2WltbJUklJSXKyMhQTk6OcnJydPbsWUnSlStXtHTpUmVmZmr79u3q7e0duskBAAMaMPLNzc3Ky8tTS0tLcM3n86m6ulonT57UyZMngw/sLi4u1o4dO9TQ0CDLsv7nIwIBAP+cASPv9XpVXl4ul8slSXr06JHa2tpUWloqt9ut/fv3q7+/X7dv31ZXV5dmzZolSVq6dKnq6+uHdHgAwN8b8EHelZWVT213dHQoJSVF5eXleumll7Ru3TrV1dVp8uTJcjqdwf2cTqfa29vtnxgAELKwb7wmJibq4MGDcrlciouLU35+vhobG9Xf3y+HwxHcz7Ksp7YBAP+8sCN/9epVNTQ0BLcty1JMTIzi4+Pl9/uD6x0dHcFLPACA6Ag78pZlqaqqSvfv31dPT49qa2uVnp6uhIQEjRw5Uk1NTZKkkydPKjU11faBAQChG/Ca/J9NnTpVa9euVV5ennp7e5WRkaHs7GxJ0p49e1RWVqZAIKDp06eroKDA9oEBAKELOfLnz58P/uzxeOTxeJ7ZZ+rUqaqrq7NnMgDAoPGJVwAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwGJEHAIMReQAwWEiRDwQCys7OVmtrqySptrZW2dnZcrvdKikp0ePHjyVJBw4c0Pz585WTk6OcnBzV1NQM3eQAgAEN+Pi/5uZmlZWVqaWlRZJ048YNHT16VMePH9eYMWO0bds2ffXVV1q9erV8Pp/27t2r5OTkoZ4bABCCAc/kvV6vysvL5XK5JEmxsbEqLy/Xiy++KIfDoSlTpqitrU2S5PP5dPjwYbndblVUVKi7u3topwcA/K0BI19ZWak5c+YEtxMSEvT2229Lkjo7O1VTU6OFCxfq4cOHmjZtmoqLi3XixAk9ePBAhw4dGrrJAQADivjGa3t7u1atWqX3339fc+fO1ZgxY3TkyBFNmjRJMTExKiwsVGNjo52zAgDCFFHkr1+/rtzcXC1ZskQbNmyQJLW1tamuri64j2VZiokZ8JI/AGAIhR35QCCgDz74QJs2bVJhYWFwfdSoUdq9e7du3boly7JUU1Oj9PR0W4cFAIQn7FPturo6dXR06PPPP9fnn38uSVqwYIE2bdqkiooKrV+/Xj09PZo9e7bWrFlj+8AAgNA5LMuyoj2EJLW2tmrhwoU6d+6cxo8fH+1xgKc4HA4Nk18V4CkDtZNPvAKAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwYg8ABiMyAOAwUKKfCAQUHZ2tlpbWyVJFy5ckNvtVkZGhvbt2xfc78qVK1q6dKkyMzO1fft29fb2Ds3UAICQDBj55uZm5eXlqaWlRZLU1dWl0tJSHTp0SN999518Pp8aGxslScXFxdqxY4caGhpkWZa8Xu+QDg8A+HsDRt7r9aq8vFwul0uSdPnyZU2YMEGJiYmKiYmR2+1WfX29bt++ra6uLs2aNUuStHTpUtXX1w/p8ACAvxcz0A6VlZVPbd+9e1dOpzO47XK51N7e/sy60+lUe3u7jaMCAMIV9o3X/v5+ORyO4LZlWXI4HH+5DgCInrAjHx8fL7/fH9z2+/1yuVzPrHd0dAQv8QAAoiPsyM+cOVM3btzQzZs31dfXp9OnTys1NVUJCQkaOXKkmpqaJEknT55Uamqq7QMDAEI34DX5Pxs5cqR27typjRs3qru7W2lpaVq0aJEkac+ePSorK1MgEND06dNVUFBg+8AAgNCFHPnz588Hf543b55OnTr1zD5Tp05VXV2dPZMBAAaNT7wCgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMGIPAAYjMgDgMHCfvzfE19//bWqq6uD262trcrJydGjR4/U1NSkuLg4SVJRUZHS09MHPykAIGwRR3758uVavny5JOnatWvasGGDioqKtGrVKlVXV8vlctk2JAAgMhFH/v/75JNPtGXLFsXFxamtrU2lpaVqb29Xenq6ioqK9J//cFUIQ2fcuHG6d+/ekB/H4XAM+THGjh2rzs7OIT8O/j0GXd8LFy6oq6tLixcvVkdHh1JSUlRVVSWv16tLly7xYG8MuXv37smyLCP+/BP/WeHfZdCRP3bsmNasWSNJSkxM1MGDB+VyuRQXF6f8/Hw1NjYOekgAQGQGFfnHjx/r559/1oIFCyRJV69eVUNDQ/B1y7IUE2PLFSEAQAQGFfmrV69q4sSJGj16tKQ/ol5VVaX79++rp6dHtbW1vLMGAKJoUKfZt27dUnx8fHB76tSpWrt2rfLy8tTb26uMjAxlZ2cPekgAQGQGFfmsrCxlZWU9tebxeOTxeAY1FADAHry3EQAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGBEHgAMRuQBwGA8gBXPvTMbF+qXyvejPYYtzmxcGO0RYJhBRT4/P1+dnZ3Bh3VXVFTo4cOH+vTTT9Xd3a3Fixdry5YttgwK/JXF/3dOlmVFewxbTHI4ZO2P9hQwScSRtyxLLS0t+uGHH4KR7+rq0qJFi/Tll1/qtdde07p169TY2Ki0tDTbBgYAhC7iyP/yyy+SpMLCQv32229asWKFpkyZogkTJigxMVGS5Ha7VV9fT+QBIEoivvH64MEDzZs3TwcPHtQXX3yhY8eOqa2tTU6nM7iPy+VSe3u7LYMCAMIX8Zl8cnKykpOTg9vLli3T/v379eabbwbXLMuSw+EY3IQAgIhFfCZ/6dIlXbx4MbhtWZYSEhLk9/uDa36/Xy6Xa3ATAgAiFnHkf//9d+3atUvd3d0KBAI6ceKEPvroI924cUM3b95UX1+fTp8+rdTUVDvnBQCEIeLLNfPnz1dzc7Pee+899ff3a+XKlUpOTtbOnTu1ceNGdXd3Ky0tTYsWLbJzXgBAGAb1PvnNmzdr8+bNT63NmzdPp06dGsxfCwCwCV9rAAAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGI/IAYDAiDwAGG9SToQ4cOKAzZ85IktLS0rR161aVlJSoqalJcXFxkqSioiKlp6cPflIAQNgijvyFCxf0448/6sSJE3I4HPrwww919uxZ+Xw+VVdXy+Vy2TknACACEV+ucTqd2rZtm2JjY/XCCy9o0qRJamtrU1tbm0pLS+V2u7V//3719/fbOS8AIAwRR37y5MmaNWuWJKmlpUVnzpzRO++8o5SUFFVVVcnr9erSpUuqq6uza1YAQJgGfeP12rVrKiws1NatW/X666/r4MGDcrlciouLU35+vhobG+2YEwAQgUFFvqmpSatXr9bHH3+sJUuW6OrVq2poaAi+blmWYmIGdW8XADAIEUf+119/1YYNG7Rnzx69++67kv6IelVVle7fv6+enh7V1tbyzhoAiKKIT7OPHj2q7u5u7dy5M7iWm5urtWvXKi8vT729vcrIyFB2drYtgwIAwhdx5MvKylRWVvY/X/N4PBEPBACwD594BQCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDEXkAMBiRBwCDDUnkv/32W2VlZSkjI0M1NTVDcQgAQAgifvzfX2lvb9e+fft0/PhxxcbGKjc3V3PnztUbb7xh96EAAAOw/Uz+woULSklJ0SuvvKLRo0crMzNT9fX1dh8GABAC28/k7969K6fTGdx2uVy6fPmy3YcBnuJwOKI9gi3Gjh0b7RFgGNsj39/f/9QvnGVZxvwCYniyLCvaIwDDlu2Xa+Lj4+X3+4Pbfr9fLpfL7sMAAEJge+TfeustXbx4UZ2dnXr06JG+//57paam2n0YAEAIbL9c8+qrr2rLli0qKChQT0+Pli1bphkzZth9GABACGyPvCS53W653e6h+KsBAGHgE68AYDAiDwAGI/IAYLAhuSYfib6+PknSnTt3ojwJADw/njTzSUP/bNhE/sl76z0eT5QnAYDnj9/v14QJE55Zd1jD5OOCXV1d8vl8cjqdGjFiRLTHAYDnQl9fn/x+v5KSkjRq1KhnXh82kQcA2I8brwBgMCIPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgMCIPhCAQCCg7O1utra3RHgUIC5EHBtDc3Ky8vDy1tLREexQgbEQeGIDX61V5eTnPKsZzadh8QRkwXFVWVkZ7BCBinMkDgMGIPAAYjMgDgMGIPAAYjO+TBwCDcSYPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgMCIPAAYj8gBgsP8Cki35cahvIcwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.boxplot(protein_links_v1.coexpression, showfliers=False);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `cooccurence`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAD7CAYAAABnoJM0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAOtElEQVR4nO3dTWhU97/H8c+YxKgEbvi3ZxJMMHCRa6GpuIgPWBgpKU4SjdHBUmowbirWTVVooGoxBLUpIlj7oNBWECSKWZQMockoDrgopmBSKIqWouAiiZmZEHtpYtLJw7mL0rlM50puzuTf0X7fr93v/H6HfM/mvMnDqM91XVcAALMW5XoAAEBuEQIAMI4QAIBxhAAAjCMEAGBcfq4HmK/JyUndvXtXjuMoLy8v1+MAwHNvZmZGiURClZWVWrJkScb+CxeCu3fvqrGxMddjAMALp729XVVVVRnXX7gQOI4j6Y8HKi0tzfE0APD8Gx4eVmNjY+r9+VcvXAj+/HFQaWmpysvLczwNALw4nvXjdH5ZDADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYl1UIurq6VFdXp82bN6u9vT1j//79+wqFQgoGgzp69Kimp6fT9u/du6fKyspsRgAAZMlzCGKxmM6cOaPLly+rs7NTV69e1YMHD9LONDc369ixY7p27Zpc11VHR0dqb2JiQsePH9fU1JT36QEAWfMcglu3bmnDhg0qLi7WsmXLFAwGFYlEUvuDg4OanJzUmjVrJEmhUCht/5NPPtGePXu8Tw4AWBCeQxCPx+U4Tmrt9/sVi8Weue84Tmo/Go1qcnJSNTU1Xr88AGCB5Hu9cXZ2Vj6fL7V2XTdt/az9RCKh8+fP6+LFi16/NABgAXn+jqC0tFSJRCK1TiQS8vv9z9wfGRmR3+/XzZs39euvv6qxsVENDQ2SpIaGBo2NjXkdBQCQBc8h2Lhxo3p7ezU6OqqJiQldv35dgUAgtV9WVqbCwkL19/dLksLhsAKBgN566y3duHFD4XBY4XA4tVdUVJTlowAAvPAcgpKSEh06dEhNTU3avn27tm7dqtWrV2vv3r26c+eOJOn06dNqa2tTTU2Nnj59qqampgUbHACwMHyu67q5HmI+BgYGVF1drWg0qvLy8lyPAwDPvbnem3yyGACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxmUVgq6uLtXV1Wnz5s1qb2/P2L9//75CoZCCwaCOHj2q6elpSVJ/f7927typhoYG7dmzR4ODg9mMAQDIgucQxGIxnTlzRpcvX1ZnZ6euXr2qBw8epJ1pbm7WsWPHdO3aNbmuq46OjtT1EydOKBwOq76+XidOnMjuKQAAnnkOwa1bt7RhwwYVFxdr2bJlCgaDikQiqf3BwUFNTk5qzZo1kqRQKKRIJKJkMqkDBw7olVdekSStWrVKjx8/zu4pAACeeQ5BPB6X4ziptd/vVywWe+a+4ziKxWJavHixGhoaJEmzs7P64osv9Oabb3odAwCQJc8hmJ2dlc/nS61d101bz7WfTCb1wQcfaHp6Wvv27fM6BgAgS55DUFpaqkQikVonEgn5/f5n7o+MjKT2x8fH9e6772p6elrnz59XQUGB1zEAAFnyHIKNGzeqt7dXo6OjmpiY0PXr1xUIBFL7ZWVlKiwsVH9/vyQpHA6n9pubm1VRUaFPP/1UixcvzvIRAADZyPd6Y0lJiQ4dOqSmpiZNTU1p586dWr16tfbu3av3339fr732mk6fPq2PPvpIY2NjevXVV9XU1KR79+4pGo1q5cqV2rFjh6Q/fr/w9ddfL9hDAQD+/3yu67q5HmI+BgYGVF1drWg0qvLy8lyPAwDPvbnem3yyGACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADAuqxB0dXWprq5OmzdvVnt7e8b+/fv3FQqFFAwGdfToUU1PT0uShoaG1NjYqJqaGu3fv1/j4+PZjAEAyILnEMRiMZ05c0aXL19WZ2enrl69qgcPHqSdaW5u1rFjx3Tt2jW5rquOjg5JUmtrq3bt2qVIJKLKykqdO3cuu6cAAHjmOQS3bt3Shg0bVFxcrGXLlikYDCoSiaT2BwcHNTk5qTVr1kiSQqGQIpGIpqamdPv2bQWDwbTrAIDc8ByCeDwux3FSa7/fr1gs9sx9x3EUi8X05MkTFRUVKT8/P+06ACA38r3eODs7K5/Pl1q7rpu2ftb+X89JylgD/w6R99/Uf5X8R67HyNovsf9WzWc3cj0G/kE8h6C0tFR9fX2pdSKRkN/vT9tPJBKp9cjIiPx+v/71r3/pt99+08zMjPLy8jLuA/5d/ikvz//M9QD4x/H8o6GNGzeqt7dXo6OjmpiY0PXr1xUIBFL7ZWVlKiwsVH9/vyQpHA4rEAiooKBAVVVV6u7uliR1dnam3QcA+Ht5DkFJSYkOHTqkpqYmbd++XVu3btXq1au1d+9e3blzR5J0+vRptbW1qaamRk+fPlVTU5MkqaWlRR0dHaqrq1NfX58OHjy4IA8DAJg/n+u6bq6HmI+BgQFVV1crGo2qvLw81+MAwHNvrvcmnywGAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxnkMwNDSkxsZG1dTUaP/+/RofH884k0wm1dzcrNraWu3YsUMPHz6UJI2Pj+vAgQOqr69XfX29vvvuO+9PAADIiucQtLa2ateuXYpEIqqsrNS5c+cyzly6dElLly5VT0+Pjhw5osOHD0uSvvrqKy1fvlxdXV26ePGi2traNDIy4v0pAACeeQrB1NSUbt++rWAwKEkKhUKKRCIZ527evKlt27ZJktauXavR0VENDQ1p3bp12r17tyTppZdeUnFxMSEAgBzJ93LTkydPVFRUpPz8P253HEexWCzjXDwel+M4qbXjOBoeHtbrr7+eutbd3a1kMqmVK1d6GQUAkKU5Q9DT06O2tra0axUVFfL5fGnX/rqWJNd10667rqtFi/73m5Cenh59/PHH+uabb1JRAQD8veZ8+9bW1qq2tjbt2tTUlNavX6+ZmRnl5eUpkUjI7/dn3FtSUqJ4PK4VK1ZIkkZGRlLnLl26pAsXLujChQtatWrVQjwLAMADT78jKCgoUFVVlbq7uyVJnZ2dCgQCGec2bdqkcDgsSerr61NhYaGWL1+uGzdu6OLFi7py5QoRAIAc8/xXQy0tLero6FBdXZ36+vp08OBBSdKVK1d09uxZSdLu3buVTCa1ZcsWnTx5UqdOnZIkffbZZ/r999/13nvvqaGhQQ0NDbpz5072TwMAmDef67puroeYj4GBAVVXVysajaq8vDzX4wDAc2+u9yafLAYA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMIwQAYBwhAADjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACM8xyCoaEhNTY2qqamRvv379f4+HjGmWQyqebmZtXW1mrHjh16+PBh2v709LTefvttffvtt17HAABkyXMIWltbtWvXLkUiEVVWVurcuXMZZy5duqSlS5eqp6dHR44c0eHDh9P2v/zySz169MjrCACABeApBFNTU7p9+7aCwaAkKRQKKRKJZJy7efOmtm3bJklau3atRkdHNTQ0JEn68ccf9fPPP+uNN97wOjsAYAF4CsGTJ09UVFSk/Px8SZLjOIrFYhnn4vG4HMdJrR3H0fDwsMbGxtTW1qbjx497HBsAsFDy5zrQ09Ojtra2tGsVFRXy+Xxp1/66liTXddOuu66rRYsWqbW1Vfv27dPLL7/sdW4AwAKZMwS1tbWqra1NuzY1NaX169drZmZGeXl5SiQS8vv9GfeWlJQoHo9rxYoVkqSRkRE5jqPe3l798ssv+vzzz/X48WP98MMPys/PT/0YCQDw95kzBP+XgoICVVVVqbu7W/X19ers7FQgEMg4t2nTJoXDYVVVVamvr0+FhYUqKyvT999/nzrz4Ycfat26dUQAAHLE818NtbS0qKOjQ3V1derr69PBgwclSVeuXNHZs2clSbt371YymdSWLVt08uRJnTp1akGGBgAsHJ/rum6uh5iPgYEBVVdXKxqNqry8PNfjAMBzb673Jp8sBgDjCAEAGEcIAMA4QgAAxhECADCOEACAcYQAAIwjBABgHCEAAOMIAQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAwjhAAgHGEAACMy8/1APM1MzMjSRoeHs7xJADwYvjzffnn+/OvXrgQJBIJSVJjY2OOJwGAF0sikVBFRUXGdZ/rum4O5vFscnJSd+/eleM4ysvLy/U4APDcm5mZUSKRUGVlpZYsWZKx/8KFAACwsPhlMQAYRwgAwDhCAADGEQIAMI4QAIBxhAAAjCMEAGAcIQAWyNjYmLZu3aqBgYFcjwLMCyEAFsBPP/2kd955R48ePcr1KMC8EQJgAXR0dKilpUV+vz/XowDz9sL9o3PA8+jkyZO5HgHwjO8IAMA4QgAAxhECADCOEACAcfx/BABgHN8RAIBxhAAAjCMEAGAcIQAA4wgBABhHCADAOEIAAMYRAgAw7n8AJ2YdoYJtmTsAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.boxplot(protein_links_v1.cooccurence, showfliers=False);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAD7CAYAAACCEpQdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPGElEQVR4nO3db0id9f/H8deVx7nFCLY6F2eoCMUg0FBo3RiLY2GoaztEpzttgkQM2giJEY6hogwmOZCE3y/sVnRjdOdELUPckXAhG0Z9542NwwyCn0radjzS+uO2c5jH63cjdsivbXbOcTvq+/kAKT/nkvP+3PC5i+uc4+V4nucJAGDKY4UeAADw6BF/ADCI+AOAQcQfAAwi/gBgkK/QA6wmmUwqFovJ7/erqKio0OMAwIaQTqeVSCRUVVWlrVu3rnh83cc/Foupqamp0GMAwIb02Wefac+ePSvW1338/X6/pL82EAgECjwNAGwMN27cUFNTU6ah/23dx//epZ5AIKCysrICTwMAG8v9Lpfzgi8AGET8AcAg4g8ABv2r+C8sLOjgwYOamZmRJI2NjSkUCqm+vl59fX2Z4yYmJhQOh9XQ0KD29nYtLi5Kkn755Rc1NTWpsbFRx44d061btx7CVgAA/9aq8b9y5YoOHTqkqakpSX+9776trU39/f0aGhpSLBbT6OioJKm1tVWdnZ0aHh6W53mKRCKSpFOnTunw4cOKRqOqqqpSf3//w9sR8Ig4jrPiC9goVo1/JBJRV1eXXNeVJF29elUVFRUqLy+Xz+dTKBRSNBrV7OysksmkampqJEnhcFjRaFR3797Vf/7zHzU0NCxbBzay+4WefwCwUaz6Vs/u7u5l38/NzS1736jruorH4yvW/X6/4vG4bt68qe3bt8vn8y1bBwAUTtYv+C4tLS07u/E8T47j3Hf93n//jrMjACisrOMfCASUSCQy3ycSCbmuu2J9fn5erutq586d+vPPP5VOp5cdDwAonKzjX11drcnJSU1PTyudTmtwcFDBYFClpaUqKSnR+Pi4JGlgYEDBYFDFxcXas2ePhoaGJElfffWVgsHg2u4CAJCVrP+8Q0lJiXp6etTS0qJUKqXa2lo1NjZKknp7e9XR0aGFhQVVVlaqublZktTV1aWTJ0/q448/1q5du/Thhx+u7S4AAFlx1vsN3GdmZlRXV6eRkRH+tg/WjQe9brXOf6VgxGrt5BO+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMIv4AYBDxBwCDiD8AGET8AcAg4g8ABhF/ADCI+AOAQcQfAAwi/gBgEPEHAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMIv4AYBDxBwCDiD8AGET8AcAg4g8ABuUV/4GBAR04cEAHDhzQmTNnJEljY2MKhUKqr69XX19f5tiJiQmFw2E1NDSovb1di4uL+U0OAMhZzvG/c+eOuru7dfbsWQ0MDOjy5cu6cOGC2tra1N/fr6GhIcViMY2OjkqSWltb1dnZqeHhYXmep0gksmabAABkJ+f4p9NpLS0t6c6dO1pcXNTi4qK2b9+uiooKlZeXy+fzKRQKKRqNanZ2VslkUjU1NZKkcDisaDS6VnsAAGTJl+sPbt++Xe+9957279+vbdu26YUXXtDc3Jz8fn/mGNd1FY/HV6z7/X7F4/H8JgcA5CznM/8ff/xRX3zxhb799ltdvHhRjz32mKampuQ4TuYYz/PkOI6Wlpb+cR0AUBg5x//SpUvau3evnnzySW3ZskXhcFjff/+9EolE5phEIiHXdRUIBJatz8/Py3Xd/CYHAOQs5/g/++yzGhsb0+3bt+V5ni5cuKDq6mpNTk5qenpa6XRag4ODCgaDKi0tVUlJicbHxyX99S6hYDC4ZpsAAGQn52v+L774oq5du6ZwOKzi4mI999xzamlp0b59+9TS0qJUKqXa2lo1NjZKknp7e9XR0aGFhQVVVlaqubl5zTYBAMiO43meV+ghHmRmZkZ1dXUaGRlRWVlZoccBJOmBr1mt818pGLFaO/mELwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMIv4AYBDxBwCDiD8AGET8AcAg4g8ABhF/ADCI+AOAQcQfAAwi/gBgEPEHAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMyiv+Fy5cUDgc1v79+3X69GlJ0tjYmEKhkOrr69XX15c5dmJiQuFwWA0NDWpvb9fi4mJ+kwMAcpZz/H/++Wd1dXWpv79fX3/9ta5du6bR0VG1tbWpv79fQ0NDisViGh0dlSS1traqs7NTw8PD8jxPkUhkzTYBAMhOzvH/5ptv9OqrryoQCKi4uFh9fX3atm2bKioqVF5eLp/Pp1AopGg0qtnZWSWTSdXU1EiSwuGwotHoWu0BAJAlX64/OD09reLiYh09elTXr1/XSy+9pN27d8vv92eOcV1X8Xhcc3Nzy9b9fr/i8Xh+kwMAcpZz/NPptC5fvqyzZ8/q8ccf17Fjx7R161Y5jpM5xvM8OY6jpaWlf1wHABRGzvF/6qmntHfvXu3cuVOS9MorrygajaqoqChzTCKRkOu6CgQCSiQSmfX5+Xm5rpvH2ACAfOR8zf/ll1/WpUuX9McffyidTuvixYtqbGzU5OSkpqenlU6nNTg4qGAwqNLSUpWUlGh8fFySNDAwoGAwuGabAABkJ+cz/+rqah05ckSHDx/W3bt3tW/fPh06dEhPP/20WlpalEqlVFtbq8bGRklSb2+vOjo6tLCwoMrKSjU3N6/ZJgAA2XE8z/MKPcSDzMzMqK6uTiMjIyorKyv0OIAkPfA1q3X+KwUjVmsnn/AFAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMIv4AYBDxB/6mqqpKjuOs+vUgq/1sVVXVI9oNcH8538YR2Ixisdi/Oo47eWGj48wfyMH9Ak/4sVFw5g/k6F7oHcch+thwOPMHAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEHwAMIv4AYBDxBwCDiD8AGLQm8T9z5oxOnjwpSRobG1MoFFJ9fb36+voyx0xMTCgcDquhoUHt7e1aXFxci6cGAOQg7/h/9913OnfunCQpmUyqra1N/f39GhoaUiwW0+joqCSptbVVnZ2dGh4elud5ikQi+T41ACBHecX/t99+U19fn44ePSpJunr1qioqKlReXi6fz6dQKKRoNKrZ2Vklk0nV1NRIksLhsKLRaN7DAwByk1f8Ozs7dfz4cT3xxBOSpLm5Ofn9/szjrusqHo+vWPf7/YrH4/k8NQAgDznH//PPP9euXbu0d+/ezNrS0pIcx8l873meHMe57zoAoDB8uf7g0NCQEomEXnvtNf3++++6ffu2ZmdnVVRUlDkmkUjIdV0FAgElEonM+vz8vFzXzW9yAEDOco7/p59+mvn/L7/8Uj/88INOnTql+vp6TU9Pq6ysTIODg3rjjTdUWlqqkpISjY+P6/nnn9fAwICCweCabAAAkL2c4/9PSkpK1NPTo5aWFqVSKdXW1qqxsVGS1Nvbq46ODi0sLKiyslLNzc1r+dQAgCw4nud5hR7iQWZmZlRXV6eRkRGVlZUVehxgBcdxtM5/jWDQau3kE74AYBDxBwCDiD8AGET8AcAg4g8ABhF/ADCI+AOAQcQfAAwi/gBgEPEHAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBa3ozF2A92blzp27evPlInuth35N6x44d+vXXXx/qc8AW4o9N6+bNm5vmJisP+x8X2MNlHwAwiPgDgEHEHwAMIv4AYBDxBwCDiD8AGET8AcAg4g8ABhF/ADCI+AOAQcQfAAwi/gBgEPEHAIOIPwAYRPwBwCDiDwAGEX8AMIj4A4BBxB8ADCL+AGBQXjdw/+ijj3T+/HlJUm1trU6cOKGxsTF98MEHSqVS2r9/v44fPy5JmpiYUHt7u27duqU9e/bo1KlT8vm4fzwenvMtdfq/7jcKPcaaON9SV+gRsMnkXN+xsTFdunRJ586dk+M4OnLkiAYHB9Xb26uzZ89q165deueddzQ6Oqra2lq1trbq9OnTqqmpUVtbmyKRiA4fPryWewGW2f+/I/I8r9BjrIlnHEfe/xR6CmwmOV/28fv9OnnypLZs2aLi4mI988wzmpqaUkVFhcrLy+Xz+RQKhRSNRjU7O6tkMqmamhpJUjgcVjQaXas9AACylHP8d+/enYn51NSUzp8/L8dx5Pf7M8e4rqt4PK65ubll636/X/F4PPepAQB5yfsF359++klvv/22Tpw4ofLycjmOk3nM8zw5jqOlpaV/XAcAFEZe8R8fH9dbb72l999/X6+//roCgYASiUTm8UQiIdd1V6zPz8/Ldd18nhoAkIec43/9+nW9++676u3t1YEDByRJ1dXVmpyc1PT0tNLptAYHBxUMBlVaWqqSkhKNj49LkgYGBhQMBtdmBwCArOX8bp9PPvlEqVRKPT09mbU333xTPT09amlpUSqVUm1trRobGyVJvb296ujo0MLCgiorK9Xc3Jz/9ACAnDjeOn8v3MzMjOrq6jQyMqKysrJCj4MNxHGcTfNWz820Fzwaq7WTT/gCgEF8xBab2mZ5V9mOHTsKPQI2GeKPTetRXSbhkgw2Ii77AIBBxB8ADCL+AGAQ8Qf+pqqqSo7jZPUlKavjq6qqCrxLgBd8gWVisVihRwAeCc78AcAg4g8ABhF/ADCI+AOAQcQfAAwi/gBgEPEHAIPW/fv80+m0JOnGjRsFngQANo57zbzX0P+27uN/796/TU1NBZ4EADaeRCKhioqKFevr/k5eyWRSsVhMfr9fRUVFhR4HADaEdDqtRCKhqqoqbd26dcXj6z7+AIC1xwu+AGAQ8QcAg4g/ABhE/AHAIOIPAAYRfwAwiPgDgEHEH8jDwsKCDh48qJmZmUKPAmSF+AM5unLlig4dOqSpqalCjwJkjfgDOYpEIurq6pLruoUeBcjauv/DbsB61d3dXegRgJxx5g8ABhF/ADCI+AOAQcQfAAzi7/kDgEGc+QOAQcQfAAwi/gBgEPEHAIOIPwAYRPwBwCDiDwAGEX8AMOj/ASsBT3lPoyfDAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.boxplot(protein_links_v1.combined_score, showfliers=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c78b81650a0bd32063743affb6953ff71b1a0dba806fbca9e2db842718495748"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4 ('master-thesis-log')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
