{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/cwoest/Documents/Academics/Data_Science_UP/master_thesis/material/GNN-material/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_theme(style=\"white\")\n",
    "\n",
    "import sys\n",
    "import os\n",
    "# Make sure everything is on the path.\n",
    "sys.path.insert(1, os.path.join(sys.path[0], '../src'))\n",
    "\n",
    "from src.preprocess.build_drug_response_matrix import (\n",
    "    get_gdsc_gene_expression,\n",
    "    cosmic_ids_to_cell_line_names\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conda 4.10.3\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%conda --version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we are going to create all the datasets.\n",
    "\n",
    "Datasets used for the __drug response matrix__ (`../data/processed/drm_full.pkl`):\n",
    "- `../data/raw/GDSC2_fitted_dose_response_25Feb20.xlsx`\n",
    "- `../data/raw/GDSC1_fitted_dose_response_25Feb20.xlsx`\n",
    "- `../data/raw/GDSC1_public_raw_data_25Feb20.csv`\n",
    "- `../data/raw/GDSC2_public_raw_data_25Feb20.csv`\n",
    "\n",
    "Datasets used for the __gene expression dataset__ (`../data/processed/gexpr_full.pkl`):\n",
    "- `../data/raw/Cell_line_RMA_proc_basalExp.txt`\n",
    "- `../data/raw/Cell_Lines_Details.xlsx`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "Datasets used for the __copy number picnic dataset__ (`../data/processed/cnvp_full.pkl`): \n",
    "- `../data/raw/cnv_abs_copy_number_picnic_20191101.csv`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "Datasets used for the __copy number gistic dataset__ (`../data/processed/cnvg_full.pkl`): \n",
    "- `../data/raw/cnv_gistic_20191101.csv`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n",
    "\n",
    "\n",
    "Datasets used for the __mutation dataset__ (`../data/processed/mut_full.pkl`): \n",
    "- `../data/raw/mutations_all_20220315.zip`\n",
    "- `../data/raw/landmark_genes.csv`\n",
    "- `../data/processed/drm_full.pkl`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW_DATA_PATH = '../../data/raw/'\n",
    "PROCESSED_DATA_PATH = '../../data/processed/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Drug response matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IC50's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `GDSC1_fitted_dose_response_25Feb20.xlsx` took 84.06795 seconds to import. It has shape (310904, 19)\n",
      "File `GDSC2_fitted_dose_response_25Feb20.xlsx` took 36.50131 seconds to import. It has shape (135242, 19)\n"
     ]
    }
   ],
   "source": [
    "GDSC1_IC50_FILE = 'GDSC1_fitted_dose_response_25Feb20.xlsx'\n",
    "GDSC2_IC50_FILE = 'GDSC2_fitted_dose_response_25Feb20.xlsx'\n",
    "\n",
    "# Read the IC50 files.\n",
    "# GDSC1\n",
    "start = time.time()\n",
    "gdsc1_ic50s = pd.read_excel(f'{RAW_DATA_PATH}{GDSC1_IC50_FILE}', header=0)\n",
    "print(f\"File `{GDSC1_IC50_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc1_ic50s.shape}\")\n",
    "\n",
    "# GDSC2\n",
    "start = time.time()\n",
    "gdsc2_ic50s = pd.read_excel(f'{RAW_DATA_PATH}{GDSC2_IC50_FILE}', header=0)\n",
    "print(f\"File `{GDSC2_IC50_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc2_ic50s.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12974350</td>\n",
       "      <td>683665</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>MM</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975300</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975647</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0   GDSC1             281       12974350     683665         MC-CAR   \n",
       "1   GDSC1             281       12975300     684055            ES3   \n",
       "2   GDSC1             281       12975647     684057            ES5   \n",
       "\n",
       "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID  DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0       SIDM00636            MM        1  Erlotinib            EGFR   \n",
       "1       SIDM00265  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "2       SIDM00263  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "\n",
       "     PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
       "0  EGFR signaling        1045          Y  0.007813       2.0  2.395685   \n",
       "1  EGFR signaling        1045          Y  0.007813       2.0  3.140923   \n",
       "2  EGFR signaling        1045          Y  0.007813       2.0  3.968757   \n",
       "\n",
       "        AUC      RMSE   Z_SCORE  \n",
       "0  0.982114  0.022521 -0.189576  \n",
       "1  0.984816  0.031840  0.508635  \n",
       "2  0.985693  0.026052  1.284229  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc1_ic50s.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | DATASET   |   NLME_RESULT_ID |   NLME_CURVE_ID |   COSMIC_ID | CELL_LINE_NAME   | SANGER_MODEL_ID   | TCGA_DESC    |   DRUG_ID | DRUG_NAME   | PUTATIVE_TARGET   | PATHWAY_NAME   |   COMPANY_ID | WEBRELEASE   |   MIN_CONC |   MAX_CONC |   LN_IC50 |      AUC |     RMSE |   Z_SCORE |\n",
      "|---:|:----------|-----------------:|----------------:|------------:|:-----------------|:------------------|:-------------|----------:|:------------|:------------------|:---------------|-------------:|:-------------|-----------:|-----------:|----------:|---------:|---------:|----------:|\n",
      "|  0 | GDSC1     |              281 |        12974350 |      683665 | MC-CAR           | SIDM00636         | MM           |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   2.39568 | 0.982114 | 0.022521 | -0.189576 |\n",
      "|  1 | GDSC1     |              281 |        12975300 |      684055 | ES3              | SIDM00265         | UNCLASSIFIED |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   3.14092 | 0.984816 | 0.03184  |  0.508635 |\n",
      "|  2 | GDSC1     |              281 |        12975647 |      684057 | ES5              | SIDM00263         | UNCLASSIFIED |         1 | Erlotinib   | EGFR              | EGFR signaling |         1045 | Y            |   0.007813 |          2 |   3.96876 | 0.985693 | 0.026052 |  1.28423  |\n",
      "\n",
      "|    | DATASET   |   NLME_RESULT_ID |   NLME_CURVE_ID |   COSMIC_ID | CELL_LINE_NAME   | SANGER_MODEL_ID   | TCGA_DESC   |   DRUG_ID | DRUG_NAME    | PUTATIVE_TARGET   | PATHWAY_NAME    |   COMPANY_ID | WEBRELEASE   |   MIN_CONC |   MAX_CONC |   LN_IC50 |      AUC |     RMSE |   Z_SCORE |\n",
      "|---:|:----------|-----------------:|----------------:|------------:|:-----------------|:------------------|:------------|----------:|:-------------|:------------------|:----------------|-------------:|:-------------|-----------:|-----------:|----------:|---------:|---------:|----------:|\n",
      "|  0 | GDSC2     |              282 |        13320532 |      749709 | HCC1954          | SIDM00872         | BRCA        |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |    9.8e-05 |        0.1 | -0.251083 | 0.975203 | 0.112373 |  0.906631 |\n",
      "|  1 | GDSC2     |              282 |        13320565 |      749710 | HCC1143          | SIDM00866         | BRCA        |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |    9.8e-05 |        0.1 |  1.34332  | 0.978464 | 0.067577 |  1.68357  |\n",
      "|  2 | GDSC2     |              282 |        13320598 |      749711 | HCC1187          | SIDM00885         | BRCA        |      1003 | Camptothecin | TOP1              | DNA replication |         1046 | Y            |    9.8e-05 |        0.1 |  1.73699  | 0.9951   | 0.045205 |  1.8754   |\n"
     ]
    }
   ],
   "source": [
    "print(gdsc1_ic50s.head(3).to_markdown())\n",
    "print()\n",
    "print(gdsc2_ic50s.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 19)\n"
     ]
    }
   ],
   "source": [
    "# Join both datasets for analysis purposes.\n",
    "gdsc_ic50s_join = pd.concat([gdsc1_ic50s, gdsc2_ic50s], ignore_index=True)\n",
    "print(gdsc_ic50s_join.shape)\n",
    "assert gdsc_ic50s_join[gdsc_ic50s_join.index.duplicated()].shape[0] == 0\n",
    "assert gdsc_ic50s_join.shape[0] == gdsc1_ic50s.shape[0] + gdsc2_ic50s.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/kv/l0jzxgbj1kggff_kd35bfzkw0000gn/T/ipykernel_10142/1587790020.py:7: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gdsc1_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC1_RAW_FILE}', header=0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `GDSC1_public_raw_data_25Feb20.csv` took 14.88413 seconds to import. It has shape (5837703, 18)\n",
      "File `GDSC2_public_raw_data_25Feb20.csv` took 15.84662 seconds to import. It has shape (6646430, 18)\n"
     ]
    }
   ],
   "source": [
    "GDSC1_RAW_FILE = 'GDSC1_public_raw_data_25Feb20.csv'\n",
    "GDSC2_RAW_FILE = 'GDSC2_public_raw_data_25Feb20.csv'\n",
    "\n",
    "# Read the raw files.\n",
    "# GDSC1\n",
    "start = time.time()\n",
    "gdsc1_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC1_RAW_FILE}', header=0)\n",
    "print(f\"File `{GDSC1_RAW_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc1_raw.shape}\")\n",
    "\n",
    "# GDSC2\n",
    "start = time.time()\n",
    "gdsc2_raw = pd.read_csv(f'{RAW_DATA_PATH}{GDSC2_RAW_FILE}', header=0)\n",
    "print(f\"File `{GDSC2_RAW_FILE}` took {time.time()-start:.5f} seconds to import. It has shape {gdsc2_raw.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | RESEARCH_PROJECT   |   BARCODE |   SCAN_ID | DATE_CREATED         |   SCAN_DATE |   CELL_ID |   MASTER_CELL_ID |   COSMIC_ID | CELL_LINE_NAME   |   SEEDING_DENSITY | DRUGSET_ID   | ASSAY   |   DURATION |   POSITION | TAG     |   DRUG_ID |   CONC |   INTENSITY |\n",
      "|---:|:-------------------|----------:|----------:|:---------------------|------------:|----------:|-----------------:|------------:|:-----------------|------------------:|:-------------|:--------|-----------:|-----------:|:--------|----------:|-------:|------------:|\n",
      "|  0 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          1 | B       |       nan |    nan |       26022 |\n",
      "|  1 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          2 | B       |       nan |    nan |       20491 |\n",
      "|  2 | Sanger_GDSC1       |    100541 |      1765 | 2010-04-18T23:00:00Z |         nan |      2415 |              365 |      924238 | K5               |               250 | 505_a_5      | a       |          3 |          3 | UN-USED |       nan |    nan |      181946 |\n",
      "\n",
      "|    | RESEARCH_PROJECT   |   BARCODE |   SCAN_ID | DATE_CREATED         | SCAN_DATE            |   CELL_ID |   MASTER_CELL_ID |   COSMIC_ID | CELL_LINE_NAME   |   SEEDING_DENSITY |   DRUGSET_ID | ASSAY   |   DURATION |   POSITION | TAG     |   DRUG_ID |   CONC |   INTENSITY |\n",
      "|---:|:-------------------|----------:|----------:|:---------------------|:---------------------|----------:|-----------------:|------------:|:-----------------|------------------:|-------------:|:--------|-----------:|-----------:|:--------|----------:|-------:|------------:|\n",
      "|  0 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          1 | UN-USED |       nan |    nan |           0 |\n",
      "|  1 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          2 | UN-USED |       nan |    nan |           0 |\n",
      "|  2 | GDSC_SA            |      3230 |      2945 | 2015-02-13T00:00:00Z | 2015-02-17T00:00:00Z |      4712 |              198 |      753608 | PC-14            |               250 |          159 | Glo     |          4 |          3 | UN-USED |       nan |    nan |         398 |\n"
     ]
    }
   ],
   "source": [
    "print(gdsc1_raw.head(3).to_markdown())\n",
    "print()\n",
    "print(gdsc2_raw.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12484133, 18)\n"
     ]
    }
   ],
   "source": [
    "# Join both datasets for analysis purposes.\n",
    "gdsc_raw_join = pd.concat([gdsc1_raw, gdsc2_raw], ignore_index=True)\n",
    "print(gdsc_raw_join.shape)\n",
    "assert gdsc_raw_join[gdsc_raw_join.index.duplicated()].shape[0] == 0\n",
    "assert gdsc_raw_join.shape[0] == gdsc1_raw.shape[0] + gdsc2_raw.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 19)\n",
      "Index(['DATASET', 'NLME_RESULT_ID', 'NLME_CURVE_ID', 'COSMIC_ID',\n",
      "       'CELL_LINE_NAME', 'SANGER_MODEL_ID', 'TCGA_DESC', 'DRUG_ID',\n",
      "       'DRUG_NAME', 'PUTATIVE_TARGET', 'PATHWAY_NAME', 'COMPANY_ID',\n",
      "       'WEBRELEASE', 'MIN_CONC', 'MAX_CONC', 'LN_IC50', 'AUC', 'RMSE',\n",
      "       'Z_SCORE'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>NLME_RESULT_ID</th>\n",
       "      <th>NLME_CURVE_ID</th>\n",
       "      <th>COSMIC_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>TCGA_DESC</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PUTATIVE_TARGET</th>\n",
       "      <th>PATHWAY_NAME</th>\n",
       "      <th>COMPANY_ID</th>\n",
       "      <th>WEBRELEASE</th>\n",
       "      <th>MIN_CONC</th>\n",
       "      <th>MAX_CONC</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12974350</td>\n",
       "      <td>683665</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>MM</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975300</td>\n",
       "      <td>684055</td>\n",
       "      <td>ES3</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>281</td>\n",
       "      <td>12975647</td>\n",
       "      <td>684057</td>\n",
       "      <td>ES5</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>UNCLASSIFIED</td>\n",
       "      <td>1</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR signaling</td>\n",
       "      <td>1045</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.007813</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET  NLME_RESULT_ID  NLME_CURVE_ID  COSMIC_ID CELL_LINE_NAME  \\\n",
       "0   GDSC1             281       12974350     683665         MC-CAR   \n",
       "1   GDSC1             281       12975300     684055            ES3   \n",
       "2   GDSC1             281       12975647     684057            ES5   \n",
       "\n",
       "  SANGER_MODEL_ID     TCGA_DESC  DRUG_ID  DRUG_NAME PUTATIVE_TARGET  \\\n",
       "0       SIDM00636            MM        1  Erlotinib            EGFR   \n",
       "1       SIDM00265  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "2       SIDM00263  UNCLASSIFIED        1  Erlotinib            EGFR   \n",
       "\n",
       "     PATHWAY_NAME  COMPANY_ID WEBRELEASE  MIN_CONC  MAX_CONC   LN_IC50  \\\n",
       "0  EGFR signaling        1045          Y  0.007813       2.0  2.395685   \n",
       "1  EGFR signaling        1045          Y  0.007813       2.0  3.140923   \n",
       "2  EGFR signaling        1045          Y  0.007813       2.0  3.968757   \n",
       "\n",
       "        AUC      RMSE   Z_SCORE  \n",
       "0  0.982114  0.022521 -0.189576  \n",
       "1  0.984816  0.031840  0.508635  \n",
       "2  0.985693  0.026052  1.284229  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(gdsc_ic50s_join.shape)\n",
    "print(gdsc_ic50s_join.columns)\n",
    "gdsc_ic50s_join.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 9)\n",
      "Missing rates:\n",
      "==============\n",
      "DATASET            0.0\n",
      "CELL_LINE_NAME     0.0\n",
      "DRUG_NAME          0.0\n",
      "DRUG_ID            0.0\n",
      "SANGER_MODEL_ID    0.0\n",
      "AUC                0.0\n",
      "RMSE               0.0\n",
      "Z_SCORE            0.0\n",
      "LN_IC50            0.0\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID SANGER_MODEL_ID       AUC  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1       SIDM00636  0.982114   \n",
       "1   GDSC1            ES3  Erlotinib        1       SIDM00265  0.984816   \n",
       "2   GDSC1            ES5  Erlotinib        1       SIDM00263  0.985693   \n",
       "\n",
       "       RMSE   Z_SCORE   LN_IC50  \n",
       "0  0.022521 -0.189576  2.395685  \n",
       "1  0.031840  0.508635  3.140923  \n",
       "2  0.026052  1.284229  3.968757  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COLS_TO_KEEP = ['DATASET', 'CELL_LINE_NAME', 'DRUG_NAME', 'DRUG_ID', \n",
    "                'SANGER_MODEL_ID', 'AUC', 'RMSE', 'Z_SCORE', 'LN_IC50']\n",
    "gdsc_base = gdsc_ic50s_join[COLS_TO_KEEP].drop_duplicates()\n",
    "print(gdsc_base.shape)\n",
    "# Percent of NaN values per column.\n",
    "print(\"Missing rates:\")\n",
    "print(\"==============\")\n",
    "print(100 * gdsc_base.isna().sum() / gdsc_base.shape[0])\n",
    "gdsc_base.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC1    70.0\n",
       "GDSC2    30.0\n",
       "Name: DATASET, dtype: float64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Occurence rate for each GDSC dataset.\n",
    "round(100 * gdsc_base.DATASET.value_counts() / gdsc_base.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save drug response matrix.\n",
    "gdsc_base.to_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1002, 13)\n",
      "|    | Sample Name   |   COSMIC identifier | Whole Exome Sequencing (WES)   | Copy Number Alterations (CNA)   | Gene Expression   | Methylation   | Drug       | GDSC                  | GDSC           | Cancer Type             | Microsatellite             | Screen Medium   | Growth Properties   |\n",
      "|    |               |                     |                                |                                 |                   |               | Response   | Tissue descriptor 1   | Tissue         | (matching TCGA label)   | instability Status (MSI)   |                 |                     |\n",
      "|    |               |                     |                                |                                 |                   |               |            |                       | descriptor 2   |                         |                            |                 |                     |\n",
      "|---:|:--------------|--------------------:|:-------------------------------|:--------------------------------|:------------------|:--------------|:-----------|:----------------------|:---------------|:------------------------|:---------------------------|:----------------|:--------------------|\n",
      "|  0 | A253          |              906794 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | nan                     | MSS/MSI-L                  | D/F12           | Adherent            |\n",
      "|  1 | BB30-HNC      |              753531 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | HNSC                    | MSS/MSI-L                  | D/F12           | Adherent            |\n",
      "|  2 | BB49-HNC      |              753532 | Y                              | Y                               | Y                 | Y             | Y          | aero_dig_tract        | head and neck  | HNSC                    | MSS/MSI-L                  | D/F12           | Adherent            |\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/cwoest/Applications/anaconda3/envs/master-thesis-log/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "cl_details = pd.read_excel(RAW_DATA_PATH + 'Cell_Lines_Details.xlsx')\n",
    "print(cl_details.shape)\n",
    "print(cl_details.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17737, 1020)\n",
      "|    | GENE_SYMBOLS   | GENE_title                                   |   DATA.906826 |   DATA.687983 |   DATA.910927 |   DATA.1240138 |   DATA.1240139 |   DATA.906792 |   DATA.910688 |   DATA.1240135 |\n",
      "|---:|:---------------|:---------------------------------------------|--------------:|--------------:|--------------:|---------------:|---------------:|--------------:|--------------:|---------------:|\n",
      "|  0 | TSPAN6         | tetraspanin 6 [Source:HGNC Symbol;Acc:11858] |       7.63202 |       7.54867 |       8.71234 |        7.79714 |        7.72927 |       7.07453 |        3.2852 |        6.96161 |\n",
      "|  1 | TNMD           | tenomodulin [Source:HGNC Symbol;Acc:17757]   |       2.96459 |       2.77772 |       2.64351 |        2.81792 |        2.95774 |       2.88968 |        2.8282 |        2.87475 |\n"
     ]
    }
   ],
   "source": [
    "gene_expression = pd.read_csv(RAW_DATA_PATH + 'Cell_line_RMA_proc_basalExp.txt', sep=\"\\t\")\n",
    "print(gene_expression.shape)\n",
    "print(gene_expression.iloc[:2, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: 50 Cosmic IDs not found in cell annotation data: \n",
      "['906815', '1330955', '907284', '1330944', '1330943', '1298232', '1299076', '1299051', '906829', '1290773', '907049', '909730', '910946', '907391', '687509', '1327761', '753536', '1240152', '925342', '687814', '1479994', '1479991', '1298355', '1331026', '1298150', '1479993', '11223344', '1330945', '1479992', '998179', '1331060', '1331028', '1503362.1', '906850', '1240211', '1299049', '908119', '1331031', '907785', '1330983.1', '906867', '1723793', '1723794', '1240156', '1299067', '906803', '1298154', '909976.1', '1659787', '905954.1']\n",
      "(1018, 17737)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/cwoest/Applications/anaconda3/envs/master-thesis-log/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample Name</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>LINC00526</th>\n",
       "      <th>PPY2</th>\n",
       "      <th>nan</th>\n",
       "      <th>nan</th>\n",
       "      <th>KRT18P55</th>\n",
       "      <th>nan</th>\n",
       "      <th>POLRMTP1</th>\n",
       "      <th>UBL5P2</th>\n",
       "      <th>TBC1D3P5</th>\n",
       "      <th>nan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAL-120</th>\n",
       "      <td>7.632023</td>\n",
       "      <td>2.964585</td>\n",
       "      <td>10.379553</td>\n",
       "      <td>3.614794</td>\n",
       "      <td>3.380681</td>\n",
       "      <td>3.324692</td>\n",
       "      <td>3.566350</td>\n",
       "      <td>8.204530</td>\n",
       "      <td>5.235118</td>\n",
       "      <td>5.369039</td>\n",
       "      <td>...</td>\n",
       "      <td>6.786925</td>\n",
       "      <td>2.997054</td>\n",
       "      <td>3.109774</td>\n",
       "      <td>7.882377</td>\n",
       "      <td>3.331134</td>\n",
       "      <td>2.852537</td>\n",
       "      <td>3.130696</td>\n",
       "      <td>9.986616</td>\n",
       "      <td>3.073724</td>\n",
       "      <td>7.284733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMS-114</th>\n",
       "      <td>7.548671</td>\n",
       "      <td>2.777716</td>\n",
       "      <td>11.807341</td>\n",
       "      <td>4.066887</td>\n",
       "      <td>3.732485</td>\n",
       "      <td>3.152404</td>\n",
       "      <td>7.827172</td>\n",
       "      <td>6.616972</td>\n",
       "      <td>5.809264</td>\n",
       "      <td>7.209653</td>\n",
       "      <td>...</td>\n",
       "      <td>5.317911</td>\n",
       "      <td>3.263745</td>\n",
       "      <td>3.059424</td>\n",
       "      <td>8.681302</td>\n",
       "      <td>2.992611</td>\n",
       "      <td>2.776771</td>\n",
       "      <td>3.260982</td>\n",
       "      <td>9.002814</td>\n",
       "      <td>3.000182</td>\n",
       "      <td>8.504804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAL-51</th>\n",
       "      <td>8.712338</td>\n",
       "      <td>2.643508</td>\n",
       "      <td>9.880733</td>\n",
       "      <td>3.956230</td>\n",
       "      <td>3.236620</td>\n",
       "      <td>3.241246</td>\n",
       "      <td>2.931034</td>\n",
       "      <td>8.191246</td>\n",
       "      <td>5.426841</td>\n",
       "      <td>5.120747</td>\n",
       "      <td>...</td>\n",
       "      <td>3.143006</td>\n",
       "      <td>3.112145</td>\n",
       "      <td>2.930254</td>\n",
       "      <td>8.707886</td>\n",
       "      <td>2.886574</td>\n",
       "      <td>2.685307</td>\n",
       "      <td>3.176239</td>\n",
       "      <td>9.113243</td>\n",
       "      <td>2.916274</td>\n",
       "      <td>7.059092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 17737 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample Name    TSPAN6      TNMD       DPM1     SCYL3  C1orf112       FGR  \\\n",
       "CAL-120      7.632023  2.964585  10.379553  3.614794  3.380681  3.324692   \n",
       "DMS-114      7.548671  2.777716  11.807341  4.066887  3.732485  3.152404   \n",
       "CAL-51       8.712338  2.643508   9.880733  3.956230  3.236620  3.241246   \n",
       "\n",
       "Sample Name       CFH     FUCA2      GCLC      NFYA  ...  LINC00526      PPY2  \\\n",
       "CAL-120      3.566350  8.204530  5.235118  5.369039  ...   6.786925  2.997054   \n",
       "DMS-114      7.827172  6.616972  5.809264  7.209653  ...   5.317911  3.263745   \n",
       "CAL-51       2.931034  8.191246  5.426841  5.120747  ...   3.143006  3.112145   \n",
       "\n",
       "Sample Name       nan       nan  KRT18P55       nan  POLRMTP1    UBL5P2  \\\n",
       "CAL-120      3.109774  7.882377  3.331134  2.852537  3.130696  9.986616   \n",
       "DMS-114      3.059424  8.681302  2.992611  2.776771  3.260982  9.002814   \n",
       "CAL-51       2.930254  8.707886  2.886574  2.685307  3.176239  9.113243   \n",
       "\n",
       "Sample Name  TBC1D3P5       nan  \n",
       "CAL-120      3.073724  7.284733  \n",
       "DMS-114      3.000182  8.504804  \n",
       "CAL-51       2.916274  7.059092  \n",
       "\n",
       "[3 rows x 17737 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# from src.preprocess.build_drug_response_matrix import (\n",
    "#     get_gdsc_gene_expression, cosmic_ids_to_cell_line_names\n",
    "# )\n",
    "\n",
    "CELL_LINE_DETAILS_FILE = 'Cell_Lines_Details.xlsx'\n",
    "GENE_EXPRESSION_RAW_FILE = 'Cell_line_RMA_proc_basalExp.txt'\n",
    "\n",
    "# Return the gene expression dataframe(n_cells x n_genes) for a set of gene symbols for all cell_lines of the GDSC cell line annotation file.\n",
    "# If the genes are None, return the data for all genes.\n",
    "gene_expr = get_gdsc_gene_expression(\n",
    "    path_cell_annotations=RAW_DATA_PATH + CELL_LINE_DETAILS_FILE,\n",
    "    path_gene_expression=RAW_DATA_PATH + GENE_EXPRESSION_RAW_FILE \n",
    ")\n",
    "print(gene_expr.shape)\n",
    "gene_expr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 7)\n",
      "|    |   Entrez ID | Symbol   | Name                   | Gene Family                | Type     |   RNA-Seq Correlation |   RNA-Seq Correlation Self-Rank |\n",
      "|---:|------------:|:---------|:-----------------------|:---------------------------|:---------|----------------------:|--------------------------------:|\n",
      "|  0 |        3638 | INSIG1   | insulin induced gene 1 | nan                        | landmark |                   nan |                             nan |\n",
      "|  1 |        2309 | FOXO3    | forkhead box O3        | Forkhead boxes             | landmark |                   nan |                             nan |\n",
      "|  2 |        1001 | CDH3     | cadherin 3             | Type I classical cadherins | landmark |                   nan |                             nan |\n"
     ]
    }
   ],
   "source": [
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "print(landmark_genes.head(3).to_markdown())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1018, 908)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>LPAR2</th>\n",
       "      <th>PEX11A</th>\n",
       "      <th>MAPKAPK2</th>\n",
       "      <th>RSU1</th>\n",
       "      <th>FIS1</th>\n",
       "      <th>PRKACA</th>\n",
       "      <th>BZW2</th>\n",
       "      <th>CD40</th>\n",
       "      <th>TLR4</th>\n",
       "      <th>HOMER2</th>\n",
       "      <th>...</th>\n",
       "      <th>CXCR4</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>TBX2</th>\n",
       "      <th>RRP1B</th>\n",
       "      <th>MLLT11</th>\n",
       "      <th>UBE3B</th>\n",
       "      <th>VAPB</th>\n",
       "      <th>CDK5R1</th>\n",
       "      <th>STAT5B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAL-120</th>\n",
       "      <td>5.519982</td>\n",
       "      <td>4.759168</td>\n",
       "      <td>4.300245</td>\n",
       "      <td>8.369014</td>\n",
       "      <td>9.344897</td>\n",
       "      <td>3.317099</td>\n",
       "      <td>9.191766</td>\n",
       "      <td>3.582643</td>\n",
       "      <td>2.993602</td>\n",
       "      <td>3.272433</td>\n",
       "      <td>...</td>\n",
       "      <td>2.835618</td>\n",
       "      <td>3.189415</td>\n",
       "      <td>6.167946</td>\n",
       "      <td>3.106098</td>\n",
       "      <td>7.621849</td>\n",
       "      <td>5.515228</td>\n",
       "      <td>4.080860</td>\n",
       "      <td>4.888738</td>\n",
       "      <td>4.164141</td>\n",
       "      <td>3.608917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DMS-114</th>\n",
       "      <td>6.541490</td>\n",
       "      <td>5.977963</td>\n",
       "      <td>3.612725</td>\n",
       "      <td>9.685341</td>\n",
       "      <td>9.122008</td>\n",
       "      <td>3.529718</td>\n",
       "      <td>6.874194</td>\n",
       "      <td>6.943208</td>\n",
       "      <td>3.053156</td>\n",
       "      <td>4.044992</td>\n",
       "      <td>...</td>\n",
       "      <td>3.507280</td>\n",
       "      <td>3.174291</td>\n",
       "      <td>5.271099</td>\n",
       "      <td>7.758626</td>\n",
       "      <td>8.088948</td>\n",
       "      <td>7.860457</td>\n",
       "      <td>4.212928</td>\n",
       "      <td>4.780611</td>\n",
       "      <td>4.723101</td>\n",
       "      <td>4.345699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAL-51</th>\n",
       "      <td>7.489398</td>\n",
       "      <td>4.040322</td>\n",
       "      <td>4.888300</td>\n",
       "      <td>8.332596</td>\n",
       "      <td>9.567052</td>\n",
       "      <td>3.272170</td>\n",
       "      <td>9.779323</td>\n",
       "      <td>3.720286</td>\n",
       "      <td>2.802697</td>\n",
       "      <td>3.619598</td>\n",
       "      <td>...</td>\n",
       "      <td>2.971266</td>\n",
       "      <td>3.671769</td>\n",
       "      <td>5.784561</td>\n",
       "      <td>2.860879</td>\n",
       "      <td>8.447288</td>\n",
       "      <td>4.578649</td>\n",
       "      <td>4.108093</td>\n",
       "      <td>4.850897</td>\n",
       "      <td>2.921327</td>\n",
       "      <td>4.064676</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 908 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "CELL_LINE_NAME     LPAR2    PEX11A  MAPKAPK2      RSU1      FIS1    PRKACA  \\\n",
       "CAL-120         5.519982  4.759168  4.300245  8.369014  9.344897  3.317099   \n",
       "DMS-114         6.541490  5.977963  3.612725  9.685341  9.122008  3.529718   \n",
       "CAL-51          7.489398  4.040322  4.888300  8.332596  9.567052  3.272170   \n",
       "\n",
       "CELL_LINE_NAME      BZW2      CD40      TLR4    HOMER2  ...     CXCR4  \\\n",
       "CAL-120         9.191766  3.582643  2.993602  3.272433  ...  2.835618   \n",
       "DMS-114         6.874194  6.943208  3.053156  4.044992  ...  3.507280   \n",
       "CAL-51          9.779323  3.720286  2.802697  3.619598  ...  2.971266   \n",
       "\n",
       "CELL_LINE_NAME    NOTCH1   PLEKHM1      TBX2     RRP1B    MLLT11     UBE3B  \\\n",
       "CAL-120         3.189415  6.167946  3.106098  7.621849  5.515228  4.080860   \n",
       "DMS-114         3.174291  5.271099  7.758626  8.088948  7.860457  4.212928   \n",
       "CAL-51          3.671769  5.784561  2.860879  8.447288  4.578649  4.108093   \n",
       "\n",
       "CELL_LINE_NAME      VAPB    CDK5R1    STAT5B  \n",
       "CAL-120         4.888738  4.164141  3.608917  \n",
       "DMS-114         4.780611  4.723101  4.345699  \n",
       "CAL-51          4.850897  2.921327  4.064676  \n",
       "\n",
       "[3 rows x 908 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Choose only the cell-line columns of the gene expressions table that are in the landmark gene file.\n",
    "inter_cols = list(set(gene_expr.columns).intersection(set(landmark_genes.Symbol)))\n",
    "gene_expr_sparse = gene_expr[inter_cols]\n",
    "gene_expr_sparse.columns.rename('CELL_LINE_NAME', inplace=True)\n",
    "print(gene_expr_sparse.shape)\n",
    "gene_expr_sparse.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 916)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>LPAR2</th>\n",
       "      <th>PEX11A</th>\n",
       "      <th>...</th>\n",
       "      <th>CXCR4</th>\n",
       "      <th>NOTCH1</th>\n",
       "      <th>PLEKHM1</th>\n",
       "      <th>TBX2</th>\n",
       "      <th>RRP1B</th>\n",
       "      <th>MLLT11</th>\n",
       "      <th>UBE3B</th>\n",
       "      <th>VAPB</th>\n",
       "      <th>CDK5R1</th>\n",
       "      <th>STAT5B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>5.377255</td>\n",
       "      <td>4.000075</td>\n",
       "      <td>...</td>\n",
       "      <td>9.345843</td>\n",
       "      <td>3.631735</td>\n",
       "      <td>6.312806</td>\n",
       "      <td>3.169560</td>\n",
       "      <td>8.001636</td>\n",
       "      <td>5.106525</td>\n",
       "      <td>3.911568</td>\n",
       "      <td>5.059999</td>\n",
       "      <td>6.212202</td>\n",
       "      <td>5.440448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>6.288406</td>\n",
       "      <td>5.435063</td>\n",
       "      <td>...</td>\n",
       "      <td>3.005560</td>\n",
       "      <td>3.435098</td>\n",
       "      <td>6.188079</td>\n",
       "      <td>2.935917</td>\n",
       "      <td>8.226306</td>\n",
       "      <td>7.938137</td>\n",
       "      <td>4.489525</td>\n",
       "      <td>5.751381</td>\n",
       "      <td>4.490407</td>\n",
       "      <td>3.992091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>6.166101</td>\n",
       "      <td>6.351613</td>\n",
       "      <td>...</td>\n",
       "      <td>3.852866</td>\n",
       "      <td>3.490747</td>\n",
       "      <td>6.345090</td>\n",
       "      <td>3.025210</td>\n",
       "      <td>8.219939</td>\n",
       "      <td>8.871089</td>\n",
       "      <td>4.315447</td>\n",
       "      <td>5.074429</td>\n",
       "      <td>5.300401</td>\n",
       "      <td>4.583204</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 916 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50     LPAR2    PEX11A  ...     CXCR4    NOTCH1   PLEKHM1      TBX2  \\\n",
       "0  2.395685  5.377255  4.000075  ...  9.345843  3.631735  6.312806  3.169560   \n",
       "1  3.140923  6.288406  5.435063  ...  3.005560  3.435098  6.188079  2.935917   \n",
       "2  3.968757  6.166101  6.351613  ...  3.852866  3.490747  6.345090  3.025210   \n",
       "\n",
       "      RRP1B    MLLT11     UBE3B      VAPB    CDK5R1    STAT5B  \n",
       "0  8.001636  5.106525  3.911568  5.059999  6.212202  5.440448  \n",
       "1  8.226306  7.938137  4.489525  5.751381  4.490407  3.992091  \n",
       "2  8.219939  8.871089  4.315447  5.074429  5.300401  4.583204  \n",
       "\n",
       "[3 rows x 916 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_geneexpr = gdsc_base.merge(right=gene_expr_sparse,\n",
    "                                     left_on=['CELL_LINE_NAME'],\n",
    "                                     right_index=True,\n",
    "                                     how='left',\n",
    "                                     suffixes=['_gdsc', '_geneexpr'])\n",
    "print(join_gdsc_geneexpr.shape)\n",
    "join_gdsc_geneexpr.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the gene expression information to a file.\n",
    "join_gdsc_geneexpr.to_pickle(PROCESSED_DATA_PATH + 'gexpr_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy number variation - picnic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `cnv_abs_copy_number_picnic_20191101.csv` took 3.31045 seconds to import. \n",
      "Shape: (24502, 988)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>M14</th>\n",
       "      <th>TE-12</th>\n",
       "      <th>TMK-1</th>\n",
       "      <th>STS-0421</th>\n",
       "      <th>PL4</th>\n",
       "      <th>PCI-4B</th>\n",
       "      <th>PCI-30</th>\n",
       "      <th>HSC-39</th>\n",
       "      <th>...</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>MMAc-SF</th>\n",
       "      <th>BE-13</th>\n",
       "      <th>MC-IXC</th>\n",
       "      <th>Ramos-2G6-4C10</th>\n",
       "      <th>CGTH-W-1</th>\n",
       "      <th>H9</th>\n",
       "      <th>GR-ST</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>MM1S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG00001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG00003</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG00004</td>\n",
       "      <td>A2M</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 988 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_ID GENE_SYMBOL  M14  TE-12  TMK-1  STS-0421  PL4  PCI-4B  PCI-30  \\\n",
       "1  SIDG00001        A1BG  3.0    3.0    3.0       4.0  6.0     3.0     2.0   \n",
       "2  SIDG00003        A1CF  3.0    3.0    3.0       4.0  3.0     3.0     3.0   \n",
       "3  SIDG00004         A2M  3.0    3.0    2.0       4.0  9.0     3.0     3.0   \n",
       "\n",
       "   HSC-39  ...  451Lu  MMAc-SF  BE-13  MC-IXC  Ramos-2G6-4C10  CGTH-W-1   H9  \\\n",
       "1     3.0  ...    4.0      3.0    4.0     2.0             2.0       4.0  4.0   \n",
       "2     3.0  ...    4.0      2.0    3.0     1.0             2.0       3.0  4.0   \n",
       "3     3.0  ...    5.0      3.0    4.0     2.0             2.0       4.0  3.0   \n",
       "\n",
       "   GR-ST  YMB-1-E  MM1S  \n",
       "1    2.0      3.0   2.0  \n",
       "2    2.0      3.0   2.0  \n",
       "3    2.0      5.0   2.0  \n",
       "\n",
       "[3 rows x 988 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "cnv_picnic_v0 = pd.read_csv(RAW_DATA_PATH + 'cnv_abs_copy_number_picnic_20191101.csv', sep=\",\", header=1)\n",
    "cnv_picnic_v0.rename(columns={\n",
    "    'Unnamed: 1': 'GENE_SYMBOL',\n",
    "    'model_name': 'GENE_ID'}, inplace=True)\n",
    "cnv_picnic_v0 = cnv_picnic_v0.iloc[1:, :]\n",
    "print(f\"File `cnv_abs_copy_number_picnic_20191101.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {cnv_picnic_v0.shape}\")\n",
    "cnv_picnic_v0.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | GENE_ID   | GENE_SYMBOL   |   M14 |   TE-12 |   TMK-1 |   STS-0421 |   PL4 |   PCI-4B |   PCI-30 |   HSC-39 |\n",
      "|---:|:----------|:--------------|------:|--------:|--------:|-----------:|------:|---------:|---------:|---------:|\n",
      "|  1 | SIDG00001 | A1BG          |     3 |       3 |       3 |          4 |     6 |        3 |        2 |        3 |\n",
      "|  2 | SIDG00003 | A1CF          |     3 |       3 |       3 |          4 |     3 |        3 |        3 |        3 |\n",
      "|  3 | SIDG00004 | A2M           |     3 |       3 |       2 |          4 |     9 |        3 |        3 |        3 |\n"
     ]
    }
   ],
   "source": [
    "print(cnv_picnic_v0.iloc[:3, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that here the columns are the cell-line names and the row are indentified by the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(986, 24502)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11AP1</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 24502 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      A1BG A1CF  A2M A2ML1 A2ML1-AS1 A2ML1-AS2 A3GALT2 A4GALT A4GNT AAAS  ...  \\\n",
       "M14    3.0  3.0  3.0   3.0       3.0       3.0     4.0    4.0   4.0  3.0  ...   \n",
       "TE-12  3.0  3.0  3.0   3.0       3.0       3.0     2.0    3.0   3.0  3.0  ...   \n",
       "TMK-1  3.0  3.0  2.0   2.0       2.0       2.0     2.0    2.0   3.0  3.0  ...   \n",
       "\n",
       "      ZWINT ZXDA ZXDB ZXDC ZYG11A ZYG11AP1 ZYG11B  ZYX ZZEF1 ZZZ3  \n",
       "M14     3.0  2.0  2.0  4.0    4.0      3.0    4.0  4.0   4.0  4.0  \n",
       "TE-12   3.0  2.0  2.0  3.0    3.0      3.0    3.0  3.0   3.0  3.0  \n",
       "TMK-1   3.0  2.0  2.0  1.0    3.0      3.0    3.0  3.0   2.0  3.0  \n",
       "\n",
       "[3 rows x 24502 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_picnic_v1 = cnv_picnic_v0.iloc[:, 1:].T\n",
    "cnv_picnic_v1 = cnv_picnic_v1.rename(columns=cnv_picnic_v1.iloc[0]).drop(cnv_picnic_v1.index[0])\n",
    "print(cnv_picnic_v1.shape)\n",
    "cnv_picnic_v1.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 7)\n",
      "(986, 966)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 966 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AARS ABCB6 ABCC5 ABCF1 ABCF3 ABHD4 ABHD6 ABL1 ACAA1 ACAT2  ... ZMIZ1  \\\n",
       "M14    3.0   3.0   4.0   3.0   4.0   2.0   3.0  3.0   3.0   2.0  ...   3.0   \n",
       "TE-12  4.0   3.0   3.0   2.0   3.0   4.0   2.0  4.0   2.0   2.0  ...   3.0   \n",
       "TMK-1  3.0   2.0   3.0   3.0   3.0   3.0   2.0  3.0   2.0   2.0  ...   2.0   \n",
       "\n",
       "      ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451 ZNF586 ZNF589 ZW10  \n",
       "M14     3.0    3.0    3.0    3.0    2.0    3.0    3.0    3.0  3.0  \n",
       "TE-12   4.0    5.0    3.0    2.0    3.0    2.0    3.0    2.0  3.0  \n",
       "TMK-1   2.0    5.0    3.0    3.0    2.0    3.0    3.0    2.0  4.0  \n",
       "\n",
       "[3 rows x 966 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------------- #\n",
    "# Sparse down the gene column dimension by only using landmark genes. #\n",
    "# ------------------------------------------------------------------- #\n",
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "\n",
    "# Check how many cell line columns of the cnv picnic table are in the landmark gene file.\n",
    "count, cols_to_keep = 0, []\n",
    "for c in cnv_picnic_v1.columns[cnv_picnic_v1.columns != 'nan']:\n",
    "    if c in landmark_genes.Symbol.tolist(): \n",
    "        count += 1\n",
    "        cols_to_keep.append(c)\n",
    "\n",
    "cnv_picnic_v2 = cnv_picnic_v1[cols_to_keep]\n",
    "assert cnv_picnic_v2.shape[1] == len(cols_to_keep)\n",
    "print(cnv_picnic_v2.shape)\n",
    "cnv_picnic_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 974)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 974 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50 AARS ABCB6  ... ZMIZ1 ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451  \\\n",
       "0  2.395685  2.0   2.0  ...   2.0   2.0    2.0    2.0    2.0    2.0    2.0   \n",
       "1  3.140923  2.0   2.0  ...   1.0   1.0    2.0    2.0    2.0    3.0    2.0   \n",
       "2  3.968757  3.0   4.0  ...   2.0   3.0    3.0    3.0    4.0    6.0    4.0   \n",
       "\n",
       "  ZNF586 ZNF589 ZW10  \n",
       "0    2.0    2.0  1.0  \n",
       "1    2.0    3.0  2.0  \n",
       "2    3.0    4.0  3.0  \n",
       "\n",
       "[3 rows x 974 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_cnv_picnic = gdsc_base.merge(right=cnv_picnic_v2,\n",
    "                                       left_on=['CELL_LINE_NAME'],\n",
    "                                       right_index=True,\n",
    "                                       how='left',\n",
    "                                       suffixes=['_gdsc', '_cnvp'])\n",
    "print(join_gdsc_cnv_picnic.shape)\n",
    "join_gdsc_cnv_picnic.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the copy number picnic information to a file.\n",
    "join_gdsc_cnv_picnic.to_pickle(PROCESSED_DATA_PATH + 'cnvp_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy number variation - gistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `cnv_gistic_20191101.csv` took 3.35268 seconds to import. \n",
      "Shape: (20669, 980)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>M14</th>\n",
       "      <th>TE-12</th>\n",
       "      <th>TMK-1</th>\n",
       "      <th>STS-0421</th>\n",
       "      <th>PL4</th>\n",
       "      <th>PCI-4B</th>\n",
       "      <th>PCI-30</th>\n",
       "      <th>HSC-39</th>\n",
       "      <th>...</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>MMAc-SF</th>\n",
       "      <th>BE-13</th>\n",
       "      <th>MC-IXC</th>\n",
       "      <th>Ramos-2G6-4C10</th>\n",
       "      <th>CGTH-W-1</th>\n",
       "      <th>H9</th>\n",
       "      <th>GR-ST</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>MM1S</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG00001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG00002</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG00003</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 980 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     GENE_ID GENE_SYMBOL  M14  TE-12  TMK-1  STS-0421  PL4  PCI-4B  PCI-30  \\\n",
       "1  SIDG00001        A1BG  0.0   -1.0    0.0       0.0  1.0    -1.0     0.0   \n",
       "2  SIDG00002    A1BG-AS1  0.0   -1.0    0.0       0.0  1.0    -1.0     0.0   \n",
       "3  SIDG00003        A1CF -1.0    0.0    0.0       1.0 -1.0     0.0     1.0   \n",
       "\n",
       "   HSC-39  ...  451Lu  MMAc-SF  BE-13  MC-IXC  Ramos-2G6-4C10  CGTH-W-1   H9  \\\n",
       "1     0.0  ...    0.0     -1.0    2.0     0.0             0.0       1.0  1.0   \n",
       "2     0.0  ...    0.0     -1.0    2.0     0.0             0.0       1.0  1.0   \n",
       "3     0.0  ...    0.0     -1.0   -1.0    -1.0             0.0       0.0  1.0   \n",
       "\n",
       "   GR-ST  YMB-1-E  MM1S  \n",
       "1    0.0      0.0   0.0  \n",
       "2    0.0      0.0   0.0  \n",
       "3    0.0      0.0   0.0  \n",
       "\n",
       "[3 rows x 980 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "cnv_gistic_v0 = pd.read_csv(RAW_DATA_PATH + 'cnv_gistic_20191101.csv', sep=\",\", header=1)\n",
    "cnv_gistic_v0.rename(columns={\n",
    "    'Unnamed: 1': 'GENE_SYMBOL',\n",
    "    'model_name': 'GENE_ID'}, inplace=True)\n",
    "cnv_gistic_v0 = cnv_gistic_v0.iloc[1:, :]\n",
    "print(f\"File `cnv_gistic_20191101.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {cnv_gistic_v0.shape}\")\n",
    "cnv_gistic_v0.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | GENE_ID   | GENE_SYMBOL   |   M14 |   TE-12 |   TMK-1 |   STS-0421 |   PL4 |   PCI-4B |   PCI-30 |   HSC-39 |\n",
      "|---:|:----------|:--------------|------:|--------:|--------:|-----------:|------:|---------:|---------:|---------:|\n",
      "|  1 | SIDG00001 | A1BG          |     0 |      -1 |       0 |          0 |     1 |       -1 |        0 |        0 |\n",
      "|  2 | SIDG00002 | A1BG-AS1      |     0 |      -1 |       0 |          0 |     1 |       -1 |        0 |        0 |\n",
      "|  3 | SIDG00003 | A1CF          |    -1 |       0 |       0 |          1 |    -1 |        0 |        1 |        0 |\n"
     ]
    }
   ],
   "source": [
    "print(cnv_gistic_v0.iloc[:3, :10].to_markdown())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that here the columns are the cell-line names and the row are indentified by the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 20669)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZUP1</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 20669 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      A1BG A1BG-AS1 A1CF  A2M A2ML1 A2MP1 A4GALT A4GNT AAAS AACS  ... ZUP1  \\\n",
       "M14    0.0      0.0 -1.0  0.0   0.0   0.0    1.0   1.0  0.0  0.0  ... -1.0   \n",
       "TE-12 -1.0     -1.0  0.0  1.0   1.0   1.0   -1.0   0.0  0.0  0.0  ... -1.0   \n",
       "TMK-1  0.0      0.0  0.0 -1.0  -1.0  -1.0   -1.0   0.0  0.0  0.0  ... -1.0   \n",
       "\n",
       "      ZW10 ZWILCH ZWINT ZXDC ZYG11A ZYG11B  ZYX ZZEF1 ZZZ3  \n",
       "M14    0.0    1.0  -1.0  1.0    1.0    1.0  2.0   2.0  2.0  \n",
       "TE-12  1.0    1.0   1.0  0.0    0.0    0.0  1.0   0.0  0.0  \n",
       "TMK-1  1.0    0.0   1.0 -2.0    0.0    0.0  2.0  -1.0  1.0  \n",
       "\n",
       "[3 rows x 20669 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_gistic_v1 = cnv_gistic_v0.iloc[:, 1:].T\n",
    "cnv_gistic_v1 = cnv_gistic_v1.rename(columns=cnv_gistic_v1.iloc[0]).drop(cnv_gistic_v1.index[0])\n",
    "print(cnv_gistic_v1.shape)\n",
    "cnv_gistic_v1.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(978, 7)\n",
      "(978, 938)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE-12</th>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMK-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 938 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      AARS ABCB6 ABCC5 ABCF1 ABCF3 ABHD4 ABHD6 ABL1 ACAA1 ACAT2  ... ZMIZ1  \\\n",
       "M14    0.0   0.0   1.0   0.0   1.0  -1.0   0.0  0.0   0.0  -1.0  ...  -1.0   \n",
       "TE-12  1.0  -1.0   0.0  -1.0   0.0   1.0  -1.0  1.0  -1.0  -1.0  ...  -1.0   \n",
       "TMK-1  0.0  -1.0   0.0   0.0   0.0   0.0  -1.0  0.0  -1.0  -1.0  ...  -1.0   \n",
       "\n",
       "      ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451 ZNF586 ZNF589 ZW10  \n",
       "M14     0.0    0.0    0.0    0.0   -1.0    0.0    0.0    0.0  0.0  \n",
       "TE-12   1.0    1.0   -1.0   -1.0    0.0   -1.0   -1.0   -1.0  1.0  \n",
       "TMK-1  -1.0    1.0    0.0    0.0   -1.0    1.0    0.0   -1.0  1.0  \n",
       "\n",
       "[3 rows x 938 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ------------------------------------------------------------------- #\n",
    "# Sparse down the gene column dimension by only using landmark genes. #\n",
    "# ------------------------------------------------------------------- #\n",
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(landmark_genes.shape)\n",
    "\n",
    "# Check how many cell line columns of the cnv picnic table are in the landmark gene file.\n",
    "count, cols_to_keep = 0, []\n",
    "for c in cnv_gistic_v1.columns[cnv_gistic_v1.columns != 'nan']:\n",
    "    if c in landmark_genes.Symbol.tolist(): \n",
    "        count += 1\n",
    "        cols_to_keep.append(c)\n",
    "\n",
    "cnv_gistic_v2 = cnv_gistic_v1[cols_to_keep]\n",
    "assert cnv_gistic_v2.shape[1] == len(cols_to_keep)\n",
    "print(cnv_gistic_v2.shape)\n",
    "cnv_gistic_v2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 946)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>AARS</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 946 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID       AUC      RMSE   Z_SCORE  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1  0.982114  0.022521 -0.189576   \n",
       "1   GDSC1            ES3  Erlotinib        1  0.984816  0.031840  0.508635   \n",
       "2   GDSC1            ES5  Erlotinib        1  0.985693  0.026052  1.284229   \n",
       "\n",
       "    LN_IC50 AARS ABCB6  ... ZMIZ1 ZMYM2 ZNF131 ZNF274 ZNF318 ZNF395 ZNF451  \\\n",
       "0  2.395685  0.0   0.0  ...   0.0   0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "1  3.140923  0.0   0.0  ...  -1.0  -1.0    0.0    0.0    0.0    1.0    0.0   \n",
       "2  3.968757 -1.0   1.0  ...  -1.0  -1.0   -1.0   -1.0    0.0    1.0    1.0   \n",
       "\n",
       "  ZNF586 ZNF589 ZW10  \n",
       "0    0.0    0.0 -2.0  \n",
       "1    0.0    1.0  0.0  \n",
       "2   -1.0    0.0 -1.0  \n",
       "\n",
       "[3 rows x 946 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "cols_to_join_on = ['CELL_LINE_NAME']\n",
    "join_gdsc_cnv_gistic = gdsc_base.merge(right=cnv_gistic_v2,\n",
    "                                       left_on=['CELL_LINE_NAME'],\n",
    "                                       right_index=True,\n",
    "                                       how='left',\n",
    "                                       suffixes=['_gdsc', '_cnvg'])\n",
    "print(join_gdsc_cnv_gistic.shape)\n",
    "join_gdsc_cnv_gistic.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the copy number gistic information to a file.\n",
    "join_gdsc_cnv_gistic.to_pickle(PROCESSED_DATA_PATH + 'cnvg_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File `mutations_all_20220315.csv` took 20.57474 seconds to import. \n",
      "Shape: (8322616, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15  \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15  \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15  \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15  \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = time.time()\n",
    "mutations_v0 = pd.read_csv(RAW_DATA_PATH + 'mutations_all_20220315.csv', sep=\",\", header=0)\n",
    "print(f\"File `mutations_all_20220315.csv` took {time.time()-start:.5f} seconds to import. \\nShape: {mutations_v0.shape}\")\n",
    "mutations_v0.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|    | gene_id   | gene_symbol   | model_id   | protein_mutation   | rna_mutation   | cdna_mutation   | cancer_driver   | cancer_predisposition_variant   | effect   |    vaf | coding   | source   | model_name        |\n",
      "|---:|:----------|:--------------|:-----------|:-------------------|:---------------|:----------------|:----------------|:--------------------------------|:---------|-------:|:---------|:---------|:------------------|\n",
      "|  0 | SIDG36104 | SOX30         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.2703 | False    | Sanger   | HCM-SANG-1082-C15 |\n",
      "|  1 | SIDG37647 | THG1L         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.75   | False    | Sanger   | HCM-SANG-1082-C15 |\n",
      "|  2 | SIDG16920 | LSM11         | SIDM02021  | -                  | r.?            | -               | False           | False                           | intronic | 0.6452 | False    | Sanger   | HCM-SANG-1082-C15 |\n"
     ]
    }
   ],
   "source": [
    "print(mutations_v0.head(3).to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\n",
      "Shape: (8322616, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG36104</td>\n",
       "      <td>SOX30</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2703</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG37647</td>\n",
       "      <td>THG1L</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.7500</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6452</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.3207delU</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3prime_UTR_variant</td>\n",
       "      <td>0.7692</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDG16920</td>\n",
       "      <td>LSM11</td>\n",
       "      <td>SIDM02021</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>0.3571</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>HCM-SANG-1082-C15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG36104       SOX30  SIDM02021                -          r.?   \n",
       "1  SIDG37647       THG1L  SIDM02021                -          r.?   \n",
       "2  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "3  SIDG16920       LSM11  SIDM02021              p.?   r.3207delU   \n",
       "4  SIDG16920       LSM11  SIDM02021                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1             -          False                          False   \n",
       "2             -          False                          False   \n",
       "3           c.?          False                          False   \n",
       "4             -          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source         model_name  \\\n",
       "0            intronic  0.2703   False  Sanger  HCM-SANG-1082-C15   \n",
       "1            intronic  0.7500   False  Sanger  HCM-SANG-1082-C15   \n",
       "2            intronic  0.6452   False  Sanger  HCM-SANG-1082-C15   \n",
       "3  3prime_UTR_variant  0.7692   False  Sanger  HCM-SANG-1082-C15   \n",
       "4          downstream  0.3571   False  Sanger  HCM-SANG-1082-C15   \n",
       "\n",
       "  SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "0             NaN            NaN  \n",
       "1             NaN            NaN  \n",
       "2             NaN            NaN  \n",
       "3             NaN            NaN  \n",
       "4             NaN            NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Find the CELL_LINE_NAME's per SANGER_MODEL_ID.\n",
    "celllines_per_sangermodelid = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']].groupby('SANGER_MODEL_ID')['CELL_LINE_NAME'].nunique()\n",
    "counts_per_sangermodelid = celllines_per_sangermodelid.values\n",
    "\n",
    "assert (counts_per_sangermodelid != 1).any() == False\n",
    "assert (counts_per_sangermodelid == 1).all()\n",
    "\n",
    "# Only take the interested columns for the mapping.\n",
    "gdsc_mapping_subset = gdsc_base[['SANGER_MODEL_ID', 'CELL_LINE_NAME']]\n",
    "\n",
    "# Only take the unique SANGER_MODEL_ID's, since these have a 1-to-1 relationship to the CELL_LINE_NAME's anyways.\n",
    "gdsc_mapping_subset = gdsc_mapping_subset.groupby('SANGER_MODEL_ID').first().reset_index(level=0)\n",
    "\n",
    "# Join the CELL_LINE_NAME's onto the mutations_all dataset, based on the model_id.\n",
    "mutations_v1 = mutations_v0.merge(right=gdsc_mapping_subset,\n",
    "                                  left_on='model_id',\n",
    "                                  right_on='SANGER_MODEL_ID',\n",
    "                                  how='left')\n",
    "print(\"We joined the mutations dataset with the GDSC table to get the SANGER_MODEL_ID & CELL_LINE_NAME for each row in the mutations dataset.\")                                  \n",
    "print(f\"Shape: {mutations_v1.shape}\")\n",
    "mutations_v1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Each `CELL_LINE_NAME` has only exactly one `SANGER_MODEL_ID`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We excluded all rows which have CELL_LINE_NAME as NaN.\n",
      "Number of unique `gene_symbol`s in the mutations dataset: 20450\n",
      "Shape: (4574963, 15)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SIDG40054</td>\n",
       "      <td>UAP1</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.2857</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>JHH-6</td>\n",
       "      <td>SIDM00615</td>\n",
       "      <td>JHH-6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SIDG13175</td>\n",
       "      <td>IPO11</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.3511</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>NCI-H727</td>\n",
       "      <td>SIDM01119</td>\n",
       "      <td>NCI-H727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SIDG14638</td>\n",
       "      <td>LIN9</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>downstream</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SIDG07173</td>\n",
       "      <td>DYM</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.5294</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>CAL-51</td>\n",
       "      <td>SIDM00933</td>\n",
       "      <td>CAL-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SIDG08926</td>\n",
       "      <td>FCAMR</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.6250</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>D-336MG</td>\n",
       "      <td>SIDM00620</td>\n",
       "      <td>D-336MG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "19  SIDG40054        UAP1  SIDM00615                -          r.?   \n",
       "20  SIDG13175       IPO11  SIDM01119                -          r.?   \n",
       "21  SIDG14638        LIN9  SIDM00933                -          r.?   \n",
       "22  SIDG07173         DYM  SIDM00933                -          r.?   \n",
       "23  SIDG08926       FCAMR  SIDM00620                -          r.?   \n",
       "\n",
       "   cdna_mutation  cancer_driver  cancer_predisposition_variant      effect  \\\n",
       "19             -          False                          False    intronic   \n",
       "20             -          False                          False    intronic   \n",
       "21             -          False                          False  downstream   \n",
       "22             -          False                          False    intronic   \n",
       "23             -          False                          False    intronic   \n",
       "\n",
       "       vaf  coding  source model_name SANGER_MODEL_ID CELL_LINE_NAME  \n",
       "19  0.2857   False  Sanger      JHH-6       SIDM00615          JHH-6  \n",
       "20  0.3511   False  Sanger   NCI-H727       SIDM01119       NCI-H727  \n",
       "21  1.0000   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "22  0.5294   False  Sanger     CAL-51       SIDM00933         CAL-51  \n",
       "23  0.6250   False  Sanger    D-336MG       SIDM00620        D-336MG  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " # Only taking the ones with a CELL_LINE_NAME.\n",
    "mutations_v2 = mutations_v1[mutations_v1['CELL_LINE_NAME'].notna()]\n",
    "print(\"We excluded all rows which have CELL_LINE_NAME as NaN.\")\n",
    "print(f\"Number of unique `gene_symbol`s in the mutations dataset: {len(np.unique(mutations_v2.gene_symbol))}\")\n",
    "print(f\"Shape: {mutations_v2.shape}\")\n",
    "mutations_v2.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We use the landmark genes to sparse down the dimension.\n",
      "We merged the most recent dataset mutations_v2 with the landmark genes to sparse it down.\n",
      "Shape: (235906, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>cancer_predisposition_variant</th>\n",
       "      <th>effect</th>\n",
       "      <th>vaf</th>\n",
       "      <th>coding</th>\n",
       "      <th>source</th>\n",
       "      <th>model_name</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>intronic</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>False</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>splice_region</td>\n",
       "      <td>0.6190</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDG40702</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5prime_UTR_variant</td>\n",
       "      <td>0.4118</td>\n",
       "      <td>False</td>\n",
       "      <td>Broad</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0  SIDG40702       VGLL4  SIDM00330                -          r.?   \n",
       "1  SIDG40702       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2  SIDG40702       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "\n",
       "  cdna_mutation  cancer_driver  cancer_predisposition_variant  \\\n",
       "0             -          False                          False   \n",
       "1           c.?          False                          False   \n",
       "2           c.?          False                          False   \n",
       "\n",
       "               effect     vaf  coding  source model_name SANGER_MODEL_ID  \\\n",
       "0            intronic  0.9926   False  Sanger      LAN-6       SIDM00330   \n",
       "1       splice_region  0.6190   False   Broad      CL-11       SIDM00944   \n",
       "2  5prime_UTR_variant  0.4118   False   Broad      SNG-M       SIDM00373   \n",
       "\n",
       "  CELL_LINE_NAME Symbol  \n",
       "0          LAN-6  VGLL4  \n",
       "1          CL-11  VGLL4  \n",
       "2          SNG-M  VGLL4  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LANDMARK_GENES_FILE = 'landmark_genes.csv' \n",
    "\n",
    "landmark_genes = pd.read_csv(RAW_DATA_PATH + LANDMARK_GENES_FILE, sep=\"\\t\")\n",
    "print(\"We use the landmark genes to sparse down the dimension.\")\n",
    "\n",
    "# Take only the rows which have a `gene_symbol` which is also present in the landmark genes table.\n",
    "mutations_v3 = mutations_v2.merge(right=landmark_genes['Symbol'],\n",
    "                                  left_on='gene_symbol',\n",
    "                                  right_on='Symbol')\n",
    "print(\"We merged the most recent dataset mutations_v2 with the landmark genes to sparse it down.\")                                  \n",
    "print(f\"Shape: {mutations_v3.shape}\")\n",
    "mutations_v3.head(3)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We only took a chosen set of columns.\n",
      "Number of unique ...\n",
      "    CELL_LINE_NAME's   :    983 (0.42% out of all rows)\n",
      "    gene_symbol's      :    956 (0.41% out of all rows)\n",
      "    protein_mutation's :  25113 (10.65% out of all rows)\n",
      "    rna_mutation's     :  28392 (12.04% out of all rows)\n",
      "    cdna_mutation's    :  22196 (9.41% out of all rows)\n",
      "    cancer_driver's    :      2 ([False  True])\n",
      "\n",
      "Shape: (235906, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>model_id</th>\n",
       "      <th>protein_mutation</th>\n",
       "      <th>rna_mutation</th>\n",
       "      <th>cdna_mutation</th>\n",
       "      <th>cancer_driver</th>\n",
       "      <th>vaf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LAN-6</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00330</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL-11</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00944</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.407-7c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.6190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNG-M</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00373</td>\n",
       "      <td>p.?</td>\n",
       "      <td>r.65c&gt;u</td>\n",
       "      <td>c.?</td>\n",
       "      <td>False</td>\n",
       "      <td>0.4118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JHOS-2</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00305</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.5698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NCI-H1436</td>\n",
       "      <td>VGLL4</td>\n",
       "      <td>SIDM00697</td>\n",
       "      <td>-</td>\n",
       "      <td>r.?</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>0.9677</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CELL_LINE_NAME gene_symbol   model_id protein_mutation rna_mutation  \\\n",
       "0          LAN-6       VGLL4  SIDM00330                -          r.?   \n",
       "1          CL-11       VGLL4  SIDM00944              p.?   r.407-7c>u   \n",
       "2          SNG-M       VGLL4  SIDM00373              p.?      r.65c>u   \n",
       "3         JHOS-2       VGLL4  SIDM00305                -          r.?   \n",
       "4      NCI-H1436       VGLL4  SIDM00697                -          r.?   \n",
       "\n",
       "  cdna_mutation  cancer_driver     vaf  \n",
       "0             -          False  0.9926  \n",
       "1           c.?          False  0.6190  \n",
       "2           c.?          False  0.4118  \n",
       "3             -          False  0.5698  \n",
       "4             -          False  0.9677  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v4 = mutations_v3[[\n",
    "    'CELL_LINE_NAME',\n",
    "    'gene_symbol',\n",
    "    'model_id',\n",
    "    'protein_mutation',\n",
    "    'rna_mutation',\n",
    "    'cdna_mutation',\n",
    "    'cancer_driver',\n",
    "    'vaf'\n",
    "]]\n",
    "print(\"We only took a chosen set of columns.\")\n",
    "n_v4 = mutations_v4.shape[0]\n",
    "uniq_cl = len(np.unique(mutations_v4.CELL_LINE_NAME))\n",
    "uniq_gs = len(np.unique(mutations_v4.gene_symbol))\n",
    "uniq_pm = len(np.unique(mutations_v4.protein_mutation))\n",
    "uniq_rna = len(np.unique(mutations_v4.rna_mutation))\n",
    "uniq_cdna = len(np.unique(mutations_v4.cdna_mutation))\n",
    "\n",
    "print(f\"\"\"Number of unique ...\n",
    "    CELL_LINE_NAME's   : {uniq_cl:6.0f} ({100*uniq_cl/n_v4:2.2f}% out of all rows)\n",
    "    gene_symbol's      : {uniq_gs:6.0f} ({100*uniq_gs/n_v4:2.2f}% out of all rows)\n",
    "    protein_mutation's : {uniq_pm:6.0f} ({100*uniq_pm/n_v4:2.2f}% out of all rows)\n",
    "    rna_mutation's     : {uniq_rna:6.0f} ({100*uniq_rna/n_v4:2.2f}% out of all rows)\n",
    "    cdna_mutation's    : {uniq_cdna:6.0f} ({100*uniq_cdna/n_v4:2.2f}% out of all rows)\n",
    "    cancer_driver's    : {len(np.unique(mutations_v4.cancer_driver)):6.0f} ({np.unique(mutations_v4.cancer_driver)})\n",
    "\"\"\")\n",
    "print(f\"Shape: {mutations_v4.shape}\")\n",
    "mutations_v4.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (983, 957)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>ABCC5</th>\n",
       "      <th>ABCF1</th>\n",
       "      <th>ABCF3</th>\n",
       "      <th>ABHD4</th>\n",
       "      <th>ABHD6</th>\n",
       "      <th>ABL1</th>\n",
       "      <th>ACAA1</th>\n",
       "      <th>ACAT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22RV1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23132-87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42-MG-BA</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>451Lu</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5637</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 957 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GENE_SYMBOL CELL_LINE_NAME  ABCB6  ABCC5  ABCF1  ABCF3  ABHD4  ABHD6  ABL1  \\\n",
       "0                    22RV1    0.0    0.0    1.0    1.0    1.0    0.0   1.0   \n",
       "1                 23132-87    0.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "2                 42-MG-BA    1.0    0.0    0.0    0.0    0.0    0.0   0.0   \n",
       "3                    451Lu    1.0    0.0    1.0    0.0    0.0    0.0   0.0   \n",
       "4                     5637    0.0    1.0    1.0    1.0    1.0    0.0   1.0   \n",
       "\n",
       "GENE_SYMBOL  ACAA1  ACAT2  ...  ZMIZ1  ZMYM2  ZNF131  ZNF274  ZNF318  ZNF395  \\\n",
       "0              0.0    0.0  ...    1.0    1.0     0.0     0.0     1.0     0.0   \n",
       "1              0.0    0.0  ...    0.0    1.0     1.0     0.0     0.0     0.0   \n",
       "2              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "3              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "4              0.0    0.0  ...    0.0    0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "GENE_SYMBOL  ZNF451  ZNF586  ZNF589  ZW10  \n",
       "0               0.0     0.0     0.0   1.0  \n",
       "1               0.0     0.0     1.0   0.0  \n",
       "2               0.0     0.0     0.0   0.0  \n",
       "3               1.0     0.0     0.0   0.0  \n",
       "4               0.0     0.0     0.0   0.0  \n",
       "\n",
       "[5 rows x 957 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_v5 = pd.pivot_table(\n",
    "    data    = mutations_v4,\n",
    "    values  = 'cancer_driver',\n",
    "    index   = ['CELL_LINE_NAME'],\n",
    "    columns = ['gene_symbol'],\n",
    "    aggfunc = np.sum,\n",
    "    dropna  = False\n",
    ")\n",
    "\n",
    "# Set mutation values: 1.0=mutation, 0.0=no_mutation\n",
    "mutations_v5[mutations_v5==0.0] = 1.0\n",
    "mutations_v5[np.isnan(mutations_v5)] = 0.0\n",
    "\n",
    "mutations_v5['CELL_LINE_NAME'] = mutations_v5.index\n",
    "mutations_v5.insert(0, 'CELL_LINE_NAME', mutations_v5.pop('CELL_LINE_NAME'))\n",
    "mutations_v5.reset_index(drop=True, inplace=True)\n",
    "mutations_v5.columns.name = 'GENE_SYMBOL'\n",
    "\n",
    "print(f\"Shape: {mutations_v5.shape}\")\n",
    "mutations_v5.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(446146, 965)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET</th>\n",
       "      <th>CELL_LINE_NAME</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>SANGER_MODEL_ID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>RMSE</th>\n",
       "      <th>Z_SCORE</th>\n",
       "      <th>LN_IC50</th>\n",
       "      <th>ABCB6</th>\n",
       "      <th>...</th>\n",
       "      <th>ZMIZ1</th>\n",
       "      <th>ZMYM2</th>\n",
       "      <th>ZNF131</th>\n",
       "      <th>ZNF274</th>\n",
       "      <th>ZNF318</th>\n",
       "      <th>ZNF395</th>\n",
       "      <th>ZNF451</th>\n",
       "      <th>ZNF586</th>\n",
       "      <th>ZNF589</th>\n",
       "      <th>ZW10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>MC-CAR</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00636</td>\n",
       "      <td>0.982114</td>\n",
       "      <td>0.022521</td>\n",
       "      <td>-0.189576</td>\n",
       "      <td>2.395685</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES3</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00265</td>\n",
       "      <td>0.984816</td>\n",
       "      <td>0.031840</td>\n",
       "      <td>0.508635</td>\n",
       "      <td>3.140923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GDSC1</td>\n",
       "      <td>ES5</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>SIDM00263</td>\n",
       "      <td>0.985693</td>\n",
       "      <td>0.026052</td>\n",
       "      <td>1.284229</td>\n",
       "      <td>3.968757</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 965 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET CELL_LINE_NAME  DRUG_NAME  DRUG_ID SANGER_MODEL_ID       AUC  \\\n",
       "0   GDSC1         MC-CAR  Erlotinib        1       SIDM00636  0.982114   \n",
       "1   GDSC1            ES3  Erlotinib        1       SIDM00265  0.984816   \n",
       "2   GDSC1            ES5  Erlotinib        1       SIDM00263  0.985693   \n",
       "\n",
       "       RMSE   Z_SCORE   LN_IC50  ABCB6  ...  ZMIZ1  ZMYM2  ZNF131  ZNF274  \\\n",
       "0  0.022521 -0.189576  2.395685    0.0  ...    0.0    1.0     0.0     0.0   \n",
       "1  0.031840  0.508635  3.140923    0.0  ...    1.0    0.0     0.0     0.0   \n",
       "2  0.026052  1.284229  3.968757    0.0  ...    0.0    1.0     0.0     0.0   \n",
       "\n",
       "   ZNF318  ZNF395  ZNF451  ZNF586  ZNF589  ZW10  \n",
       "0     0.0     0.0     1.0     0.0     0.0   0.0  \n",
       "1     0.0     0.0     0.0     0.0     0.0   0.0  \n",
       "2     0.0     0.0     0.0     0.0     0.0   0.0  \n",
       "\n",
       "[3 rows x 965 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc_base = pd.read_pickle(PROCESSED_DATA_PATH + 'drm_full.pkl')\n",
    "\n",
    "# Join the sparsed CNV data to the GDSC table.\n",
    "join_gdsc_mut= gdsc_base.merge(right=mutations_v5,\n",
    "                               on=['CELL_LINE_NAME'],\n",
    "                               how='left',\n",
    "                               suffixes=['_gdsc', '_mut'])\n",
    "print(join_gdsc_mut.shape)\n",
    "join_gdsc_mut.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the GDSC table with the mutational information to a file.\n",
    "join_gdsc_mut.to_pickle(PROCESSED_DATA_PATH + 'mut_full.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO \n",
    "\n",
    "- [x] Write generalized methods for all datasets and integrate in separate `.py` scripts\n",
    "- [ ] Later on: separate table/dataset with drug concentration for each row in the final drug-response matrix. (use raw GDSC dataset for that)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c78b81650a0bd32063743affb6953ff71b1a0dba806fbca9e2db842718495748"
  },
  "kernelspec": {
   "display_name": "Python 3.10.4 ('master-thesis-log')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
